---
title: "Plots for Figure 4"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2, 
  fig.height = 4,
  dpi = 700
)
knitr::opts_knit$set(root.dir = "/Users/burtsche/Documents/01_repos/Glycoproteomics/")
```

## Packages

```{r, message=F, warning =F, include=}
library(tidyverse)
library(ggplot2); theme_set(cowplot::theme_cowplot(font_size = 15))
library("reshape2")
library(ggrepel)
library(knitr)
library(VennDiagram)
library(Biostrings)
library(ComplexHeatmap)
library(viridis)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
filter <- dplyr::filter
```

```{r}
options(ggplot2.discrete.colour= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
options(ggplot2.discrete.fill= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))

order <- c("sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small")
pal <- c("lightgrey", "#d7301fff", "#377eb8ff", "#41ae76ff","#88419dff" ,"#ff7f00ff", "#ffe300ff","#f781bfff", "mediumseagreen", "grey")
names(pal) <- c("FP", "sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small", "hit", "no hit")

order_O <- c("O-sialylated", "O-fucosylated", "O-hexose", "O-hybrid", "O-phospho", "O-sulfated", "O-GlcNac")
pal_O <- c("lightgrey",  "#d7301fff", "#377eb8ff", "#41ae76ff","#88419dff" ,"#ff7f00ff", '#a65628ff',"#f781bfff")
names(pal_O) <- c("FP", "O-sialylated", "O-fucosylated", "O-hexose", "O-hybrid", "O-phospho", "O-sulfated", "O-GlcNac")
```

```{r}
load("data/paper/revision1/processed/2024-01-24_res_microbiome_mousebrain_Nglyco.RData")
load("data/paper/revision1/processed/2024-01-24_res_microbiome_mousebrain_FPTPP.RData")
```

# Main

## A

experimental design

## B

```{r}
ComplexHeatmap::Heatmap(
  res_microbiome_mousebrain_Nglyco$psm %>%
    select(matches(res_microbiome_mousebrain_Nglyco$TMT_info_glyco$sample)) %>%
    setNames(paste0(res_microbiome_mousebrain_Nglyco$TMT_info_glyco$microbiome, "_", res_microbiome_mousebrain_Nglyco$TMT_info_glyco$replicate)) %>%
    cor(method = "spearman"),
  col = corrplot::COL1("Reds"),
  # cluster_columns = F,
  show_column_names = FALSE,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  rect_gp = gpar(col = "white"),
  right_annotation = row_anno <- HeatmapAnnotation(
  "microbiome" = res_microbiome_mousebrain_Nglyco$TMT_info_glyco$microbiome,
  "sex" = res_microbiome_mousebrain_Nglyco$TMT_info_glyco$sex,
  col = list(
    microbiome = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"),
    sex = c("male" = "#E1BE6A", "female" = "#40B0A6")
  ),
  which = "row",
  show_annotation_name = FALSE
), 
name = "corr"
)
```


```{r}
m <- res_microbiome_mousebrain_Nglyco$psm %>%
    select(matches(res_microbiome_mousebrain_Nglyco$TMT_info_glyco$sample)) %>%
    setNames(paste0(res_microbiome_mousebrain_Nglyco$TMT_info_glyco$microbiome, "_", res_microbiome_mousebrain_Nglyco$TMT_info_glyco$replicate))

m[m == 0] <- NA
m <- drop_na(m)
m %>%  
  cor(method = "spearman") %>% 
  melt()%>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_glyco %>% mutate(Var1 = paste0(microbiome, "_", replicate)) %>%  distinct(`Var1`, `sex`)) %>% 
  mutate(sex1= sex) %>% 
  select(-sex)%>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_glyco %>% mutate(Var2 = paste0(microbiome, "_", replicate)) %>%  distinct(`Var2`, `sex`)) %>% 
  mutate(sex2= sex) %>% 
  select(-sex) %>% 
  mutate(Var1 = str_replace(Var1, "_rep\\d", ""),
         Var2 = str_replace(Var2, "_rep\\d", "")) %>% 
  filter(Var1 == Var2 & sex1==sex2) %>% 
  distinct(Var1, sex1, value) %>% 
  group_by(Var1, sex1) %>% 
  summarise(mean_corr = mean(value), min_corr = min(value))
```

## C 
both PCAs
number of peptides and proteins

```{r}
# m <- acast(psm_glyco_brains_anno_corrected, sample ~ glycoID, value.var = "cleaned_glycosignal", fun.aggregate = median, fill = 0)
# m[is.na(m)] <- 0
m <- acast(res_microbiome_mousebrain_Nglyco$psm_anno, glycoID ~ sample, value.var = "cleaned_glycosignal", fill = 0, fun.aggregate = median)
m[m== 0] <- NA
m <- m %>% 
  as.data.frame() %>% 
  drop_na()%>% 
  as.matrix()%>% 
  t()

pca <- prcomp(m)
plot(pca)
var = (pca$sdev)^2/sum(pca$sdev^2)*100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_glyco) %>%
  ggplot(aes(x = PC1, y = PC2, colour = microbiome, shape = sex)) +
  geom_point(size = 3) +
  scale_colour_manual(values  = c("germfree" = "#7D9FC5", "community" = "#C11555",  "monocolonized" = "darkseagreen")) +
  scale_shape_manual(values = c(15, 17)) +
  labs(
       x = paste0("PC1 (", as.character(round(var[1])), "%)"),
       y = paste0("PC2 (", as.character(round(var[2])), "%)"))

pca_rotation <- pca$rotation %>% as.data.frame()

rm(m, pca_rotation)
```

```{r}
m <- dcast(res_microbiome_mousebrain_Nglyco$psm_FP_anno, Protein.ID ~ sample, value.var = "quant_norm", fill = NA) %>% 
  drop_na()%>% 
  column_to_rownames("Protein.ID") %>% 
  as.matrix() %>% 
  t()

pca <- prcomp(m)
plot(pca)

var = (pca$sdev)^2/sum(pca$sdev^2)*100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_FP) %>%
  ggplot(aes(x = round(PC1), y = PC2, colour = microbiome, shape = sex)) +
  geom_point(size = 3) +
  scale_colour_manual(values  = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
  scale_shape_manual(values = c(15, 17)) +
  labs(
       x = paste0("PC1 (", as.character(round(var[1])), "%)"),
       y = paste0("PC2 (", as.character(round(var[2])), "%)"))

pca_rotation <- pca$rotation %>% as.data.frame()
rm(m, pca_rotation)
```
## D
volcano

```{r}
ggplot(res_microbiome_mousebrain_Nglyco$limma_results %>% 
       mutate(dataset = ifelse(dataset == "FP", "full proteome", "glycoproteome"),
              direction= ifelse(logFC > 0 & hit == "hit", "up", "no change"),
              direction= ifelse(logFC < 0 & hit == "hit", "down", direction)) %>% 
        mutate(contrast = str_replace(contrast, "-", "-\n"),
               contrast = factor(contrast, levels = c("monocolonized-\ngermfree", "community-\ngermfree", "community-\nmonocolonized"))) %>% 
        filter(dataset == "glycoproteome"),
       aes(x = logFC, y = -log10(adj.P.Val), colour = direction)) +
  geom_point(alpha = 0.5, stroke=0) +
  scale_colour_manual(values = c("dodgerblue4" , "lightgrey","indianred")) +
  facet_grid(dataset~contrast) +
  #theme_bw(14) +
  theme(panel.grid = element_blank()) +
  labs(x = "log2 fold-change", y = "-log10 adjusted p-value")
```

```{r}
res_microbiome_mousebrain_Nglyco$limma_results%>% 
       mutate(dataset = ifelse(dataset == "FP", "full proteome", "glycoproteome"),
              direction= ifelse(logFC > 0 & hit == "hit", "up", "no change"),
              direction= ifelse(logFC < 0 & hit == "hit", "down", direction)) %>% 
  filter(hit == "hit") %>% 
  group_by(dataset,contrast, direction) %>% 
  count()
```


## E
```{r}
res_microbiome_mousebrain_Nglyco$limma_results %>%
  #mutate(dataset = ifelse(dataset == "full proteome", "FP", "glyco")) %>%
  filter(hit == "hit") %>%
  # mutate(class =ifelse(logFC > 0, "up", "down")) %>%
  distinct(dataset, Protein.ID) %>%
  dcast(Protein.ID ~ dataset) %>%
  mutate(
    FP = ifelse(is.na(FP), FALSE, TRUE),
    glyco = ifelse(is.na(glyco), FALSE, TRUE)
  ) %>%
  ggplot() +
  ggvenn::geom_venn(
    aes(
      A = `FP`,
      B = `glyco`
    ),
    fill_color = c("darkorange", "darkblue", "darkred", "darkgreen"),
    fill_alpha = 1,
    stroke_alpha = 1,
    stroke_color = "white",
    set_name_size = 5,
    # set_names = c("HEK", "HeLa"),
    text_size = 4,
    text_color = "white",
    show_percentage = FALSE
  ) +
  coord_fixed() +
  theme_void() +
  labs(title = "Full proteome and \nglycoproteome all contrasts")
```

## F
mitochondria, resp complex
g-protein
```{r}




processes <- bind_rows(
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(grepl("Gng", Gene)) %>%
    distinct(Gene, Protein.Description) %>%
    mutate(
      biol_process =  "Gng\nproteins"
    ),
  data.frame(
    Gene = c('Ndufab1','Cox19','Slc25a14','Uqcrh','Uqcr11','Cox14','Cox7a2l','Ndufa4', 'Sdhaf4', 'Cox5b', 'Coa6', 'Sdhaf2', 'Ppp1r1a', 'Mt3', 'S100b', 'Cox17', 'Sdhd', 'Adrb1', 'Pdha2', 'Cox7a2', 'Cox7a1', 'Sdhc'),
    biol_process = "Cellular\nrespiration"
  )
)
 


hits <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "FP" ) %>% 
  mutate(ID = Gene)

col_fun = circlize::colorRamp2(c(-1.5, 0, 1.5), c("dodgerblue4", "white", "indianred"), transparency = 0)

anno <- processes %>% 
  inner_join(res_microbiome_mousebrain_Nglyco$limma_results%>% filter(dataset == "FP") %>% distinct(Gene, contrast, logFC), multiple = "all") %>% 
  mutate(ID = Gene) %>% 
  as.data.frame() %>% 
  filter(ID %in% hits$ID)%>% 
  dcast(Gene +ID  + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median) 

ComplexHeatmap::Heatmap(
  processes %>% 
  inner_join(res_microbiome_mousebrain_Nglyco$limma_results %>% 
               filter(dataset == "FP") %>%  
               distinct(Gene, contrast, logFC), multiple = "all") %>% 
              mutate(ID = Gene) %>% 
  filter(ID %in% hits$ID & contrast != "community-monocolonized") %>% 
    mutate(contrast = str_replace(contrast, "-", "-\n"), contrast = factor(contrast, levels = c("monocolonized-\ngermfree", "community-\ngermfree"))) %>% 
  acast(ID + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median),
  col = col_fun,
  cluster_rows = T, cluster_columns = F,
  show_column_names = T, 
  row_split = anno$biol_process,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  right_annotation = HeatmapAnnotation(
  "process" = anno$biol_process,
  col = list(
    process = c("Cellular\nrespiration" = "grey60", "Gng\nproteins" = "palegreen4")
  ),
  which = "row",
  show_annotation_name = FALSE
),
name = "log2 fold-change"
)
```


## G

Axon guidance
Ion channels (chennel in description)
receptors (gaba and glutam,ate)

```{r}
processes <- bind_rows(
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(grepl("Gria|Gabra|Glra|Grin|Grik|Grm|Gabbr|Gabrb|Gabrg|Gabrd", Gene)) %>%
    distinct(Gene, Protein.Description) %>%
    mutate(
      biol_process = str_extract(Protein.Description, "Glutamate|glutamate|Gamma"),
      biol_process = ifelse(grepl("Glutamate|glutamate", biol_process), "Glutamate\nreceptor", "GABA\nreceptor")
    ),
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(hit == "hit") %>%
    filter(grepl("channel", Protein.Description)) %>% distinct(Gene, Protein.Description) %>%
    mutate(biol_process = "Ion\nchannels"),
  data.frame(
    Gene = c("Sema4a", "Plxna4", "Dscam", "Grin2b", "Epha4", "Col6a1", "Ncstn", "Sema4d", "Dcc", "Dpysl2", "Plxnb3", "Plxna1", "Dag1", "Col6a2", "Itgb3", "Reln", "Grin1", "Itgb1"),
    biol_process = "Axon\nguidance"
  )
) %>% 
  mutate(biol_process = factor(biol_process, levels = c("GABA\nreceptor","Glutamate\nreceptor", "Ion\nchannels","Axon\nguidance")))
 


hits <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" & hit == "hit") %>% 
  mutate(ID = paste0(Gene,"_", Modified.Peptide))

col_fun = circlize::colorRamp2(c(-1.5, 0, 1.5), c("dodgerblue4", "white", "indianred"), transparency = 0)

anno <- processes %>% 
  inner_join(res_microbiome_mousebrain_Nglyco$limma_results%>% filter(dataset == "glyco") %>% distinct(Gene, Modified.Peptide, contrast, logFC), multiple = "all") %>% 
  mutate(ID = paste0(Gene,"_", Modified.Peptide)) %>% 
  as.data.frame() %>% 
  filter(ID %in% hits$ID)%>% 
  dcast(Gene +ID  + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median) 

ComplexHeatmap::Heatmap(
  processes %>% 
  inner_join(res_microbiome_mousebrain_Nglyco$limma_results %>% 
               filter(dataset == "glyco") %>%  
               distinct(Gene, Modified.Peptide, contrast, logFC), multiple = "all") %>% 
              mutate(ID = paste0(Gene,"_", Modified.Peptide)) %>% 
  filter(ID %in% hits$ID & contrast != "community-monocolonized") %>% 
    mutate(contrast = str_replace(contrast, "-", "-\n"), contrast = factor(contrast, levels = c("monocolonized-\ngermfree", "community-\ngermfree"))) %>% 
  acast(ID + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median),
  col = col_fun,
  cluster_rows = T,cluster_columns = F,
  show_column_names = T, 
  row_split = anno$biol_process,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  right_annotation = HeatmapAnnotation(
  "process" = anno$biol_process,
  col = list(
    process = c("GABA\nreceptor" = "goldenrod", "Glutamate\nreceptor" = "darkred", "Ion\nchannels" = "darkcyan", "Axon\nguidance" = "mediumvioletred")
  ),
  which = "row",
  show_annotation_name = FALSE
),
name = "log2 fold-change"
)
```

## H

```{r, eval = F}
hits <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" & hit == "hit") %>% 
  distinct(Gene) %>% 
  filter(!is.na(Gene))

FCs <- res_microbiome_mousebrain_Nglyco$limma_results  %>% 
  filter(dataset == "glyco"& contrast == "community-germfree") %>% 
  mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>% 
  distinct(glycoID, logFC, hit)


for (hit in c("Grin2a")){
  p <- res_microbiome_mousebrain_Nglyco$psm_anno %>% 
  filter(Gene == hit) %>% 
    mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>% 
    left_join(FCs, by = "glycoID") %>% 
    filter(!is.na(Gene)) %>% 
  ggplot(aes(x = as.factor(glyco_position), y = cleaned_glycosignal, colour = microbiome, group = reorder(Modified.Peptide, -abs(logFC)))) +
  #geom_boxplot() + 
  geom_point(position =position_dodge(width = 0.7), alpha = 0.5, size =2) +
  theme(legend.position = "right") +
  scale_colour_manual(values  = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
  facet_wrap(~Gene, scales = "free", ncol = 2)  +
  labs(x = "glycosite positon", y = "log2 normalized\nglycopeptide intensity",
       title = unique(na.omit(res_microbiome_mousebrain_Nglyco$limma_results$Protein.Description[res_microbiome_mousebrain_Nglyco$limma_results$Gene == hit])))
  plot(p)
}

for (hit in c("Podxl2")){
  p <- res_microbiome_mousebrain_Nglyco$psm_anno %>% 
  filter(Gene == hit) %>% 
    mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>% 
    left_join(FCs, by = "glycoID") %>% 
    filter(!is.na(Gene)) %>% 
  ggplot(aes(x = as.factor(glyco_position), y = cleaned_glycosignal, colour = microbiome, group = reorder(Modified.Peptide, -abs(logFC)))) +
  #geom_boxplot() + 
  geom_point(position =position_dodge(width = 0.7), alpha = 0.5, size =2) +
  theme(legend.position = "right") +
  scale_colour_manual(values  = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
  facet_wrap(~Gene, scales = "free", ncol = 2)  +
  labs(x = "glycosite positon", y = "log2 normalized\nglycopeptide intensity",
       title = unique(na.omit(res_microbiome_mousebrain_Nglyco$limma_results$Protein.Description[res_microbiome_mousebrain_Nglyco$limma_results$Gene == hit]))) +
    theme(legend.position = "none")
  plot(p)
}


rm(hits, hit, p)
```


## I

structure

## J

```{r}
res_microbiome_mousebrain_Nglyco$limma_results %>% 
  mutate(glyco_position = n_position)%>% 
  filter(Gene == "Slc1a2" & contrast != "community-monocolonized") %>% 
  mutate(direction= ifelse(logFC > 0, "1", "-1")) %>% 
  distinct(Modified.Peptide, Observed.Modifications, glyco_position, hit,direction)%>%
  group_by(glyco_position, hit, direction) %>% 
  count() %>% 
  drop_na() %>%
  group_by(glyco_position) %>% 
  mutate(sum = sum(n), fraction = n/sum) %>% 
  filter(hit == "hit") %>% 
  mutate(n = n*as.numeric(direction), label = abs(n)) %>% 
  ggplot(aes(x = factor(glyco_position), y = n, fill = direction)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept =  0) +
  geom_text(aes(label = label)) +
  #geom_text(aes(label = round(fraction, 3))) +
  scale_fill_manual(values = c("dodgerblue4", "indianred3"))+
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())+
  coord_flip() +
  labs(x = "", y = "")

```

## K

```{r}
hits <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" & hit == "hit") %>% 
  ungroup() %>% 
  distinct(Gene) %>% 
  filter(Gene %in% res_fpTPP_mousebrain_microbiome$AUCs$Gene)

no_hits <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco") %>% 
  distinct(Gene) %>% 
  filter(Gene %in% res_fpTPP_mousebrain_microbiome$AUCs$Gene & !(Gene %in% hits$Gene))

library(org.Mm.eg.db)
#GO:0015630 microtub cy
# 0007010
cytoskeleton <- AnnotationDbi::select(org.Mm.eg.db, keytype="GOALL", keys=c("GO:0005198"), columns="UNIPROT")

res_fpTPP_mousebrain_microbiome$AUCs %>% 
  filter(absAUC < 5) %>% 
  mutate(anno= ifelse(Gene %in% hits$Gene, "microbiome\nhit", "other"),
          anno= ifelse(Gene %in% no_hits$Gene, "microbiome\nno hit", anno),
         anno= ifelse(ID %in% cytoskeleton$UNIPROT, "structural\nmolecule\nactivity", anno)) %>% 
  mutate(anno = factor(anno,levels = c("microbiome\nhit", "microbiome\nno hit", "other",  "structural\nmolecule\nactivity"))) %>% 
  ggplot(aes(x = anno, y = mean_AUC)) +
  geom_hline(yintercept = 0) +
  ggforce::geom_sina(size = 3, alpha = 0.3, stroke = NA, colour = "grey") +
    stat_summary(aes(anno),fun = median, geom = "crossbar",
               position = position_dodge(0.9), colour = "black") +
  stat_compare_means(comparisons = list( c("microbiome\nno hit", "microbiome\nhit"),c("microbiome\nno hit", "other"), c("microbiome\nhit", "other"), c("structural\nmolecule\nactivity", "other")), label = "p.signif") +
  theme(legend.position = "none")+
  labs(y = "deltaAUC community-germfree\n(zoomed in)", x = "")
```


## L

```{r}
res_fpTPP_mousebrain_microbiome$FP_irs_norm_sclaed %>%
  filter(Gene == "Cntnap1") %>% 
  separate(sample, into = c("group", "replicate"), sep = "_") %>% 
    ggplot(aes(x = as.numeric(temperature), 
               y = quant_scaled, 
               colour = group, fill = group, shape = replicate, group = paste(replicate, group))) +
  geom_hline(yintercept = c(1), linetype = 3) +
    geom_point() +
    #geom_line() +
    geom_smooth(alpha = 0.2,  method = 'loess', formula = 'y ~ x') +
    scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555")) +
   scale_fill_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555")) +
  labs(x = "temperature (C)", y = "soluble fraction", subtitle = "Cntnap1")
```


## M


```{r}
res_fpTPP_mousebrain_microbiome$FP_irs_norm_sclaed %>%
  #filter(grepl("Dnm1$|Dnm2|Dnm3", Gene)) %>% 
  #filter(grepl("Myl6$|Myl4|Myl12b", Gene)) %>% 
  #filter(grepl("Myh10|Myh11|Myh9", Gene)) %>% 
  #filter(grepl("Actb$|Acta2$|Actg1$", Gene)) %>% 
  filter(grepl("Tuba1b$|Tuba4a|Tuba8|Tubal3|Tubb1|Tubb2a|Tubb2b|Tubb3|Tubb4a|Tubb4b|Tubb5|Tubb6|Tubg1|Tubg2|Actb$|Acta2$|Actg1$|Myh10|Myh11|Myh9$|Myl12b|Dnm1$|Dnm2|Dnm3", Gene)) %>% 
  mutate(tmp = ifelse(grepl("Tuba", Gene), "Tub alpha", "Tub gamma"),
         tmp = ifelse(grepl("Tubb", Gene), "Tub beta", tmp),
         tmp = ifelse(grepl("Actb$|Acta2$|Actg1$", Gene), "Actins", tmp),
          tmp = ifelse(grepl("Myh10|Myh11|Myh9|Myl12b", Gene), "Myosins", tmp),
          tmp = ifelse(grepl("Dnm1$|Dnm2|Dnm3", Gene), "Dyneins", tmp)) %>% 
  separate(sample, into = c("group", "replicate"), sep = "_") %>% 
    ggplot(aes(x = as.numeric(temperature), 
               y = quant_scaled, 
               colour = group,
               fill =group,
               #shape = tmp, 
               group = paste(Gene, group))) +
  geom_hline(yintercept = 1, linetype =3) +
    geom_point(size = 0.5) +
    geom_smooth(size = 0.5,  method = 'loess', formula = 'y ~ x', se = T, alpha = 0.2) +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555")) +
   scale_fill_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555")) +
  facet_wrap(~tmp) +
  cowplot::panel_border() +
  labs(x = "temperature (C)", y = "soluble fraction")
```


# supplementary

## A

```{r}
ComplexHeatmap::Heatmap(
  res_microbiome_mousebrain_Nglyco$psm_FP %>%
    select(Protein.ID, matches(res_microbiome_mousebrain_Nglyco$TMT_info_FP$sample)) %>%
    group_by(Protein.ID) %>%
    summarise(across(everything(), list(sum))) %>%
    select(-Protein.ID) %>%
    setNames(paste0(res_microbiome_mousebrain_Nglyco$TMT_info_FP$microbiome, "_", res_microbiome_mousebrain_Nglyco$TMT_info_FP$replicate)) %>%
    cor(method = "spearman"),
  col = corrplot::COL1("Reds"),
  cluster_columns = F,
  cluster_rows = F,
  show_column_names = FALSE,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  rect_gp = gpar(col = "white"),
  right_annotation = HeatmapAnnotation(
  "microbiome" = res_microbiome_mousebrain_Nglyco$TMT_info_FP$microbiome,
  "sex" = res_microbiome_mousebrain_Nglyco$TMT_info_FP$sex,
  col = list(
    microbiome = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"),
    sex = c("male" = "#E1BE6A", "female" = "#40B0A6")
  ),
  which = "row",
  show_annotation_name = FALSE
), 
name = "corr"
)
```

## B

```{r}
ggplot(
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(dataset == "FP") %>% 
    mutate(
      contrast = str_replace(contrast, "-", "-\n"),
      contrast = factor(contrast, levels = c("monocolonized-\ngermfree", "community-\ngermfree", "community-\nmonocolonized"))
    ),
  aes(x = logFC, y = -log10(P.Value), colour = direction)
) +
  geom_point(alpha = 0.5, stroke = 0) +
  scale_colour_manual(values = c("dodgerblue4", "lightgrey", "indianred")) +
  facet_grid(dataset ~ contrast) +
  # theme_bw(14) +
  theme(panel.grid = element_blank()) +
  labs(x = "log2 fold-change", y = "-log10 p-value")
```


## C

```{r}
hits <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" & hit == "hit")

res_microbiome_mousebrain_Nglyco$limma_results %>% 
  mutate(hit_anno = ifelse(Protein.ID %in% hits$Protein.ID, "at least one hit", "no hit")) %>% 
  filter(dataset == "glyco") %>%
  distinct(hit_anno, glyco_position, Protein.ID, Observed.Modifications) %>% 
  group_by(hit_anno, Protein.ID, glyco_position) %>% 
  count(feature = "forms per site")%>% 
  group_by(hit_anno, n) %>% 
  summarise(a = n()) %>% 
  ungroup() %>% 
  mutate(anno = ifelse(n > 30, "big", "other")) %>% 
  group_by(hit_anno, anno) %>% 
  mutate(a_sum = sum(a))%>% 
  ungroup() %>% 
  mutate(a = ifelse(n > 30, a_sum , a),
         n = ifelse(n > 30,  ">30", as.character(n))) %>% 
  distinct() %>%
  mutate(freq = a/sum(a)*100) %>% 
  mutate(n = factor(n, levels = c(as.character(seq(1:30)), ">30")))%>% 
  ggplot(aes(x = n, y = freq, fill = hit_anno)) +
  geom_bar(position = "dodge", stat = "identity") +
  scale_x_discrete(breaks = c("1", "10", "20", ">30")) +
  # scale_fill_manual(values = c("darkslateblue")) +
  labs(y = "frequency [%]", x = "# gycocompositions per site")

rm(hits)
```


## D

```{r}
ggplot(
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(dataset == "glyco") %>% 
    mutate(
      contrast = str_replace(contrast, "-", "-\n"),
      contrast = factor(contrast, levels = c("monocolonized-\ngermfree", "community-\ngermfree", "community-\nmonocolonized"))
    ),
  aes(x = logFC, y = -log10(P.Value), colour = direction)
) +
  geom_point(alpha = 0.5, stroke = 0) +
  scale_colour_manual(values = c("dodgerblue4", "lightgrey", "indianred")) +
  facet_grid(contrast~ glycan_type) +
  # theme_bw(14) +
  theme(panel.grid = element_blank()) +
  labs(x = "log2 fold-change", y = "-log10 p-value")
```

## E nad F

```{r}
set.seed(1)

hitfrac_df <- bind_rows(
res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" ) %>%  
  group_by(contrast, Protein.ID,glyco_position, glycan_type, hit) %>% 
  count() %>% 
  dcast(contrast + Protein.ID + glyco_position + glycan_type ~ hit, fill = 0) %>% 
  mutate(all_gf = hit + `no hit`) %>% 
  filter(all_gf > 1) %>% 
  mutate(hit_fraction = hit/all_gf, data = "true"),
res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" ) %>% 
  mutate(hit = sample(hit)) %>% 
  group_by(contrast, Protein.ID,glyco_position, glycan_type, hit) %>% 
  count() %>% 
  dcast(contrast + Protein.ID + glyco_position + glycan_type ~ hit, fill = 0) %>% 
  mutate(all_gf = hit + `no hit`) %>% 
  filter(all_gf > 1) %>% 
  mutate(hit_fraction = hit/all_gf, data = "random")
)
 
  hitfrac_df %>% 
  filter(hit_fraction > 0 ) %>% 
  mutate(glycan_type = factor(glycan_type, levels = names(pal))) %>% 
  ggplot(aes(x = hit_fraction, y = -1, colour = data)) +
    ggforce::geom_sina(orientation = "y", stroke = 0,alpha = 0.5, size =2) +
geom_density(aes(x = hit_fraction, colour = data), inherit.aes = FALSE, size = 0.8) +
     geom_hline(yintercept = 0, colour = "grey30") +
    theme(legend.position = "none") +
    scale_colour_manual(values = c("grey", "darkblue")) +
    scale_y_continuous(labels = scales::label_number(accuracy = 1), limits = c(-1.5, 7)) +
  labs(x = "fraction of significantly\n regulated glycoforms per site", y = "data | density")

  hitfrac_df %>% 
  filter(hit_fraction > 0 ) %>% 
  mutate(glycan_type = factor(glycan_type, levels = names(pal))) %>% 
    filter(grepl("fuc|sia|comp|high", glycan_type)) %>% 
  ggplot(aes(x = hit_fraction, y = -1, colour = glycan_type)) +
    ggforce::geom_sina(orientation = "y", stroke = 0,alpha = 0.5, size =1) +
geom_density(aes(x = hit_fraction,  colour = glycan_type), inherit.aes = FALSE, size = 0.8) +
     geom_hline(yintercept = 0, colour = "grey30", linetype = 1) +
    theme(legend.position = "none") +
    scale_color_brewer(palette = "Set1")+
     scale_y_continuous(labels = scales::label_number(accuracy = 1), limits = c(-1.5, 7)) +
  labs(x = "fraction of significantly\n regulated glycoforms per site", y = "data | density")
```

## G and H

```{r}
res_microbiome_mousebrain_Nglyco$limma_results <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  mutate(glyco_position = n_position)
set.seed(1)
dirfraction_df <- bind_rows(
dirfraction_random_df <- res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" ) %>% 
  mutate(direction = ifelse(logFC > 0, "up", "down")) %>% 
  mutate(direction_s = sample(direction, replace = F)) %>% 
  group_by(contrast, Protein.ID, glyco_position, glycan_type, hit, direction_s) %>% 
  count() %>% 
  filter(hit == "hit") %>% 
  ungroup() %>% 
  dcast(contrast + Protein.ID + glyco_position + glycan_type ~ direction_s, value.var = "n", fill = 0) %>% 
  mutate(all_gf = up + down) %>% 
  filter(all_gf > 1) %>% 
  mutate(direction_fraction = up/all_gf) %>% 
  mutate(data = "random"),
res_microbiome_mousebrain_Nglyco$limma_results %>% 
  filter(dataset == "glyco" ) %>% 
  mutate(direction = ifelse(logFC > 0, "up", "down")) %>% 
  group_by(contrast, Protein.ID,glyco_position, glycan_type, hit, direction) %>% 
  count() %>% 
  filter(hit == "hit") %>% 
  ungroup() %>% 
  dcast(contrast + Protein.ID + glyco_position + glycan_type ~ direction, value.var = "n", fill = 0) %>% 
  mutate(all_gf = up + down) %>% 
  filter(all_gf > 1) %>% 
  mutate(direction_fraction = up/all_gf) %>% 
  mutate(data = "true")
)

dirfraction_df %>% 
  mutate(glycan_type = factor(glycan_type, levels = names(pal))) %>%
  #filter(data == "true" & grepl("fuc|sia|com|high", glycan_type)) %>%
  ggplot(aes(x = direction_fraction, y = -0.5, colour = data)) +
  #geom_boxplot(orientation = "y") +
  ggforce::geom_sina(orientation = "y", stroke = 0,alpha = 0.5, size =2) +
  geom_density(aes(x = direction_fraction,  colour = data), inherit.aes = FALSE, size = 0.8) +
  geom_hline(yintercept = 0, colour = "grey60") +
  scale_colour_manual(values = c("grey", "darkblue")) +
  theme(legend.position = "none") +
  scale_y_continuous(labels = scales::label_number(accuracy = 1), limits = c(-1, 2.5)) +
  labs(x = "fraction of upregulated\n glycoforms per site", y = "data | density")

dirfraction_df %>% 
  mutate(glycan_type = factor(glycan_type, levels = names(pal))) %>%
  filter(data == "true" & grepl("fuc|sia|com|high", glycan_type)) %>%
  ggplot(aes(x = direction_fraction, y = -0.5, colour = glycan_type)) +
  #geom_boxplot(orientation = "y") +
  ggforce::geom_sina(orientation = "y", stroke = 0,alpha = 0.5, size =2) +
  geom_density(aes(x = direction_fraction,  colour = glycan_type), inherit.aes = FALSE, size = 0.8) +
  geom_hline(yintercept = 0, colour = "grey60", linetype = 1) +
  scale_y_continuous(labels = scales::label_number(accuracy = 1), limits = c(-1, 2.5)) +
  scale_color_brewer(palette = "Set1") +
  theme(legend.position = "none") +
  labs(x = "fraction of upregulated\n glycoforms per site", y = "glycan type | density")
```

## I

workflow cartoon

## J

```{r}
FCs <- res_microbiome_mousebrain_Nglyco$limma_results  %>% 
  filter(dataset == "glyco"& contrast == "community-germfree") %>% 
  mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>% 
  distinct(glycoID, logFC, hit)

res_microbiome_mousebrain_Nglyco$psm_anno %>%
  filter(Gene == "Cntnap1") %>%
  mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>%
  left_join(FCs, by = "glycoID") %>%
  filter(!is.na(Gene) & hit == "hit") %>%
  ggplot(aes(x = as.factor(glyco_position), y = cleaned_glycosignal, colour = microbiome, group = reorder(Modified.Peptide, -abs(logFC)))) +
  # geom_boxplot() +
  geom_point(position = position_dodge(width = 0.7), alpha = 0.5, size = 2) +
  theme(legend.position = "right") +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
  facet_wrap(~Gene, scales = "free", ncol = 2) +
  labs(x = "glycosite positon", y = "log2 normalized\nglycopeptide intensity")
 
res_microbiome_mousebrain_Nglyco$psm_anno %>%
  filter(Gene == "Cntnap1") %>%
  mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>%
  left_join(FCs, by = "glycoID") %>%
  filter(!is.na(Gene) & hit == "hit") %>%
  distinct(glycoID) %>%
  nrow()
```

## K

```{r}
int <- c("Axon guidance", "Structural constituent of cytoskeleton","Myelin sheath","RHO GTPases activate IQGAPs", "Tight junction", "Recycling pathway of L1",  "Structural molecule activity")


res_fpTPP_mousebrain_microbiome$TPP_enrichment  %>%  
  filter(description %in% int & category != "COMPARTMENTS") %>% 
  mutate(description = factor(description, levels = int)) %>% 
  ggplot(aes(y = description, x = -log10(fdr), size =number_of_genes)) +
  geom_vline(xintercept = -log10(0.01), linetype = 3) +
  geom_point(colour = "#C11555") +
  scale_size(range = c(3, 10)) +
  theme(panel.grid.major.y = element_line(colour = "lightgrey")) +
  labs(y  = "")
```



