---
title: "Plots for Supp Figures"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2, 
  fig.height = 4,
  dpi = 700
)
knitr::opts_knit$set(root.dir = "/Users/burtsche/Documents/01_repos/Glycoproteomics")
```

## Packages

```{r, message=F, warning =F, include=}
library(tidyverse)
library(ggplot2); theme_set(cowplot::theme_cowplot(font_size = 15))
library("reshape2")
library(ggrepel)
library(knitr)
library(VennDiagram)
library(Biostrings)
library(ComplexHeatmap)
library(viridis)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
```

```{r}
options(ggplot2.discrete.colour= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
options(ggplot2.discrete.fill= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))

order <- c("sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small")
pal <- c("lightgrey", "#d7301fff", "#377eb8ff", "#41ae76ff","#88419dff" ,"#ff7f00ff", "#ffe300ff","#f781bfff", "mediumseagreen", "grey")
names(pal) <- c("FP", "sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small", "hit", "no hit")

order_O <- c("O-sialylated", "O-fucosylated", "O-hexose", "O-hybrid", "O-phospho", "O-sulfated", "O-GlcNac")
pal_O <- c("lightgrey",  "#d7301fff", "#377eb8ff", "#41ae76ff","#88419dff" ,"#ff7f00ff", '#a65628ff',"#f781bfff")
names(pal_O) <- c("FP", "O-sialylated", "O-fucosylated", "O-hexose", "O-hybrid", "O-phospho", "O-sulfated", "O-GlcNac")
tilted <-  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust = 1))
blank <- theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
```

```{r}
setwd("/Users/burtsche/Documents/01_repos/Glycoproteomics")
```


```{r}
locations_of_interest <-c("Nucleus", "Cytoplasm", "Mitochondrion", "Endoplasmic reticulum", "Golgi apparatus", "Lysosome", "Plasma membrane", "Extracellular")

compartment_anno <- tt <- read_tsv("comparison_to_datasets/combined_loc_df.tsv")%>%
  mutate(description = ifelse(grepl("Extracellular", description), "Extracellular", description)) %>% 
  filter(description %in% locations_of_interest) %>% 
  group_by(gene) %>%
  filter(score > 2.5) %>%
  mutate(species = ifelse(grepl("MUS", ensembl_prot), "mus musculus", "homo sapiens")) %>% 
  mutate(Gene = gene) %>%
  select(-gene) %>% 
  group_by(Gene) %>% 
  mutate(n_annos =n())

pm_proteins <- compartment_anno %>% 
  filter(description == "Plasma membrane")

extracell_proteins <- compartment_anno %>% 
  filter(description == "Extracellular")

compartment_anno <- compartment_anno %>% 
  mutate(description = ifelse(Gene %in% extracell_proteins$Gene, "Extracellular", description)) %>% 
  mutate(description = ifelse(Gene %in% pm_proteins$Gene, "Plasma membrane", description)) %>% 
  distinct(species, Gene, description)
```


# S1

##A


```{r}
list.files(path = "psm/lysis_methods/immonium/", full.names = TRUE)%>%
  lapply(readr::read_tsv) %>% 
  purrr::map(. %>% select(title, `immonium mz`, `immonium int`)) %>% 
  bind_rows(.id = "sample")%>% 
  mutate(ID = str_extract(title,"HEK.+[lysis, IPA]"),
         ID = str_replace(ID, "\\..+File:.+$", ""))%>% 
  mutate(lysis_method = ifelse(grepl("SDS", ID), "SDS\nlysis", "adapted\nlysis")) %>% 
  filter(sample %in% c("1", "2", "5", "6")) %>% 
  group_by(sample, lysis_method) %>% 
  count(name = "n_ions")%>% 
  group_by(lysis_method) %>% 
  summarise(n = mean(n_ions), sd= sd(n_ions)) %>% 
  ggplot(aes(x = lysis_method, y = n, fill = lysis_method)) +
  geom_bar(stat = "identity") +
  geom_errorbar(aes(ymin=n-sd, ymax=n+sd), colour = "black", width = 0.2) +
  #geom_text(aes(label=n),vjust=-0.25)+
  labs(x = "", y = "# contaminant marker ion")
```

## B

```{r}
load("results/2024-05-02_res_HEK_Hela_optimisation_psm_processed.RData")

res_HEK_Hela_optimisation$psm_total %>% 
  filter(experiment %in% c("SDS_lysis", "singleshot")) %>%  
  mutate(lysis_method = ifelse(grepl("SDS", experiment), "SDS\nlysis", "adapted\nlysis")) %>%
  distinct(Protein.ID, Modified.Peptide, Observed.Modifications, lysis_method, modification, replicate) %>% 
  group_by(lysis_method, modification, replicate) %>%
  count() %>% 
  group_by(lysis_method, modification) %>%
  summarise(mean_IDnumber = mean(n), sd = sd(n)) %>%
  ggplot(aes(x = lysis_method, y = mean_IDnumber, fill = modification)) +
  geom_bar(pos = "dodge", stat = "identity") +
  geom_errorbar(aes(ymin=mean_IDnumber-sd, ymax=mean_IDnumber+sd), width=.2, position = position_dodge(.9)) 

```

## C
scan range rep 3 is weird

```{r}
res_HEK_Hela_optimisation$psm_total %>% 
  filter(experiment == "scan_ranges") %>% 
  mutate(scan_range =  str_extract(sample, "450-2000|700-2000")) %>% 
  ggplot(aes(x = `Calibrated.Observed.M/Z`, fill = modification)) +
  geom_histogram(pos = "dodge") +
  facet_grid(replicate~scan_range)

res_HEK_Hela_optimisation$psm %>% 
  filter(experiment == "scan_ranges") %>% 
  mutate(scan_range =  str_extract(sample, "450-2000|700-2000")) %>% 
  group_by(scan_range, replicate) %>% 
  summarise(n_ = (n_distinct(Protein.ID, Modified.Peptide, Observed.Modifications))) %>% 
  group_by(scan_range) %>% 
  summarise(m = mean(n_))
```

## D


```{r}
res_HEK_Hela_optimisation$psm_total %>% 
  filter(experiment == "scan_ranges") %>% 
  mutate(scan_range =  str_extract(sample, "450-2000|700-2000"))  %>%
  mutate(bins = cut(`Calibrated.Observed.M/Z`, breaks = seq(400, 2000, 100))) %>% 
  group_by(scan_range, bins, modification, replicate) %>%
  count() %>% 
  group_by(scan_range, bins, modification) %>%
  summarise(mean_IDnumber = mean(n), sd = sd(n)) %>%
  ggplot(aes(x = bins, y = mean_IDnumber, fill = modification)) +
  geom_bar(pos = "dodge", stat = "identity") +
  geom_errorbar(aes(ymin=mean_IDnumber-sd, ymax=mean_IDnumber+sd), width=.2, position = position_dodge(.9)) +
  facet_wrap(~scan_range)+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size =8)) +
  labs(x = "binned m/z")
```
## E

```{r}
#specificity

res_HEK_Hela_optimisation$psm_total %>% 
  filter(experiment == "scan_ranges") %>% 
  mutate(scan_range =  str_extract(sample, "450-2000|700-2000"))  %>% 
  group_by(scan_range, modification, replicate) %>%
  count() %>% 
  group_by(scan_range, modification) %>%
  summarise(mean_IDnumber = mean(n), sd = sd(n)) %>%
  mutate(scan_range = ifelse(grepl("450", scan_range), "450-2000", "700-2000")) %>% 
  ggplot(aes(x = scan_range, y = mean_IDnumber, fill = modification)) +
  geom_bar(pos = "dodge", stat = "identity") +
  geom_errorbar(aes(ymin=mean_IDnumber-sd, ymax=mean_IDnumber+sd), width=.2, position = position_dodge(.9)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))



res_HEK_Hela_optimisation$psm_total %>% 
  filter(experiment == "scan_ranges") %>% 
  mutate(scan_range =  str_extract(sample, "450-2000|700-2000"))  %>% 
  group_by(scan_range,replicate, modification) %>% 
  count() %>% 
  dcast(scan_range+replicate~modification) %>% 
  mutate(specificity = glycosylation/(unmodified+glycosylation)*100) %>% 
  group_by(scan_range) %>%
  summarise(mean_specificity = mean(specificity), sd = sd(specificity)) %>%
   mutate(scan_range = ifelse(grepl("450", scan_range), "450-2000", "700-2000")) %>% 
  ggplot(aes(x = scan_range, y = mean_specificity, fill = scan_range)) +
  geom_bar(pos = "dodge", stat = "identity") +
  geom_errorbar(aes(ymin=mean_specificity-sd, ymax=mean_specificity+sd), width=.2, position = position_dodge(.9)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(y = "mean specificity")

res_HEK_Hela_optimisation$psm_total %>% 
  filter(experiment == "scan_ranges") %>% 
  mutate(scan_range =  str_extract(sample, "450-2000|700-2000"))  %>% 
  distinct(Protein.ID, Modified.Peptide, Observed.Modifications, scan_range, modification, replicate) %>% 
  group_by(scan_range, modification, replicate) %>%
  count() %>% 
  group_by(scan_range, modification) %>%
  summarise(mean_IDnumber = mean(n), sd = sd(n)) %>%
  mutate(scan_range = ifelse(grepl("450", scan_range), "450-2000", "700-2000")) %>% 
  ggplot(aes(x = scan_range, y = mean_IDnumber, fill = modification)) +
  geom_bar(pos = "dodge", stat = "identity") +
  geom_errorbar(aes(ymin=mean_IDnumber-sd, ymax=mean_IDnumber+sd), width=.2, position = position_dodge(.9)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(y = "mean glycopeptide IDs")

```

## F

```{r}
load("results/2024-05-06_compdatasets_IDnumber_peptides.RData")

comp_glycopeptides %>%   
group_by(dataset) %>% 
  count() %>% 
  ggplot(aes(x = dataset, y = n, fill = dataset)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = n), vjust = -1) +
  lims(y=c(0, 200000)) +
  scale_fill_manual(values = c("springgreen4", "darkorange", "mediumvioletred", 'purple4' )) +
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust = 1)) +
  labs(x = "", y = "ID number")
```

## G - out of order

```{r}
source("~/Documents/generic_functions.R")
riley_msf_182 <- glyco_psm_processing(path = "comparison_to_datasets/AI-ETD_182glycans_psm.tsv", dataset = "riley_msfragger_182glycans", species = "moouse_brain" )
riley_msf_1799 <- glyco_psm_processing(path = "comparison_to_datasets/AI_ETD_1799glycans_psm.tsv", dataset = "riley_msfragger_1799glycans", species = "moouse_brain" )

us_182 <- glyco_psm_processing(path = "psm/Mouse_brain_182glycans_psm.tsv", dataset = "this_study_182glycans", species = "mouse_brain" )
us_182_hcd <- glyco_psm_processing(path = "psm/Brain_glyco_steppedHCD_182glycans_psm.tsv", dataset = "this_study_182glycans", species = "mouse_brain" )

bind_rows(
  res_species$psm %>%  filter(grepl("mouse", species)),
  riley_msf_1799$psm,
  riley_msf_182$psm,
  us_182$psm,
  us_182_hcd$psm
) %>% 
  mutate(dataset = ifelse(grepl("LF", dataset), "this_study_1799glycans", dataset)) %>% 
  group_by(dataset) %>% 
  summarise(
    glycoproteins = n_distinct(Protein.ID),
    glycopeptides = n_distinct(Protein.ID, Modified.Peptide, Observed.Modifications),
    glycosites = n_distinct(Protein.ID, n_position)
  ) %>% 
  melt() %>% 
  mutate(variable = factor(variable, levels = c("glycoproteins","glycosites", "glycopeptides"))) %>% 
  ggplot(aes(x = variable, y = value, fill = dataset)) +
    geom_bar(stat = "identity", pos = "dodge") +
  facet_wrap(~variable, scales= "free") +
  labs(x = "", y = "count") +
  blank
```

## H

```{r}
locations_of_interest <-c("Nucleus", "Cytoplasm", "Mitochondrion", "Endoplasmic reticulum", "Golgi apparatus", "Lysosome", "Plasma membrane", "Extracellular")

compartment_anno <- tt <- read_tsv("comparison_to_datasets/combined_loc_df.tsv")%>%
  mutate(description = ifelse(grepl("Extracellular", description), "Extracellular", description)) %>% 
  filter(description %in% locations_of_interest) %>% 
  group_by(gene) %>%
  filter(score > 2.5) %>%
  mutate(species = ifelse(grepl("MUS", ensembl_prot), "mus musculus", "homo sapiens")) %>% 
  mutate(Gene = gene) %>%
  select(-gene) %>% 
  group_by(Gene) %>% 
  mutate(n_annos =n())

pm_proteins <- compartment_anno %>% 
  filter(description == "Plasma membrane")

extracell_proteins <- compartment_anno %>% 
  filter(description == "Extracellular")

compartment_anno <- compartment_anno %>% 
  mutate(description = ifelse(Gene %in% extracell_proteins$Gene, "Extracellular", description)) %>% 
  mutate(description = ifelse(Gene %in% pm_proteins$Gene, "Plasma membrane", description)) %>% 
  distinct(species, Gene, description)

load("2024-05-02_combined_lf_data_human_and_mouse.RData")
res_species$psm %>% 
  filter(grepl("mouse", species)) %>% 
  mutate(ID = paste0(Protein.ID, "_", Gene, "_", Modified.Peptide, "_", Observed.Modifications, "_", glycan_type, "_", n_position))%>%
  distinct(Gene, ID) %>% 
  inner_join(compartment_anno %>%  filter(species == "mus musculus")) %>% 
  distinct(Gene, ID, description) %>%  
   mutate(description = factor(description, levels =locations_of_interest)) %>% 
  ggplot(aes(x = description)) +
  geom_bar()+
  tilted +
  labs(y = "# glycoforms")

```

## I

```{r}
# load( "data/paper/revision1/processed/2024-01-08_Mouse_brain_lf_singleshot_psm_processed.RData")
# load("data/paper/revision1/processed/2024-01-08_combined_lf_data_human_and_mouse.RData")
load("results/2024-05-02_Mouse_brain_lf_singleshot_psm_processed.RData")
tt <- bind_rows(
  "single-shot" = res_singleshot_mousebrain$psm %>%  filter(grepl("mouse", species)) %>%  mutate(replicate =as.character(replicate)),
  "PGC fractionation" = res_species$psm %>%  filter(grepl("mouse", species)),
  .id = "dataset"
) 

tt %>% 
  ggplot(aes(x = as.numeric(mass), colour = dataset)) +
  geom_density(adjust = 5, size = 1)+
  scale_color_manual(values = c("slateblue4", "hotpink")) +
  labs(x = "gylcopeptide mass")

tt %>%  
  ggplot(aes(x = Peptide.Length, colour = dataset)) +
  geom_density(adjust = 5,  size = 1)+
   scale_color_manual(values = c("slateblue4", "hotpink")) +
  labs(x = "Peptide.Length")

```

## additional
### 1 & 2

```{r}
load("results/2024-05-09_df_msfragger_byonic.RData")

p <- venn.diagram(disable.logging = T,main = "compositions",
  x = list(
    df_msfragger_byonic %>% filter(dataset == "byonic") %>% distinct(glycan_content) %>% unlist(),
    df_msfragger_byonic %>% filter(dataset == "msfragger") %>% distinct(glycan_content) %>% unlist()
  ),
  filename = NULL,
  category.names = c("byonic", "msfragger"),
  col = c("#21908dff", "purple4"),
  fill = c(alpha("#21908dff", 1), alpha("purple4", 1)),
  cex = 0.8,
  cex.prop = NULL,
  cat.cex = 1,
  cat.default.pos = "text",
  fontfamily = "sans",
  ext.percent = .005,
  cat.dist = c(0.055, 0.055)
)

grid::grid.draw(p)
dev.off()

p <- venn.diagram(disable.logging = T,main = "site + compositions",
  x = list(
    df_msfragger_byonic %>% filter(dataset == "byonic") %>% distinct(comp_site) %>% unlist(),
    df_msfragger_byonic %>% filter(dataset == "msfragger") %>% distinct(comp_site) %>% unlist()
  ),
  filename = NULL,
  category.names = c("byonic", "msfragger"),
  col = c("#21908dff", "purple4"),
  fill = c(alpha("#21908dff", 1), alpha("purple4", 1)),
  cex = 0.8,
  cex.prop = NULL,
  cat.cex = 1,
  cat.default.pos = "text",
  fontfamily = "sans",
  ext.percent = .005,
  cat.dist = c(0.055, 0.055)
)

grid::grid.draw(p)
dev.off()

```

### 3

```{r}
load("results/2024-05-09_df_msfragger_riley.RData")
df_msfragger_riley %>% 
  ggplot(aes(x = variable, y = value, fill = dataset)) +
  geom_bar(stat = "identity", pos = "dodge") +
  blank +
  scale_fill_manual(values = c("darkorange", 'darkslateblue'))+
  facet_wrap(~variable, scales = "free") +
  labs(x = "", y = "ID number")
```




# S2

## A


```{r}
load("results/2024-05-06_overlap_df_brainatlas_glycoproteins.RData")

p <- venn.diagram(
  x = list(
    overlap_df %>% filter(data == "mouse\nbrain atlas") %>% select(Gene) %>% unlist(),
    overlap_df %>% filter(data == "mouse brain fractionated") %>% select(Gene) %>% unlist()
  ),
  filename = NULL,
  category.names = c("", ""),
  col = c("#21908dff", "purple4"),
  fill = c(alpha("#21908dff", 1), alpha("purple4", 1)),
  cex = 0.8,
  cex.prop = NULL,
  cat.cex = 0.2,
  cat.default.pos = "text",
  fontfamily = "sans",
  ext.percent = .005,
  cat.dist = c(0.055, 0.055)
)

dev.off()
grid::grid.draw(p)
```


## B

```{r}
load("results/2024-05-06_uniprot_annotation_comparison.RData")
```


```{r}
uniprot_anno_comparison %>%
  filter(evidence != "combined" & evidence != "similarity") %>%
  distinct(evidence, Protein.ID, n_position) %>%
  group_by(evidence) %>%
  summarise(glycoproteins = n_distinct(Protein.ID), glycosites = n_distinct(Protein.ID, n_position)) %>%
  melt() %>%
  ggplot(
    aes(x = variable, y = value, fill = evidence)
  ) +
  geom_bar(stat = "identity", pos = "stack") + # position = position_stack(reverse = TRUE)) +
  # scale_fill_manual(values = c("peru", "wheat3", "grey70","turquoise4")) +
  scale_fill_manual(values = c("peru", "wheat3", "turquoise4")) +
  labs(x = "", y = "count")
```


## C

```{r}
load("results/2024-05-06_uniprotannotationER_string.RData")
```


```{r}
interesting_pathways <- data.frame(
  description = c(
    "Cell adhesion",  "Organelle", 
    "Plasma membrane",
    "Extracellular region", "Signaling", "Intracellular"
  ))
ER_notanno_mM %>%
  inner_join(interesting_pathways) %>%
  filter(category %in% c("Component", "Function", "Process")) %>%
  #mutate(glycan_type = factor(glycan_type, levels = order)) %>%
  ggplot(aes(x = -log10(fdr), y = description, colour = evidence, size = number_of_genes)) +
  geom_vline(xintercept = -log10(0.05), linetype = 3) +
  geom_point() +
  #scale_colour_viridis_c(option = "rocket", direction = -1, begin = 0.4, end = 0.8) +
  theme(panel.grid.major.y = element_line(colour = "grey")) +
  labs( y = "", x = "-log10 false-discovery rate")
```

## D

```{r}
load("results/2024-01-18_df_Nglyco_anno_subcell.RData")

df_Nglyco_anno_subcell %>% 
  mutate(description = factor(description, levels = locations_of_interest)) %>% 
  mutate(or = ifelse(or == "annoated", "reported in uniprot", "not reported in uniprot")) %>% 
  ggplot(aes(x = description, fill = or)) +
  geom_bar(pos = "fill")+
  #geom_text(aes(label = after_stat(count)), stat = "count", position = "fill", vjust =-1, size =3) +
  tilted +
  ylim(c(0, 1)) +
  labs(y = "fraction", fill = "annotation status") 
```

## E

```{r}
load("results/2024-05-06_overlap_df_brainatlas.RData")
p <- venn.diagram(
  x = list(
    overlap_df %>% filter(data == "mouse\nbrain atlas") %>% select(Gene) %>% unlist(),
    overlap_df %>% filter(data == "this study") %>% select(Gene) %>% unlist()
  ),
  filename = NULL,
  category.names = c("", ""),
  col = c("mediumseagreen", "#440154ff"),
  fill = c(alpha("mediumseagreen", 0.5), alpha("#440154ff", 0.5)),
  cex = 0.8,
  cex.prop = NULL,
  cat.cex = 0,
  cat.default.pos = "text",
  fontfamily = "sans",
  ext.percent = .005,
  # cat.pos = c(-27, 27),
  cat.dist = c(0.055, 0.055)
)

dev.off()
grid::grid.draw(p)
```

## F


```{r}
list <- riley_glycopsm %>% 
  distinct(glycan_type, glycan_content) %>% 
  mutate(dataset = "Riley et al") %>% 
  bind_rows(res_species$psm %>% filter(grepl("mouse", species)) %>%  
  distinct(glycan_type, glycan_content) %>% 
  mutate(dataset = "us") ) %>% 
  group_by(glycan_type, glycan_content) %>% 
  mutate(n_datasets = n())%>% 
  ungroup() %>% 
  mutate(anno = ifelse(n_datasets == 2, "both", dataset)) %>% 
  distinct(glycan_content, dataset)

x <- split(list , list$dataset)
x <- list("Riley et al" = x$`Riley et al`$glycan_content, "us" = x$us$glycan_content)

v <- nVennR::plotVenn(x, setColors = c("darkorange", "#420978"), opacity=0.3, borderWidth=4,  fontScale = 2, labelRegions = F, nCycles = 30000)

```

## G

```{r}
load("results/2024-05-06_ER_string_NvsO.RData")

intracell_org_genes <- ER_NvsO_string %>%
  filter(description == "Intracellular organelle" & category == "Component") %>%  
  mutate(Genes = str_extract_all(inputGenes, "[:alnum:]+,")) %>% 
  unnest() %>% 
  mutate(Genes = str_replace_all(Genes,",", "")) %>% 
  distinct(Genes)

membrane_genes <- ER_NvsO_string %>%
  filter(description == "Plasma membrane" & category == "Component" ) %>%  
  mutate(Genes = str_extract_all(inputGenes, "[:alnum:]+,")) %>% 
  unnest() %>% 
  mutate(Genes = str_replace_all(Genes,",", "")) %>% 
  distinct(Genes)

micro_annotation <- res_mousebrain_fractionated_Oglyco_combined$psm %>%  
  distinct(Gene, Modified.Peptide, Observed.Modifications) %>% 
  group_by(Gene) %>%
  summarise(n_gforms = n_distinct(Modified.Peptide, Observed.Modifications)) %>% 
  mutate(n_glycoforms =ifelse(n_gforms == 1, "only one glycoform", "1-5 glycoforms"),
         n_glycoforms =ifelse(n_gforms > 1, "more than one glycoform", n_glycoforms)) %>% 
  drop_na() %>% 
  distinct(Gene, n_glycoforms)
```


```{r}
bind_rows(
"N_vs_O" = res_species$psm %>%
  filter(grepl("mouse", species)) %>%
  distinct(Protein.ID, Modified.Peptide, n_position, Observed.Modifications, mass) %>%
  mutate(data = "N-glyco") %>%
  bind_rows(res_mousebrain_fractionated_Oglyco_combined$psm %>%
    distinct(Protein.ID, Modified.Peptide, Observed.Modifications, mass) %>%
    mutate(data = "O-glyco")),

"subellular localisation"  = res_mousebrain_fractionated_Oglyco_combined$psm %>%
    distinct(Gene,Protein.ID, Modified.Peptide, Observed.Modifications, mass) %>%
  ungroup() %>% 
  mutate(compartment = ifelse(Gene %in% intracell_org_genes$Genes, "intracellular organelle", "else"),
         compartment = ifelse(Gene %in% membrane_genes$Genes, "plasma membrane", compartment)),

"microheterogeneity" = res_mousebrain_fractionated_Oglyco_combined$psm %>%
    distinct(Gene,Protein.ID, Modified.Peptide, Observed.Modifications, mass) %>%
  ungroup() %>% 
  inner_join(micro_annotation),

.id = "type"
) %>% 
  mutate(feature = paste0(data, "_", compartment, "_", n_glycoforms),
         feature = str_replace_all(feature, "NA_|NA|_NA", "")) %>% 
  mutate(feature = factor(feature, levels = c("N-glyco", "O-glyco", "intracellular organelle", "plasma membrane", "else","more than one glycoform", "only one glycoform"))) %>% 
  filter(type != "microheterogeneity") %>% 
  ggplot(aes(x = mass, fill = feature, colour = feature)) +
  geom_density(alpha = 0.4,adjust = 1.5) +
  scale_fill_manual(values = c("darkslateblue", "darkgoldenrod", "darkgreen", "darkred", "grey60", "darkblue", "cornflowerblue")) +
  scale_colour_manual(values = c("darkslateblue", "darkgoldenrod", "darkgreen", "darkred", "grey60", "darkblue", "cornflowerblue")) +
  facet_wrap(~type, scales = "free") +
  labs(x = "glycan mass")
```

```{r}
rm(intracell_org_genes, membrane_genes, micro_annotation, ER_NvsO_string)
```

## H

```{r}
diag_ions <- read_tsv("diagnostic_ions/dataset01.diagnosticIons_O.tsv")
load("results/2023-03-01res_microbiome_mousebrain_Oglyco.RData")

tt <- diag_ions %>% 
  select(Peptide, matches("144\\.06"), matches("138\\.05"), matches("168")) 


res_microbiome_mousebrain_Oglyco$psm %>% 
  left_join(tt, relationship = "many-to-many") %>% 
  distinct(Protein.ID, Gene, Observed.Modifications,glycan_type, Peptide,`ox_138.0550_intensity`, `ox_144.0656_intensity`, `ox_168.0655_intensity`) %>% 
  left_join(compartment_anno) %>% 
  #filter(Curated_SVM_predictions == "NUCLEUS") %>% 
  select(description, Protein.ID, Peptide,`ox_138.0550_intensity`, `ox_144.0656_intensity`, `ox_168.0655_intensity`) %>% 
  mutate(ratio = `ox_144.0656_intensity`/`ox_138.0550_intensity`)%>% 
  drop_na() %>% 
  filter(is.finite(ratio)) %>% 
  mutate(description = factor(description, levels = locations_of_interest)) %>% 
  ggplot(aes(x = description, y = ratio)) +
  ggforce::geom_sina(alpha = 0.2, colour = "grey80") +
  geom_boxplot(fill =NA) +
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust=1)) +
  labs(y = "138/144 ratio")

```

## I

```{r}
diag_ions <- read_tsv("diagnostic_ions/dataset01.diagnosticIons.tsv")
load("results/2024-04-29_res_microbiome_mousebrain_Nglyco.RData")

tt <- diag_ions %>% 
  select(Peptide, matches("144\\.06"), matches("138\\.05")) 

res_microbiome_mousebrain_Nglyco$psm %>% 
  distinct(Protein.ID, Gene, Observed.Modifications,glycan_type,Peptide, Modified.Peptide) %>%
  filter(Peptide %in% tt$Peptide) %>% 
  inner_join(tt, relationship = "many-to-many")%>% 
  distinct(Protein.ID, Gene, Observed.Modifications,glycan_type, Peptide,`ox_138.0550_intensity`, `ox_144.0656_intensity`) %>% 
  left_join(compartment_anno) %>% 
  #filter(Curated_SVM_predictions == "NUCLEUS") %>% 
  select(description, Protein.ID, Peptide,`ox_138.0550_intensity`, `ox_144.0656_intensity`) %>% 
  mutate(ratio = `ox_144.0656_intensity`/`ox_138.0550_intensity`) %>% 
  drop_na() %>% 
  filter(is.finite(ratio)) %>% 
  ggplot(aes(x = description, y = ratio)) +
  #ggforce::geom_sina(alpha = 0.2) +
  geom_boxplot(fill =NA) +
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust=1)) +
  labs(y = "144/138 ratio")
```

### additional



```{r}
riley_glycopsm <- readxl::read_excel("comparison_to_datasets/riley_etal_glycopsms.xlsx", sheet = 2) %>%
  mutate(glycan_content = Glycans) %>%
  separate(glycan_content, into = c("hexnac", "hexnac_content", "hex", "hex_content", "end"), sep = "\\(|\\)", extra = "merge", remove = FALSE) %>%
  mutate(
    hexnac_content = as.numeric(hexnac_content),
    hex_content = as.numeric(hex_content),
    end = ifelse(is.na(end), "", end)
  ) %>%
  mutate(glycan_type = case_when(
    grepl("Phospho", glycan_content) ~ "phospho",
    grepl("NeuAc", glycan_content) ~ "sialylated",
    grepl("Fuc", glycan_content) ~ "fucosylated",
    hexnac_content == 2 & hex_content > 3 & end == "" ~ "high mannose",
    hexnac_content == 2 & hex_content %in% 1:3 & end == "" ~ "paucimannose",
    hexnac_content <= 2 & hex == "" & end == "" ~ "small",
    TRUE ~ "complex/hybrid"
  )) %>%
  mutate(
    mass = str_extract(Mods, "NGlycan / \\d+\\."),
    mass = str_replace(mass, "\\.", ""),
    mass = as.numeric(str_replace(mass, "NGlycan / ", ""))
  ) %>%
  mutate(Gene = AnnotationDbi::mapIds(org.Mm.eg.db::org.Mm.eg.db, Uniprot, "SYMBOL", "UNIPROT"))

riley_glycopsm %>% 
  distinct(glycan_type, glycan_content) %>% 
  mutate(dataset = "Riley et al") %>% 
  bind_rows(res_species$psm %>% filter(grepl("mouse", species)) %>%  
  distinct(glycan_type, glycan_content) %>% 
  mutate(dataset = "us") ) %>% 
  group_by(glycan_type, glycan_content) %>% 
  mutate(n_datasets = n())%>% 
  ungroup() %>% 
  mutate(anno = ifelse(n_datasets == 2, "both", dataset))%>% 
  ggplot(aes(x = glycan_type, fill = anno)) +
  geom_bar(pos = "fill")+
  scale_fill_manual(values=  c("grey", "darkorange", "#420978")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust =1))+
  labs(y = "fraction of compositions")
```



```{r}
tt <- res_species$psm  %>%
  filter(grepl("mouse", species)) %>% 
  distinct(Gene) %>%
  mutate(dataset = "N-glyco") %>%
  bind_rows(res_mousebrain_fractionated_Oglyco_combined$psm  %>%
    distinct(Gene) %>%
    mutate(dataset = "O-glyco")) %>%
  group_by(Gene) %>%
  mutate(n = n()) %>%
  ungroup() %>%
  mutate(dataset = ifelse(n == 2, "N- and O-glyco", dataset)) %>%
  as.data.frame()

res_mousebrain_fractionated_Oglyco_combined$psm %>% 
  distinct(Gene, Peptide, Observed.Modifications) %>% 
  inner_join(tt) %>% 
   inner_join(compartment_anno %>%  filter(species == "mus musculus")) %>% 
  mutate(description = factor(description, levels = locations_of_interest)) %>% 
  ggplot(aes(x = description, fill = dataset)) +
  geom_bar(pos = "fill") +
  tilted +
  labs(y = "fraction")
```


#S3

## A

```{r}
load("results/2024-05-06_protein_abundance_info.RData")

protein_abundance_info %>% 
  group_by(Protein.ID) %>% 
  summarise(n_sites = n_distinct(n_position), mean_quant = mean(mean_quant)) %>% 
  ggplot(aes(n_sites, mean_quant)) +
  geom_point(colour = "lightgrey") +
  geom_boxplot(aes(group = n_sites)) +
  labs(y = "mean log10 IBAQ", x = "# of sites")
```

## B

```{r}
protein_abundance_info %>% 
  ggplot(aes(bin, mean_quant)) +
  ggforce::geom_sina(colour = "lightgrey") +
  geom_boxplot() +
  labs(x = "binned number of glycoforms per site", y = "mean log10 IBAQ") 
```

```{r}
rm(protein_abundance_info)
```


## C- martin

## D

```{r}
site_mito_anno <- res_species$psm %>%
  filter(grepl("mouse", species)) %>%
  distinct(species, Protein.ID, n_position, Observed.Modifications) %>%
  group_by(species, Protein.ID, n_position) %>%
  summarise(n_forms_sites = n_distinct(Observed.Modifications)) %>%
  ungroup() %>%
  group_by(species) %>% 
  mutate(bin = as.character(cut_number(n_forms_sites, n = 3))) %>%
  mutate(bin = factor(bin, levels = c("[1,2]", "(2,9]",  "(9,667]")))

res_species$psm %>%
  filter(grepl("mouse", species)) %>% 
  ungroup() %>% 
  distinct(species, Protein.ID, n_position, Peptide) %>% 
  mutate(len = str_count(Peptide)) %>% 
  group_by(species, Protein.ID, n_position) %>% 
  summarise(len = min(as.numeric(len))) %>% 
  left_join(site_mito_anno) %>%
  ggplot(aes(x = bin, y = len)) +
  geom_violin() +
  geom_boxplot(width = 0.4) +
  stat_compare_means(comparisons = list(c("[1,2]", "(2,9]"), c("[1,2]", "(9,667]")), method = "wilcox.test") 

res_species$psm %>%
  filter(grepl("mouse", species)) %>% 
  ungroup() %>% 
  distinct(species, Protein.ID, n_position, Peptide) %>% 
  mutate(len = str_count(Peptide)) %>% 
  # group_by(species, Protein.ID, n_position) %>% 
  # summarise(len = min(as.numeric(len))) %>% 
  left_join(site_mito_anno) %>% 
  ggplot(aes(x = len, y = n_forms_sites)) +
  geom_point() +
  geom_smooth(method = "lm") +
  stat_cor() +
  labs(y = "number of glycoforms on site", x = "length of peptides covering site")
```

## F- clement

## G

```{r}
res_mousebrain_fractionated_Oglyco_combined$psm %>%
  distinct(glycan_type, Protein.ID, Modified.Peptide, Observed.Modifications) %>%
  group_by(glycan_type) %>%
  count(feature = "glyco\npeptides") %>%
  bind_rows(
    res_mousebrain_fractionated_Oglyco_combined$psm %>%
      distinct(glycan_type, Protein.ID) %>%
      group_by(glycan_type) %>%
      count(feature = "glyco\nproteins")
  ) %>%
  mutate(glycan_type = factor(glycan_type, levels = order_O)) %>%
  ggplot(aes(x = glycan_type, y = n, fill = glycan_type, group = n)) +
  geom_bar(aes(group = n), stat = "identity", pos = "dodge") +
  scale_fill_manual(values = pal_O)+
  facet_wrap(~feature, scales = "free", drop = T) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(x = "", y = "ID number")
```

## H

```{r}
res_species$psm %>%  
  filter(grepl("mouse", species)) %>% 
  distinct(Protein.ID, Modified.Peptide, Observed.Modifications, glycan_type, n_position) %>% 
  group_by(Protein.ID, n_position) %>%
  summarise(n_gforms = n_distinct(Modified.Peptide, Observed.Modifications)) %>% 
  filter(n_gforms == 1)%>% 
  drop_na() %>% 
  inner_join(res_species$psm %>%  
  filter(grepl("mouse", species)) %>% 
  distinct(Protein.ID, Modified.Peptide, Observed.Modifications, glycan_type, n_position))%>% 
  distinct(Protein.ID, n_position, glycan_type) %>% 
  group_by(glycan_type) %>%  
  count()  %>% 
  ungroup() %>% 
  group_by(glycan_type) %>%  
  mutate(freq = n/n_all*100) %>% 
  mutate(glycan_type = factor(glycan_type, levels = order)) %>% 
  ggplot(aes(x = glycan_type, y = freq, fill = glycan_type)) +
  geom_bar(stat = "identity", pos = "dodge") +
  geom_text(aes(label = round(freq)), vjust = -1) +
  scale_fill_manual(values = pal)  +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust =1), legend.position = "none") +
  lims(y = c(0, 50)) +
  labs(x = "", y = "frequency on\nmonoglycosylated sites")
```

## I

```{r}
load("results/2024-05-06_df_Nglyco_anno_subcell.RData")

res_species$psm %>% 
  filter(grepl("mouse", species)) %>% 
   mutate(ID = paste0(Protein.ID, "_", Gene, "_", Modified.Peptide, "_", Observed.Modifications, "_", glycan_type, "_", n_position))%>%
  distinct(Gene,glycan_type,  ID) %>%  
  inner_join(compartment_anno %>%  filter(species == "mus musculus")) %>% 
  mutate(glycan_type = factor(glycan_type, levels = names(pal))) %>% 
   mutate(description = factor(description, levels =locations_of_interest)) %>% 
  ggplot(aes(x = description, fill = glycan_type)) +
  geom_bar(pos = "fill") +
  tilted +
  scale_fill_manual(values = pal) +
  labs(y = "fraction of glycoforms")
  
```

## J

```{r}
res_species$psm %>%
  #filter(!grepl("mouse", species)) %>% 
  distinct(Protein.ID,Gene, Observed.Modifications, n_position) %>% 
  group_by(Protein.ID, Gene, n_position) %>% 
  count%>% 
  left_join(compartment_anno)%>% 
  drop_na() %>% 
  mutate(description = factor(description, levels = locations_of_interest)) %>% 
  ggplot(aes(y = log2(n), x = description)) +
  ggforce::geom_sina(colour = "grey") +
  stat_summary(geom = "crossbar", aes(group = description), fun = median)+
  theme(legend.position = "none", axis.text.x = element_text(angle = 45, hjust =1, vjust=1)) +
  labs(y = "log2 number\nof glycoforms\n(microheterogeneity)")
```


## K

```{r}
load("results/2024-05-02_Mouse_brain_lf_singleshot_psm_processed.RData")
res_species$psm %>% 
  mutate(model_system = str_extract(Spectrum,"HEK|HeLa")) %>% 
  mutate(model_system =ifelse(is.na(model_system), "mouse\nbrain", model_system)) %>% 
  distinct(model_system, glycan_type, Protein.ID, Modified.Peptide, Observed.Modifications, n_position) %>% 
bind_rows(res_singleshot_mousebrain$psm %>% 
            mutate(model_system = "mouse\nsingleshot")%>% 
  distinct(model_system, glycan_type, Protein.ID, Modified.Peptide, Observed.Modifications, n_position)) %>% 
  group_by(model_system, glycan_type) %>% 
  summarise(glycopeptides = n_distinct(Protein.ID, Modified.Peptide, Observed.Modifications),
            glycoproteins = n_distinct(Protein.ID),
            glycosites = n_distinct(Protein.ID, n_position)) %>%  
  melt(variable.name = "feature", value.name = "n") %>% 
  mutate(glycan_type = factor(glycan_type, levels = order),
         feature = factor(feature, levels = c("glycopeptides", "glycosites", "glycoproteins"))) %>% 
  ggplot(aes(x =model_system, y = n, fill = glycan_type)) +
  geom_bar(stat = "identity", pos = "dodge") +
  scale_fill_manual(values = pal)  +
  facet_wrap(~feature, scales = "free", drop = T) +
  labs(x = "", y = "ID number")
```

## tmp

```{r}
load("results/2024-05-02_Mouse_brain_lf_singleshot_psm_processed.RData")
order <- c("sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small")
res_species$psm %>% 
  mutate(model_system = str_extract(Spectrum,"HEK|HeLa")) %>% 
  mutate(model_system =ifelse(is.na(model_system), "mouse brain\nfracionated", model_system)) %>% 
  distinct(model_system, glycan_type, Protein.ID, Modified.Peptide, Observed.Modifications, n_position) %>% 
bind_rows(res_singleshot_mousebrain$psm %>% 
            mutate(model_system = "mouse\nsingleshot") %>% 
  distinct(model_system, glycan_type, Protein.ID, Modified.Peptide, Observed.Modifications, n_position)) %>% 
  group_by(model_system, glycan_type) %>% 
  summarise(glycopeptides = n_distinct(Protein.ID, Modified.Peptide, Observed.Modifications),
            glycoproteins = n_distinct(Protein.ID),
            glycosites = n_distinct(Protein.ID, n_position)) %>%  
  melt(variable.name = "feature", value.name = "n") %>% 
  mutate(glycan_type = factor(glycan_type, levels = order),
         feature = factor(feature, levels = c("glycopeptides", "glycosites", "glycoproteins"))) %>%  
  filter(grepl("mouse", model_system) & feature == "glycopeptides") %>%  
  group_by(model_system) %>% 
  mutate(n_all = sum(n),
         freq = n/n_all*100) %>% 
  dcast(glycan_type~model_system, value.var= "freq") %>%
  ggplot(aes(x = `mouse brain\nfracionated`, y = `mouse\nsingleshot`, colour = glycan_type)) +
  geom_abline(slope = 1, linetype = 3) +
  geom_point(size = 3) +
  scale_colour_brewer(palette = "Set1")
```

## L

```{r}
load("results/2024-05-06_ER_string_N_glycantype.RData")
```

```{r}
interesting_pathways <- data.frame(#category = c("Process", "COMPARTMENTS",  "Function", "Component"),
           description = c("Cell adhesion","Cell migration", "Organelle", "Response to stimulus", "Plasma membrane", "Transmembrane signaling receptor activity", "Extracellular region", "Signaling", "Axon", "Nervous system development"))
order <- c("sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small")

ER_string_glycantype %>% 
  inner_join(interesting_pathways) %>% 
  filter(category != "RCTM") %>% 
  mutate(glycan_type = factor(glycan_type, levels = order))%>% 
  ggplot(aes(x = glycan_type, y = description, colour = -log10(fdr), size = number_of_genes))+
  geom_point() +
  scale_colour_viridis_c(option = "rocket", direction = -1, begin = 0.4, end = 0.8)  +
  theme(axis.text.x =element_text(angle = 45, hjust =1, vjust =1)) +
  labs(x="", y="")

rm(interesting_pathways, ER_string_glycantype)
```

## M

```{r}
library(org.Mm.eg.db)
lysosome_cc <- AnnotationDbi::select(org.Mm.eg.db, keytype="GOALL", keys="GO:0005764", columns="UNIPROT")
order <- c("sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small")

res_species$psm %>% 
  filter(grepl("mouse", species)) %>% 
  mutate(lysosome_anno = ifelse(Protein.ID %in% lysosome_cc$UNIPROT, "lysosomal", "other")) %>% 
  group_by(glycan_type, lysosome_anno) %>% 
  summarise(# n_lys_proteins = n_distinct(Protein.ID),
            n_lys_glycoforms = n_distinct(Protein.ID, Modified.Peptide, Observed.Modifications)) %>% 
  dcast(glycan_type ~ lysosome_anno) %>% 
  mutate(ratio = lysosomal/other*100)%>% 
  mutate(glycan_type = factor(glycan_type, levels = order)) %>% 
  ggplot(aes(x = glycan_type, y = ratio, fill = glycan_type)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = pal) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), legend.position = "none") +
  labs(y = "fraction of glycoforms \n on lysosomal proteins (%)")

matrix <- res_species$psm %>% 
  filter(grepl("mouse", species)) %>% 
  mutate(lysosome_anno = ifelse(Protein.ID %in% lysosome_cc$UNIPROT, "lysosomal", "other"),
         phospho = ifelse(glycan_type == "phospho", "phospho", "other")) %>% 
  dplyr::select(Protein.ID, Modified.Peptide, Observed.Modifications,phospho, lysosome_anno) %>% 
  group_by(phospho, lysosome_anno) %>% 
  summarise(n_lys_glycoforms = n_distinct(Protein.ID, Modified.Peptide, Observed.Modifications)) %>% 
  acast(phospho ~ lysosome_anno) %>% 
  as.data.frame()

chisq.test(matrix, matrix$lysosomal, correct = F)

rm(lysosome_cc, matrix)
```

#S4

## A

```{r}
upset_matrix <- ComplexHeatmap::make_comb_mat(res_species$psm %>%
  filter(species == "mouse\nbrain") %>%
  distinct(glycan_type, Protein.ID, n_position) %>%
  mutate(ID = paste0(Protein.ID, "_", n_position)) %>%
  select(glycan_type, ID) %>%
  mutate(val = 1) %>%
  pivot_wider(names_from = glycan_type, values_from = val, values_fill = 0))

ComplexHeatmap::UpSet(
  column_title = "site-level",
  upset_matrix,
  lwd = 1,
  pt_size = unit(3, "mm"),
  bg_pt_col = "ghostwhite",
  # bg_col = "white",
  comb_col = c("red", "darkblue", "black", "darkgreen", "goldenrod", "darkorchid4", "brown")[ComplexHeatmap::comb_degree(upset_matrix)]
)
```

## B

```{r}
upset_matrix <- ComplexHeatmap::make_comb_mat(res_species$psm %>%
  filter(species == "mouse\nbrain") %>%
  distinct(glycan_type, Protein.ID) %>%
  mutate(ID = paste0(Protein.ID, "_")) %>%
  select(glycan_type, ID) %>%
  mutate(val = 1) %>%
  pivot_wider(names_from = glycan_type, values_from = val, values_fill = 0))

ComplexHeatmap::UpSet(
  column_title = "protein-level",
  upset_matrix,
  lwd = 1,
  pt_size = unit(3, "mm"),
  bg_pt_col = "ghostwhite",
  comb_col = c("red", "darkblue", "black", "darkgreen", "goldenrod", "darkorchid4", "brown")[ComplexHeatmap::comb_degree(upset_matrix)]
)
```

```{r}
rm(upset_matrix)
```

## C

```{r}
 morethan1site <- res_species$psm %>% 
  filter(grepl("mouse", species)) %>% 
  distinct(Protein.ID, n_position) %>% 
  group_by(Protein.ID) %>% 
  count %>% 
  filter(n > 2)


tt <- res_species$psm %>% 
  filter(grepl("mouse", species)) %>% 
  distinct(Gene, Protein.ID, n_position, Observed.Modifications) %>%
  group_by(Gene, Protein.ID, n_position) %>%
  summarise(n_forms_sites = n_distinct(Observed.Modifications)) %>%
  ungroup() %>%
  mutate(bin = as.character(cut_number(n_forms_sites, n = 3))) %>% 
  mutate(bin = factor(bin, levels = c("[1,2]", "(2,9]", "(9,667]")))

  
    tt %>%  filter(Protein.ID %in% morethan1site$Protein.ID) %>% 
  group_by(Protein.ID, bin) %>% 
  mutate(n_bin = n()) %>% 
  group_by(Protein.ID) %>% 
  mutate(n_protein = n()) %>% 
  distinct(Protein.ID,bin, n_bin, n_protein) %>% 
  mutate(ratio = n_bin/n_protein) %>% 
  ggplot(aes(x =ratio, colour = bin)) +
  geom_density(size =1, adjust = 1.5)+
  labs(x = "fraction of sites per protein \n (more than two sites)")
```

## D

```{r}
intracell_org_genes <- ER_NvsO_string %>%
  filter(description == "Intracellular organelle" & category == "Component" & dataset == "O-glyco") %>%  
  mutate(Genes = str_extract_all(inputGenes, "[:alnum:]+,")) %>% 
  unnest() %>% 
  mutate(Genes = str_replace_all(Genes,",", "")) %>% 
  distinct(Genes)

membrane_genes <- ER_NvsO_string %>%
  filter(description == "Plasma membrane" & category == "Component" & dataset == "O-glyco") %>%  
  mutate(Genes = str_extract_all(inputGenes, "[:alnum:]+,")) %>% 
  unnest() %>% 
  mutate(Genes = str_replace_all(Genes,",", "")) %>% 
  distinct(Genes)

res_mousebrain_fractionated_Oglyco_combined$psm %>% 
  distinct(glycan_type,Gene, Protein.ID, Modified.Peptide, Observed.Modifications) %>% 
  mutate(biol_process = ifelse(Gene %in% intracell_org_genes$Genes, "intracellular organelle", "else"),
         biol_process = ifelse(Gene %in% membrane_genes$Genes, "plasma membrane", biol_process))%>% 
  group_by(biol_process, glycan_type) %>% 
  count(feature = "n_glycopeptides") %>% 
  group_by(biol_process) %>% 
  mutate(n_all = sum(n), freq = n/n_all*100) %>% 
  mutate(glycan_type = factor(glycan_type, levels = order_O)) %>% 
  ggplot(aes(x =glycan_type, y = freq, fill = glycan_type, group = n)) +
  geom_bar(aes(group = n), stat = "identity", pos = "dodge") +
    geom_label(aes(label = round(freq))) +
  scale_fill_manual(values = pal_O) +
  facet_wrap(~biol_process, scales = "free", drop = T) +
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust = 1), legend.position = "none") +
  labs(x = "", y = "frequency (100%)")

#rm(ER_NvsO_string, intracell_org_genes, membrane_genes)
```

## E

```{r}
load("results/2024-05-06_mousebrain_fractionated_Oglyco_combined_psm_processed.RData")

pepwith1 <- res_mousebrain_fractionated_Oglyco_combined$psm %>% 
  distinct(Protein.ID, Peptide, glycan_content) %>% 
  group_by(Protein.ID, Peptide) %>% 
  count %>% 
  filter(n==1) %>% 
  mutate(ID = paste0(Protein.ID, "_", Peptide))

res_mousebrain_fractionated_Oglyco_combined$psm %>% 
  distinct(Protein.ID, Peptide, glycan_content) %>% 
  mutate(ID = paste0(Protein.ID, "_", Peptide))%>% 
  filter(ID %in% pepwith1$ID) %>% 
  group_by(glycan_content) %>% 
  count() %>% 
  arrange(-n) %>% 
  head(5) %>% 
  ggplot(aes(y =  glycan_content, x=n)) +
  geom_bar(stat = "identity") +
  labs(x = "Freq of compositions\n on O-glycopeptides\n with one glycoform")
```

## F

```{r}

res_mousebrain_fractionated_Oglyco_combined$psm %>%
  filter(grepl("mouse", species)) %>%
  distinct(Protein.ID, Gene, Observed.Modifications, glycan_type, Peptide) %>%
  inner_join(compartment_anno) %>%
  drop_na() %>%
  mutate(glycan_type = factor(glycan_type, levels = names(pal_O))) %>%
  mutate(description = factor(description, levels = locations_of_interest)) %>%
  ggplot(aes(x = description, fill = glycan_type)) +
  geom_bar(pos = "fill") +
  scale_fill_manual(values = pal_O) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
  labs(y = "fraction of glycoforms")
```


## G

```{r}
load("results/2024-05-06_microheterogeneityER_string_Oglyco.RData")
```


```{r}
ER_string_O_microhet %>%
  filter(category %in% c("Component", "Function", "Process")) %>% 
  filter(description %in% c("Organelle", "Cell junction", "Plasma membrane", "Regulation of signaling", "Cell adhesion")) %>% 
  ggplot(aes(x = -log10(fdr), y = reorder(description, -number_of_genes), colour = n_glycoforms, size = number_of_genes)) +
  geom_point() +
  geom_vline(xintercept = -log10(0.05), linetype = 3) +
  scale_color_manual(values = c("darkblue", "lightblue")) +
  theme(panel.grid.major.y = element_line(colour = "lightgrey")) +
  labs(y= "")
rm(ER_string_O_microhet)
```

# S5 - Martin







