---
title: "Supplementary tables"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2,
  fig.height = 4,
  dpi = 700
)
knitr::opts_knit$set(root.dir = "/Users/burtsche/Documents/Glycoproteomics/")
```

## Packages

```{r, message=F, warning =F, include=}
library(tidyverse)
library(ggplot2)
theme_set(cowplot::theme_cowplot(font_size = 15))
library("reshape2")
library(openxlsx)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
filter <- dplyr::filter
```

```{r}
options(ggplot2.discrete.colour = c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
options(ggplot2.discrete.fill = c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
```

# Data

## processed files

```{r}
# Data1
load("results/2024-05-02_res_HEK_Hela_optimisation_psm_processed.RData")
load("results/2024-05-02_Mouse_brain_lf_singleshot_psm_processed.RData")

# Data2
load("results/2024-05-02_mousebrain_fractionated_Nglyco_psm_processed.RData")
load("results/2024-04-26_res_mousebrain_steppedHCD_psm_processed.RData")
load("results/2024-05-06_mousebrain_fractionated_Oglyco_psm_processed.RData")
load("results/2024-05-06_mousebrain_fractionated_steppedHCD_Oglyco_psm_processed.RData")
load("results/2024-05-02_HEK_HeLa_lf_PGC_psm_processed.RData")

# Data 3
load("results/2024-09-26_res_2FF_timecourse.RData")

# Data4
load("data/paper/revision1/processed/2024-02-01_res_surface_acc.RData")

# Data 5
load("results/2024-04-24_res_mouse_tissue_atlas.RData")

# Data 6
load("results/2024-01-16_res_mouse_brain_solubility.RData")

# Data 7
load("results/2024-04-29_res_microbiome_mousebrain_Nglyco.RData")

# Data 8
load("results/2024-01-24_res_microbiome_mousebrain_FPTPP.RData")
```

## enrichments

```{r}
load("results/2024-05-06_ER_string_NvsO.RData")
load("results/2024-05-06_microheterogeneityER_string_Oglyco.RData")
load("results/2024-05-06_microheterogeneityER_string.RData")
load("results/2024-05-06_ER_string_N_glycantype.RData")
load("results/2024-05-06_domain_ORA_Nglyco.RData")
```

## other

```{r}
mouse_experiments <- readxl::read_excel("data/paper/revision2/supplementary_data/Mouse_experiments_information.xlsx")
```

# Data 1 -optimisation

```{r}
optimization_data <- list(
  "explanation" = readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 1),
  "glycan_database" = read_tsv("data/updated_db.tsv"),
  "N-glycosylation_mousebrain" =
    res_singleshot_mousebrain$psm_total %>%
      group_by(Protein.ID, Gene, Protein.Description, n_position, Peptide, Modified.Peptide, glycan_type, Observed.Modifications, mass, mass_bin = bin, modification, replicate) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(
        data = "N-glyco_singleshot",
        glycan_type = ifelse(modification == "unmodified", NA, glycan_type),
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]")
      ),
  "N-glyco_HeLa_scanranges" =
    res_HEK_Hela_optimisation$psm %>%
      filter(experiment == "scan_ranges") %>%
      mutate(scan_range = str_extract(sample, "450-2000|700-2000")) %>%
      group_by(scan_range, Protein.ID, Gene, Protein.Description, n_position, Peptide, Modified.Peptide, glycan_type, Observed.Modifications, mass, mass_bin = bin, modification, replicate) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(
        data = "N-glyco_scanranges",
        glycan_type = ifelse(modification == "unmodified", NA, glycan_type),
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]")
      )
)

optimization_data <- lapply(optimization_data, as.data.frame)

write.xlsx(optimization_data, file = "data/paper/revision2/supplementary_data/Supplementary_Data_1.xlsx")
```


# Data 2 - PGC fractionation lf

```{r}
deep_mouse_brain <- list(
  "explanation" = readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 2),
  "N-glycosylation" =
    res_mousebrain_fractionated_Nglyco$psm_total %>%
      group_by(Protein.ID, Gene, Protein.Description, n_position, Peptide, Modified.Peptide, glycan_type, Observed.Modifications, mass, mass_bin = bin, modification) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(
        data = "N-glyco",
        glycan_type = ifelse(modification == "unmodified", NA, glycan_type),
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = ifelse(modification == "unmodified", NA, Modified.Peptide)
      ),
  "O-glycosylation" =
    res_mousebrain_fractionated_Oglyco$psm %>%
      group_by(Protein.ID, Gene, Protein.Description, Peptide, Protein.Start, Protein.End, glycan_type, Observed.Modifications, mass) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(data = "O-glyco", modification = "glycosylation"),
  "N-glycosylation_steppedHCD" =
    res_mousebrain_steppedHCD$psm_total %>%
      group_by(Protein.ID, Gene, Protein.Description, n_position, Peptide, Modified.Peptide, glycan_type, Observed.Modifications, mass, mass_bin = bin, modification) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(
        data = "N-glyco_steppedHCD",
        glycan_type = ifelse(modification == "unmodified", NA, glycan_type),
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = ifelse(modification == "unmodified", NA, Modified.Peptide)
      ),
  "O-glycosylation_steppedHCD" =
    res_mousebrain_fractionated_Oglyco_steppedHCD$psm %>%
      group_by(Protein.ID, Gene, Protein.Description, Peptide, Protein.Start, Protein.End, glycan_type, Observed.Modifications, mass) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(data = "O-glyco_steppedHCD", modification = "glycosylation"),
  "N-glycosylation_HEK" =
    res_HEK_HeLa_lf_PGC$psm_total %>%
      filter(grepl("HEK", Spectrum)) %>%
      group_by(Protein.ID, Gene, Protein.Description, n_position, Peptide, Modified.Peptide, glycan_type, Observed.Modifications, mass, mass_bin = bin, modification) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(
        data = "N-glyco_HEK",
        glycan_type = ifelse(modification == "unmodified", NA, glycan_type),
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = ifelse(modification == "unmodified", NA, Modified.Peptide)
      ),
  "N-glycosylation_HeLa" =
    res_HEK_HeLa_lf_PGC$psm_total %>%
      filter(grepl("HeLa|Hela", Spectrum)) %>%
      group_by(Protein.ID, Gene, Protein.Description, n_position, Peptide, Modified.Peptide, glycan_type, Observed.Modifications, mass, mass_bin = bin, modification) %>%
      summarise(Hyperscore = max(Hyperscore, na.rm = T)) %>%
      mutate(
        data = "N-glyco_HeLa",
        glycan_type = ifelse(modification == "unmodified", NA, glycan_type),
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = ifelse(modification == "unmodified", NA, Modified.Peptide)
      )
)
deep_mouse_brain <- lapply(deep_mouse_brain, as.data.frame)
write.xlsx(deep_mouse_brain, file = "data/paper/revision2/supplementary_data/Supplementary_Data_2.xlsx")
```

# Data 3 - 2FF

```{r}
quantitative_data_2FF <- list(
  "explanation" = readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 3),
  "results_N-glycosylation" =
    res_2FF_timecourse$limma_results %>%
    inner_join(res_2FF_timecourse$psm %>% group_by(Protein.ID, Modified.Peptide, Observed.Modifications) %>% summarise(Hyperscore = max(Hyperscore, na.rm = T))) %>%
      mutate(
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")
      ) %>%
      select(Protein.ID, Gene, Protein.Description, n_position, Modified.Peptide, glycan_type, Observed.Modifications, Hyperscore, contrast, logFC, adj.P.Val, hit, direction)
)
quantitative_data_2FF <- lapply(quantitative_data_2FF, as.data.frame)
write.xlsx(quantitative_data_2FF, file = "data/paper/revision2/supplementary_data/Supplementary_Data_3.xlsx")
```

# Data 4 - Surface

```{r}
quantitative_data_surface <- list(
  "explanation" =  readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 4),
  "DEresults_affectedsites" =
    res_surface_acc$limma_results %>%
      mutate( Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
           Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")) %>%
      select(dataset, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications, contrast, logFC, adj.P.Val, hit, direction),
  "DEresults_complete" =
    res_surface_acc$limma_results_complete %>%
      mutate( Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
           Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")) %>%
      select(dataset, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications, contrast, logFC, adj.P.Val, hit, direction),
    "Nglyco_psm" =
    res_surface_acc$psm  %>%
    distinct(dataset, treatment, replicate, Protein.ID, Gene, n_position, Peptide, Modified.Peptide, glycan_type, Observed.Modifications, quant_norm) %>%
    mutate(data = "N-glyco_surfaceacc", 
           Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
           Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
           Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]")
           )
)
quantitative_data_surface <- lapply(quantitative_data_surface, as.data.frame)
write.xlsx(quantitative_data_surface, file = "data/paper/revision2/supplementary_data/Supplementary_Data_4.xlsx")
```

# Data 5 - tissues

```{r}
quantitative_data_tissue <- list(
  "explanation" = readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 5),
  "N-glyco_liver" =
    res_mouse_tissue_atlas$glyco_psm_corrected %>%
      filter(tissue == "liver") %>%
      inner_join(res_mouse_tissue_atlas$glyco_psm %>% group_by(Protein.ID, Modified.Peptide, Observed.Modifications) %>% summarise(Hyperscore = max(Hyperscore, na.rm = T))) %>%
      mutate(
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")
      ) %>%
      distinct(tissue, mouse, group, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications, Hyperscore, quant_abundance_corrected = cleaned_glycosignal) %>%
      group_by(tissue, mouse, group, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications, Hyperscore) %>%
      summarise(quant_abundance_corrected = median(quant_abundance_corrected)),
  "N-glyco_kidney" =
    res_mouse_tissue_atlas$glyco_psm_corrected %>%
      filter(tissue == "kidney") %>%
      inner_join(res_mouse_tissue_atlas$glyco_psm %>% group_by(Protein.ID, Modified.Peptide, Observed.Modifications) %>% summarise(Hyperscore = max(Hyperscore, na.rm = T))) %>%
      mutate(
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")
      ) %>%
      distinct(tissue, mouse, group, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications,Hyperscore, quant_abundance_corrected = cleaned_glycosignal) %>%
      group_by(tissue, mouse, group, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications,Hyperscore) %>%
      summarise(quant_abundance_corrected = median(quant_abundance_corrected)),
  "N-glyco_brain" =
    res_mouse_tissue_atlas$glyco_psm_corrected %>%
      filter(tissue == "brain") %>%
      inner_join(res_mouse_tissue_atlas$glyco_psm %>% group_by(Protein.ID, Modified.Peptide, Observed.Modifications) %>% summarise(Hyperscore = max(Hyperscore, na.rm = T))) %>%
      mutate(
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")
      ) %>%
      distinct(tissue, mouse, group, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications,Hyperscore, quant_abundance_corrected = cleaned_glycosignal) %>%
      group_by(tissue, mouse, group, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications, Hyperscore) %>%
      summarise(quant_abundance_corrected = median(quant_abundance_corrected))
)
quantitative_data_tissue <- lapply(quantitative_data_tissue, as.data.frame)
write.xlsx(quantitative_data_tissue, file = "data/paper/revision2/supplementary_data/Supplementary_Data_5.xlsx")
```

# Data 6 - SPP

```{r}
quantitative_data_spp <- list(
  "explanation" =  readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 6),
  "solubility_ratios" =
    res_mouse_brain_solubility$solubility_ratios %>%
    inner_join(res_mouse_brain_solubility$glyco_psm %>% group_by(Protein.ID, Modified.Peptide, Observed.Modifications) %>% summarise(Hyperscore = max(Hyperscore, na.rm = T))) %>%
      mutate( Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
           Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")) %>%
      select(Protein.ID, Modified.Peptide, Observed.Modifications, Hyperscore, group = variable, ratio = value),
  "DE_results" =
   res_mouse_brain_solubility$limma_results %>%
      mutate( Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
             Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
           Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")) %>%
    filter(dataset == "ratio_comp_all") %>% 
      select(dataset, Protein.ID, Gene, n_position, Modified.Peptide, glycan_type, Observed.Modifications, contrast, logFC, adj.P.Val, hit, direction)
)
quantitative_data_spp <- lapply(quantitative_data_spp, as.data.frame)
write.xlsx(quantitative_data_spp, file = "data/paper/revision2/supplementary_data/Supplementary_Data_6.xlsx")
```


# Data 7 - microbiome

```{r}
quantitative_data_microbiome <- list(
  "explanation" = readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 7),
  "mouse_experiments" = mouse_experiments,
  "results_N-glycosylation" =
    res_microbiome_mousebrain_Nglyco$limma_results %>%
    inner_join(res_microbiome_mousebrain_Nglyco$psm %>% group_by(Protein.ID, Modified.Peptide, Observed.Modifications) %>% summarise(Hyperscore = max(Hyperscore, na.rm = T))) %>%
      filter(dataset == "glyco") %>%
      mutate(
        direction = ifelse(logFC > 0, "up", "down"),
        data = "N-glyco",
        Modified.Peptide = str_replace_all(Modified.Peptide, "N\\[", "N\\[glycan_"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M\\[ox\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "\\[\\d+\\]", "\\[mod\\]"),
        Modified.Peptide = str_replace_all(Modified.Peptide, "^n\\[mod\\]", "")
      ) %>%
      select(Protein.ID, Gene, Protein.Description, n_position, Modified.Peptide, glycan_type, Observed.Modifications, Hyperscore, contrast, logFC, adj.P.Val, hit, direction),
  "results_fullproteome" =
    res_microbiome_mousebrain_Nglyco$limma_results %>%
      filter(dataset == "FP") %>%
      mutate(
        direction = ifelse(logFC > 0, "up", "down"),
        data = "full proteome"
      ) %>%
      select(Protein.ID, Gene, Protein.Description, contrast, logFC, adj.P.Val, hit, direction)
)
quantitative_data_microbiome <- lapply(quantitative_data_microbiome, as.data.frame)
write.xlsx(quantitative_data_microbiome, file = "data/paper/revision2/supplementary_data/Supplementary_Data_7.xlsx")
```

# Data 8 - tpp

```{r}
quantitative_tpp <- list(
  "explanation" = readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 8),
  "TPP_quant_info" =
    res_fpTPP_mousebrain_microbiome$FP_irs_norm_sclaed %>%
      select(Gene, ID, temperature, sample, quant_irs_norm, quant_irs),
  "GPMelt_results" =
    res_fpTPP_mousebrain_microbiome$res_GPmelt_df %>%
      select(Protein.ID, Gene, pVal, BHadjustpVal)
)
quantitative_tpp <- lapply(quantitative_tpp, as.data.frame)
write.xlsx(quantitative_tpp, file = "data/paper/revision2/supplementary_data/Supplementary_Data_8.xlsx")
```


# Data 9 - enrichments

```{r}
micro_anno <- data.frame(bin = unique(ER_string_microheterogeneity$bin), microheterogeneity = c("low", "middle", "high"))

enrichment_tables <- list(
  "explanation" = readxl::read_excel("data/paper/revision2/supplementary_data/suppdata_firstpage.xlsx", sheet = 9),
  "N_and_O_string_enrichment" =
    ER_NvsO_string %>%
      dplyr::select(dataset, description, category, number_of_genes, number_of_genes_in_background, fdr, inputGenes),
  "N_domain_enrichment" =
    domain_enrichment_N %>%
      filter(is.finite(odds_ratio)) %>%
      dplyr::select(glycan_type, interpro_description, sites_in_domain = n_type_domain, FT, odds_ratio, adj.p.val, hit),
  "N_and_O_microhet_enrichment" =
    ER_string_microheterogeneity <- bind_rows(
      "O-glyco" = ER_string_O_microhet %>% dplyr::rename(microheterogeneity = n_glycoforms),
      "N-glyco" = ER_string_microheterogeneity %>% left_join(micro_anno),
      .id = "dataset"
    ) %>%
      dplyr::select(dataset, microheterogeneity, description, category, number_of_genes, number_of_genes_in_background, fdr, inputGenes),
  "N_glycan_type_enrichment" = ER_string_glycantype %>%
    dplyr::select(glycan_type, description, category, number_of_genes, number_of_genes_in_background, fdr, inputGenes) %>%
    mutate(dataset = "N-glyco")
)
enrichment_tables <- lapply(enrichment_tables, as.data.frame)
write.xlsx(enrichment_tables, file = "data/paper/revision2/supplementary_data/Supplementary_Data_9.xlsx")
```

```{r}
```
