---
title: "Comparison to other datasets"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2,
  fig.height = 4,
  dpi = 700
)
knitr::opts_knit$set(root.dir = "~/Documents/Glycoproteomics/")
```

## Packages

```{r, message=F, warning =F, include=}
library(tidyverse)
library(ggplot2)
theme_set(cowplot::theme_cowplot(font_size = 15))
library("reshape2")
library(ggrepel)
library(knitr)
library(Biostrings)
library(VennDiagram)
library("org.Mm.eg.db")

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
filter <- dplyr::filter

source("~/Documents/generic_functions.R")
```

```{r}
options(ggplot2.discrete.colour = c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
options(ggplot2.discrete.fill = c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
```

# Load data

```{r}
load("data/paper/processed/2023-05-16_mousebrain_fractionated_Nglyco_psm_processed.RData")
#load("data/paper/processed/2023-05-16_mousebrain_fractionated_Oglyco_psm_processed.RData")
load("data/paper/processed/2023-05-16_Mouse_brain_lf_singleshot_psm_processed.RData")
```

```{r}
Fasta_ref <- readFasta("data/Mus_musculus_ref_proteome.fasta")
```

Mouse proteome atlas: https://www.nature.com/articles/s41592-022-01526-y (Supp Data 1)
Riley: https://www.nature.com/articles/s41467-019-09222-w#Sec2 (Supp Data 5)
Liu: https://www.nature.com/articles/s41467-017-00535-2 (Supp Data 3)

```{r}
uniprot_glyco_Mm = read_tsv("/Users/burtsche/Documents/uniprot/221219_uniprot_Mm_glycosites.tsv")

evidence_annot <- data.frame(
  code = c("ECO:0000269", "ECO:0000303", "ECO:0000305", "ECO:0000250", "ECO:0000255", "ECO:0000256", "ECO:0000259", "ECO:0000312", "ECO:0000313", "ECO:0007744", "ECO:0007829"),
  evidence = c("literature", "literature", "literature", "similarity", "sequence analysis", "sequence analysis", "sequence analysis", "combined", "combined", "combined", "combined")
)
uniprot_glyco <- uniprot_glyco_Mm %>% 
  mutate(Protein.ID = Entry) %>% 
  mutate(glycosites = str_extract_all(Glycosylation, "CARBOHYD \\d+")) %>% 
  unnest(cols = c(glycosites)) %>% 
  mutate(n_position= as.numeric(str_extract(glycosites, "\\d+"))) %>% 
  filter(!is.na(Glycosylation)) %>% 
  mutate(residue = str_extract(Glycosylation, "[[:upper:]]-linked"),
         code = str_extract(Glycosylation,"ECO\\:\\d+"))%>% 
  distinct(Protein.ID, n_position, Glycosylation, code) %>% 
  left_join(evidence_annot)

uniprot_anno_comparison <- res_mousebrain_fractionated_Nglyco$psm %>% 
  distinct(Gene,Protein.ID, n_position, glycan_type) %>% 
  ungroup() %>% 
  left_join(uniprot_glyco, by = c("Protein.ID", "n_position")) %>% 
  mutate(annotation = ifelse(is.na(Glycosylation), "not annotated", "annotated"))%>% 
  mutate(evidence = ifelse(is.na(evidence), "not annotated", evidence)) %>% 
  mutate(evidence = factor(evidence, levels = c("literature","sequence analysis", "similarity", "combined",  "not annotated")))

#save(uniprot_anno_comparison, file =paste0("data/paper/comparison_to_datasets/", Sys.Date(), "_uniprot_annotation_comparison.RData"))
```

```{r}
# riley et al
riley_glycopsm <- readxl::read_excel("data/paper/comparison_to_datasets/riley_etal_glycopsms.xlsx", sheet = 2) %>%
  mutate(glycan_content = Glycans) %>%
  separate(glycan_content, into = c("hexnac", "hexnac_content", "hex", "hex_content", "end"), sep = "\\(|\\)", extra = "merge", remove = FALSE) %>%
  mutate(
    hexnac_content = as.numeric(hexnac_content),
    hex_content = as.numeric(hex_content),
    end = ifelse(is.na(end), "", end)
  ) %>%
  mutate(glycan_type = case_when(
    grepl("Phospho", glycan_content) ~ "phospho",
    grepl("NeuAc", glycan_content) ~ "sialylated",
    grepl("Fuc", glycan_content) ~ "fucosylated",
    hexnac_content == 2 & hex_content > 3 & end == "" ~ "high mannose",
    hexnac_content == 2 & hex_content %in% 1:3 & end == "" ~ "paucimannose",
    hexnac_content <= 2 & hex == "" & end == "" ~ "small",
    TRUE ~ "complex/hybrid"
  )) %>%
  mutate(
    mass = str_extract(Mods, "NGlycan / \\d+\\."),
    mass = str_replace(mass, "\\.", ""),
    mass = as.numeric(str_replace(mass, "NGlycan / ", ""))
  ) %>%
  mutate(Gene = mapIds(org.Mm.eg.db, Uniprot, "SYMBOL", "UNIPROT"))

kuster_mousebrain <- readxl::read_excel("data/paper/comparison_to_datasets/kuster_etal_proteins_across_tissues.xlsx", sheet = 1) %>%
  select(
    "PROTEIN_ID", "DESCRIPTION", "GENE_NAME",
    matches(c(
      "Cerebellum", "Hindbrain", "Frontal.lobe",
      "Hippocampus", "Occipital.lobe", "Olfactory.bulb",
      "Temporal.lobe"
    ))
  ) %>%
  select(
    "PROTEIN_ID", "DESCRIPTION", "GENE_NAME",
    matches(c("IDENTIFICATION"))
  ) %>%
  melt(id.vars = c("PROTEIN_ID", "DESCRIPTION", "GENE_NAME"), variable.name = "tissue", value.name = "ID_type") %>%
  mutate(tissue = str_replace(tissue, "IDENTIFICATION TYPE ", "")) %>%
  filter(!is.na(ID_type))

kuster_mousebrain_IBAQ <- readxl::read_excel("data/paper/comparison_to_datasets/kuster_etal_proteins_across_tissues.xlsx", sheet = 1) %>%
  select(
    "PROTEIN_ID", "DESCRIPTION", "GENE_NAME",
    matches(c(
      "Cerebellum", "Hindbrain", "Frontal.lobe",
      "Hippocampus", "Occipital.lobe", "Olfactory.bulb",
      "Temporal.lobe"
    ))
  ) %>%
  select(
    "PROTEIN_ID", "DESCRIPTION", "GENE_NAME",
    matches(c("IBAQ"))
  ) %>%
  melt(id.vars = c("PROTEIN_ID", "DESCRIPTION", "GENE_NAME"), variable.name = "tissue", value.name = "log10IBAQ") %>%
  mutate(tissue = str_replace(tissue, "IBAQ \\(Log10\\) ", "")) %>%
  filter(!is.na(log10IBAQ)) %>%
  setNames(c("Protein.ID", "Protein.Description", "Gene", "tissue", "log10IBAQ"))

hilic_glycodata <- read_csv("data/paper/comparison_to_datasets/liu_etal_hilic_dataset.csv") %>%
  separate(Proteins, into = c("prefix", "Protein.ID", "longID"), remove = F, sep = "\\|") %>%
  separate(longID, into = c("Gene", "organism"), sep = "_") %>%
  distinct(Protein.ID, n_position = ProSite, Gene, mass = GlyMass, unique_identifier = `Unique Identifier`)
```

# Comparison to mouse proteome atlas brain data (Kuster lab)

https://www.nature.com/articles/s41592-022-01526-y (Supp Data 1)

## Overlap of glycoproteins with expressed brain proteins

```{r}
overlap_df <- res_mousebrain_fractionated_Nglyco$psm %>%
  distinct(Protein.ID, Gene, data) %>%
  bind_rows(kuster_mousebrain %>%
    distinct(Protein.ID = PROTEIN_ID, Gene = GENE_NAME) %>%
    mutate(data = "mouse\nbrain atlas")) %>%
  filter(!is.na(Gene))

p <- venn.diagram(
  x = list(
    overlap_df %>% filter(data == "mouse\nbrain atlas") %>% select(Gene) %>% unlist(),
    overlap_df %>% filter(data == "mouse brain fractionated") %>% select(Gene) %>% unlist()
  ),
  filename = NULL,
  category.names = c("", ""),
  col = c("#440154ff", "#21908dff"),
  fill = c(alpha("#440154ff", 0.5), alpha("#21908dff", 0.5)),
  cex = 0.8,
  cex.prop = NULL,
  cat.cex = 0.2,
  cat.default.pos = "text",
  fontfamily = "sans",
  ext.percent = .005,
  # cat.pos = c(-27, 27),
  cat.dist = c(0.055, 0.055)
)


grid::grid.draw(p)
 dev.off()
rm(p)
```

## Overlap of glycoproteins with expressed brain glycoproteins

```{r}
overlap_df <- res_mousebrain_fractionated_Nglyco$psm  %>%
  distinct(Protein.ID, Gene, data) %>%
  bind_rows(kuster_mousebrain %>%
    distinct(Protein.ID = PROTEIN_ID, Gene = GENE_NAME) %>%
    mutate(data = "mouse\nbrain atlas") %>%
    filter(Protein.ID %in% uniprot_glyco$Protein.ID)) %>%
  filter(!is.na(Gene))

p <- venn.diagram(
  x = list(
    overlap_df %>% filter(data == "mouse\nbrain atlas") %>% select(Gene) %>% unlist(),
    overlap_df %>% filter(data == "mouse brain fractionated") %>% select(Gene) %>% unlist()
  ),
  filename = NULL,
  category.names = c("", ""),
  col = c("#21908dff", "purple4"),
  fill = c(alpha("#21908dff", 1), alpha("purple4", 1)),
  cex = 0.8,
  cex.prop = NULL,
  cat.cex = 0.2,
  cat.default.pos = "text",
  fontfamily = "sans",
  ext.percent = .005,
  cat.dist = c(0.055, 0.055)
)


grid::grid.draw(p)
dev.off()
rm(p)
```

## Effects of basal protein abundnace

```{r}
protein_abundance_info <- kuster_mousebrain_IBAQ %>%
  group_by(Protein.ID) %>%
  summarise(mean_quant = mean(log10IBAQ)) %>%
  filter(Protein.ID %in% res_mousebrain_fractionated_Nglyco$psm$Protein.ID) %>%
  left_join(res_mousebrain_fractionated_Nglyco$psm %>%
    group_by(Protein.ID, n_position) %>%
    summarise(n_forms_sites = n_distinct(Observed.Modifications)) %>%
    ungroup() %>%
    mutate(bin = cut_number(n_forms_sites, n = 3)))

protein_abundance_info %>%
  group_by(Protein.ID) %>%
  summarise(n_sites = n_distinct(n_position), mean_quant = mean(mean_quant)) %>%
  ggplot(aes(n_sites, mean_quant)) +
  geom_point() +
  geom_boxplot(aes(group = n_sites)) +
  labs(y = "mean log10 IBAQ", x = "# of sites")

protein_abundance_info %>%
  ggplot(aes(n_forms_sites, mean_quant)) +
  geom_point() +
  geom_smooth()

protein_abundance_info %>%
  ggplot(aes(bin, mean_quant)) +
  ggforce::geom_sina() +
  geom_boxplot() +
  labs(x = "binned number of glycoforms per site", y = "mean log10 IBAQ")

#save(protein_abundance_info, file = paste0("data/paper/comparison_to_datasets/", Sys.Date(), "_protein_abundance_info.RData"))
```


# Comparison to recent glycoproteomic studies 

Riley: https://www.nature.com/articles/s41467-019-09222-w#Sec2 (Supp Data 5)
Liu: https://www.nature.com/articles/s41467-017-00535-2 (Supp Data 3)

## Comparison of glycan type frequencies

```{r}
order <- c("sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small")

comp_glycantypes <- res_mousebrain_fractionated_Nglyco$psm %>%
  distinct(Protein.ID, n_position, mass, glycan_type) %>%
  mutate(
    mass = str_extract(mass, "\\d+\\."),
    mass = as.numeric(str_replace(mass, "\\.", "")),
    dataset = "This study"
  ) %>%
  distinct(Protein.ID, n_position, mass, dataset, glycan_type) %>%
  bind_rows(riley_glycopsm %>%
    distinct(Protein.ID = Uniprot, n_position = GlycoSite, mass, glycan_type) %>%
    mutate(dataset = "Riley et al.")) %>%
  group_by(dataset, glycan_type) %>%
  count() %>%
  group_by(dataset) %>%
  mutate(freq = n / sum(n) * 100) %>%
  dcast(glycan_type ~ dataset, value.var = "freq")

comp_glycantypes %>%
  mutate(glycan_type = factor(glycan_type, levels = order)) %>%
  ggplot(aes(`Riley et al.`, `This study`, colour = glycan_type)) +
  geom_abline(slope = 1, linetype = 3) +
  geom_point(size = 4, shape = 15) +
  scale_colour_brewer(palette = "Set1") +
  xlim(c(0, 70)) +
  ylim(c(0, 70)) +
  labs(x = "type frequency Riley et al.", y = "type frequency this study")

#save(comp_glycantypes, file = paste0("data/paper/comparison_to_datasets/", Sys.Date(), "_comp_glycantypes.RData"))
```


## Comparison of glycan mass distributions

```{r}
massrange_comparison <- res_mousebrain_fractionated_Nglyco$psm %>%
  distinct(Protein.ID, n_position, mass) %>%
  mutate(
    mass = str_extract(mass, "\\d+\\."),
    mass = as.numeric(str_replace(mass, "\\.", "")),
    dataset = "This study"
  ) %>%
  distinct(Protein.ID, n_position, mass, dataset) %>%
  bind_rows(riley_glycopsm %>%
    distinct(Protein.ID = Uniprot, n_position = GlycoSite, mass) %>%
    mutate(dataset = "Riley et al.")) %>%
  bind_rows(hilic_glycodata %>%
    mutate(
      mass = str_extract(mass, "\\d+\\."),
      mass = as.numeric(str_replace(mass, "\\.", ""))
    ) %>%
    distinct(Gene, n_position, mass) %>%
    mutate(dataset = "Liu et al.")) %>%
  mutate(ID = paste0(Protein.ID, "_", n_position, "_", mass))

res_mousebrain_fractionated_Nglyco$psm %>%
  distinct(Protein.ID, n_position, mass) %>%
  mutate(
    mass = str_extract(mass, "\\d+\\."),
    mass = as.numeric(str_replace(mass, "\\.", "")),
    dataset = "This study"
  ) %>%
  distinct(Protein.ID, n_position, mass, dataset) %>%
  bind_rows(riley_glycopsm %>%
    distinct(Protein.ID = Uniprot, n_position = GlycoSite, mass) %>%
    mutate(dataset = "Riley et al.")) %>%
  bind_rows(hilic_glycodata %>%
    mutate(
      mass = str_extract(mass, "\\d+\\."),
      mass = as.numeric(str_replace(mass, "\\.", ""))
    ) %>%
    distinct(Gene, n_position, mass) %>%
    mutate(dataset = "Liu et al.")) %>%
  mutate(ID = paste0(Protein.ID, "_", n_position, "_", mass)) %>%
  ggplot(aes(x = mass, fill = dataset, colour = dataset)) +
  geom_histogram(position = "dodge") +
  scale_colour_manual(values = c("springgreen4", "darkorange", "purple4")) +
  scale_fill_manual(values = c("springgreen4", "darkorange", "purple4"))

#save(massrange_comparison, file = paste0("data/paper/comparison_to_datasets/", Sys.Date(), "_massrange_comparison.RData"))
```

## Comparison of glycopeptide ID numbers

```{r}
comp_glycopeptides <- res_mousebrain_fractionated_Nglyco$psm %>%
  mutate(dataset = "this study, fractionated", ID = paste(Modified.Peptide, Observed.Modifications)) %>%
  distinct(Protein.ID, ID, dataset) %>%
  bind_rows(res_singleshot$psm %>%
    filter(replicate == "replicate_1") %>%
    mutate(dataset = "this study, 1x singleshot", ID = paste(Modified.Peptide, Observed.Modifications)) %>%
    distinct(Protein.ID, ID, dataset)) %>%
  bind_rows(hilic_glycodata %>%
    distinct(Protein.ID, ID = unique_identifier) %>%
    mutate(dataset = "Liu et al.")) %>%
  bind_rows(riley_glycopsm %>%
    distinct(Protein.ID = Uniprot, ID = seqWithMods) %>%
    mutate(dataset = "Riley et al."))

comp_glycopeptides %>%
  group_by(dataset) %>%
  count() %>%
  ggplot(aes(x = dataset, y = n, fill = dataset)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = n), vjust = -1) +
  lims(y = c(0, 170000)) +
  scale_fill_manual(values = c("springgreen4", "darkorange", "mediumvioletred", "purple4")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(x = "", y = "ID number")

#save(comp_glycopeptides, file = paste0("data/paper/comparison_to_datasets/", Sys.Date(), "_compdatasets_IDnumber_peptides.RData"))
```

# Comparison to uniprot information

```{r}
# annotated glycosites
ggplot(uniprot_anno_comparison %>% 
           filter(evidence != "combined" & evidence != "similarity") %>% 
            distinct(evidence, Protein.ID, n_position) %>%
           group_by(evidence) %>%
           summarise(n= n()),
         aes(x= "mouse brain", y= n, fill = evidence)) +
    geom_bar(stat = "identity",pos = "stack") + # position = position_stack(reverse = TRUE)) +
    # scale_fill_manual(values = c("peru", "wheat3", "grey70","turquoise4")) +
    scale_fill_manual(values = c("peru", "wheat3", "turquoise4")) +
    labs(x= "", y = "# of glycosites")
```


```{r}

```

