---
title: "Extract glycosylation motifs"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: inline
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2,
  fig.height = 4,
  dpi = 700
)
knitr::opts_knit$set(root.dir = "~/Documents/Glycoproteomics/")
```

## Packages

```{r, message=F, warning =F, include=}
library(tidyverse)
library(ggplot2)
theme_set(theme_bw())
library("reshape2")
library(ggrepel)
library(knitr)
library(ggpubr)
library(Biostrings)
library(rmotifx)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by

source("~/Documents/generic_functions.R")
```

```{r}
options(ggplot2.discrete.colour = c("darkorange", "darkblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
options(ggplot2.discrete.fill = c("darkorange", "darkblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
```

# Data

```{r}
load("data/paper/processed/2023-05-16_mousebrain_fractionated_Nglyco_psm_processed.RData")
```

# Functions

```{r}
aggrregated_motifsearch <- function(input, Fasta) {
  # function to run motif search with an aggregated central motif

  # extract sequence windows around glycosylation site from Fasta file
  peptides_with_seqwindow <- getSequenceWindow_adapted(
    input %>%
      ungroup() %>%
      distinct(Protein.ID, n_position, .keep_all = T), Fasta
  ) %>%
    mutate(
      length = str_count(sequence_window),
      sequence_window = str_replace(sequence_window, "N.S|N.T", "X")
    )

  print(nrow(peptides_with_seqwindow))

  # prepare fasta to use as background
  # also extract windows here and make sure they have the correct length
  Fasta_pad <- Fasta %>%
    filter(Protein.ID %in% peptides_with_seqwindow$Protein.ID) %>%
    mutate(
      sequence = str_replace_all(sequence, "N.S|N.T", "X"),
      padded_sequence = paste0(str_dup("_", 10), sequence, str_dup("_", 10))
    )

  # define size of the window
  pattern <- paste0("(?=(.{9}", "X", ".{9}))")

  # format input
  fg.seqs <- peptides_with_seqwindow %>%
    ungroup() %>%
    filter(!is.na(sequence_window)) %>%
    distinct(Protein.ID, sequence_window) %>%
    mutate(size = str_count(sequence_window)) %>%
    filter(size == 19)

  bg.seqs <- str_match_all(paste(unlist(Fasta_pad$padded_sequence), collapse = ""), pattern)[[1]][, 2]

  # set occurence threshold
  occ_threshold <- (nrow(fg.seqs) / 100) * 5

  # central aa has to be a real amino acid, so rename
  bg.seqs <- str_replace(bg.seqs, "X", "N")
  fg.seqs$sequence_window <- str_replace(fg.seqs$sequence_window, "X", "N")
  print(head(fg.seqs))
  print(head(bg.seqs))

  # run motifx search
  mot <- motifx(fg.seqs$sequence_window,
    unlist(bg.seqs),
    central.res = "N",
    min.seqs = occ_threshold,
    pval.cutoff = 1e-5
  )

  print(nrow(mot))
  rm(peptides_with_seqwindow, bg.seqs, fg.seqs, occ_threshold, Fasta_pad, pattern)
  return(mot)
}
aggrregated_motifsearch2 <- function(input, Fasta) {
  # function to run motif search with an aggregated central motif and degenerated aa code
  peptides_with_seqwindow <- getSequenceWindow_adapted(
    input %>%
      ungroup() %>%
      distinct(Protein.ID, n_position, .keep_all = T), Fasta
  ) %>%
    mutate(
      length = str_count(sequence_window),
      sequence_window = str_replace_all(sequence_window, "N.S|N.T", "X"),
      sequence_window = str_replace_all(sequence_window, "S|T|N", "S"),
      sequence_window = str_replace_all(sequence_window, "R|K|H", "R"),
      sequence_window = str_replace_all(sequence_window, "D|E", "D"),
      sequence_window = str_replace_all(sequence_window, "L|I|M", "L"),
      sequence_window = str_replace_all(sequence_window, "W|Y|F", "W"),
      sequence_window = str_replace_all(sequence_window, "G|A|V", "G")
    )

  print(nrow(peptides_with_seqwindow))

  # prepare fasta to use as background
  Fasta_pad <- Fasta %>%
    filter(Protein.ID %in% peptides_with_seqwindow$Protein.ID) %>%
    mutate(
      sequence = str_replace_all(sequence, "N.S|N.T", "X"),
      sequence = str_replace_all(sequence, "S|T|N", "S"),
      sequence = str_replace_all(sequence, "R|K|H", "R"),
      sequence = str_replace_all(sequence, "D|E", "D"),
      sequence = str_replace_all(sequence, "L|I|M", "L"),
      sequence = str_replace_all(sequence, "W|Y|F", "W"),
      sequence = str_replace_all(sequence, "G|A|V", "G"),
      padded_sequence = paste0(str_dup("_", 10), sequence, str_dup("_", 10))
    )

  pattern <- paste0("(?=(.{9}", "X", ".{9}))")

  fg.seqs <- peptides_with_seqwindow %>%
    ungroup() %>%
    filter(!is.na(sequence_window)) %>%
    distinct(Protein.ID, sequence_window) %>%
    mutate(size = str_count(sequence_window)) %>%
    filter(size == 19)

  bg.seqs <- str_match_all(paste(unlist(Fasta_pad$padded_sequence), collapse = ""), pattern)[[1]][, 2]

  occ_threshold <- (nrow(fg.seqs) / 100) * 5

  bg.seqs <- str_replace(bg.seqs, "X", "N")
  fg.seqs$sequence_window <- str_replace(fg.seqs$sequence_window, "X", "N")
  print(head(fg.seqs))
  print(head(bg.seqs))

  mot_t <- motifx(fg.seqs$sequence_window,
    unlist(bg.seqs),
    central.res = "N",
    min.seqs = occ_threshold,
    pval.cutoff = 1e-4
  )


  print(nrow(mot_t))
  rm(peptides_with_seqwindow, bg.seqs, fg.seqs, occ_threshold, Fasta_pad, pattern)
  return(mot_t)
}
getSequenceWindow_adapted <- function(input, fasta) {
  # adaptation of getSequenceWindow() which enlarges the sequence windo so that we can collapse NxT|S and still have a -10-10 window

# rename columns and filter for unique peptides
  input <- input  %>% 
  select(Protein.ID, "Modified.Peptide","Peptide.ID") %>% 
    mutate(
      Modified.Peptide =  stringr::str_replace_all(Modified.Peptide, "M\\[\\d+\\]", "M"),
      Modified.Peptide =  stringr::str_replace_all(Modified.Peptide, "n\\[\\d+\\]", "n"),
      Modified.Peptide =  stringr::str_replace_all(Modified.Peptide, "N\\[\\d+\\]", "N[M]")) %>% 
    distinct(Peptide.ID, .keep_all = T) %>% 
  filter(!is.na(Modified.Peptide) & grepl("\\[", Modified.Peptide))
  
  input_annotated <- input %>%
    mutate(
      residue_of_interest = stringr::str_extract_all(Modified.Peptide, ".\\["),
      residue_of_interest = stringr::str_replace(residue_of_interest, "\\[", ""),
      native_sequence = stringr::str_replace_all(Modified.Peptide, "\\[M\\]", ""),
      residue_position_peptide = stringr::str_locate_all(Modified.Peptide, ".\\[")) %>% 
    unnest() %>% 
    mutate(residue_position_peptide = residue_position_peptide[, 1]) %>%
    unnest() %>% 
    distinct(Protein.ID, Modified.Peptide, residue_position_peptide, .keep_all = T)%>%
     # the residue position of peptides with multiple modificatons is wrong and has to be corrected by substracting the nchars [M]
    group_by(Protein.ID, Modified.Peptide) %>%
    mutate(
      rank = order(order(residue_position_peptide)),
      rank = rank - 1,
      rank = rank * 3,
      residue_position_peptide = residue_position_peptide - rank
    )
  
  
  input_fasta_merged <- input_annotated %>%
    left_join(fasta %>% select(Protein.ID, sequence), by = "Protein.ID") %>%
    rowwise() %>%
    mutate(
      protein_sequence_length = str_count(sequence),
      sequence = paste0(str_dup("_", 9), sequence, str_dup("_", 11)),
      peptide_location_start = stringr::str_locate(sequence, native_sequence)[1],
      residue_position_inprotein = as.numeric(peptide_location_start) + as.numeric(residue_position_peptide),
      # adapt area around residue of interest
      window_start = residue_position_inprotein - 10,
      window_end = residue_position_inprotein + 10,
      sequence_window = substr(sequence, window_start, window_end),
      residue_position_inprotein = residue_position_inprotein-10
    ) %>%
    select(-residue_position_peptide, -peptide_location_start, -window_start, -window_end, -native_sequence, -sequence) %>%
    # remove proteins with no sequence in FASTA eg contaminats
    filter(!is.na(sequence_window))%>% 
    distinct(Peptide.ID, residue_of_interest, .keep_all = T)%>%
    ungroup() %>% 
    select(-Modified.Peptide, -rank)
}
```

# Motif search with collapsed central motif

## motif search

This code was originally written to run the analysis for different species, but not shown here.

```{r}
# combine fastafiles to run search for mouse and human data at the same time
# Fasta <- readFasta("data/220523_homospaiens_uniprot_reviewed.fasta")
Fasta_mM <- readFasta("data/Mus_musculus_swissprot.fasta")
Fasta_combined <- bind_rows(
  # "human\ncell line" = Fasta %>% select(Protein.ID, sequence),
  "mouse\nbrain" = Fasta_mM %>% select(Protein.ID, sequence),
  .id = "species"
) %>%
  nest(Fasta = c(Protein.ID, sequence))
```

```{r}
# run motif search using all sites
motif_species <- res_mousebrain_fractionated_Nglyco$psm %>%
  mutate(Peptide.ID = paste0(Protein.ID, "_", Modified.Peptide)) %>%
  distinct(Peptide.ID, Protein.ID, Modified.Peptide, n_position) %>%
  nest(psm = c(Peptide.ID, Protein.ID, Modified.Peptide, n_position)) %>%
  # dummy column to match with fasta
  mutate(species = "mouse\nbrain" )%>% 
  left_join(Fasta_combined, by = "species") %>%
  mutate(mot = purrr::map2(psm, Fasta, .f = aggrregated_motifsearch)) %>%
  select(-Fasta, -psm) %>%
  unnest(cols = c(mot))

# run motif search per glycan type
motif_glycan_type <- res_mousebrain_fractionated_Nglyco$psm  %>%
  mutate(Peptide.ID = paste0(Protein.ID, "_", Modified.Peptide)) %>%
  distinct(Peptide.ID, Protein.ID, Modified.Peptide, n_position, glycan_type) %>%
  nest(psm = c(Peptide.ID, Protein.ID, Modified.Peptide, n_position)) %>%
  # dummy column to match with fasta
  mutate(species = "mouse\nbrain" )%>% 
  left_join(Fasta_combined, by = "species") %>%
  mutate(mot = purrr::map2(psm, Fasta, .f = aggrregated_motifsearch)) %>%
  select(-Fasta, -psm) %>%
  unnest()

# run motif search per microheterogeneity bin
motif_glycan_bin <- res_mousebrain_fractionated_Nglyco$psm  %>%
  mutate(Peptide.ID = paste0(Protein.ID, "_", Modified.Peptide)) %>%
  distinct(Protein.ID, n_position, Observed.Modifications, .keep_all = T) %>%
  select(Protein.ID, n_position, Observed.Modifications, Modified.Peptide) %>%
  group_by(Protein.ID, n_position) %>%
  mutate(n = n()) %>%
  ungroup() %>%
  distinct(Protein.ID, n_position, n, .keep_all = T) %>%
  mutate(bin = cut_number(n, n = 3), Peptide.ID = (paste0(Protein.ID, "_", Modified.Peptide))) %>%
  select(-Observed.Modifications, -n) %>%
  nest(psm = c(Peptide.ID, Protein.ID, Modified.Peptide, n_position)) %>%
  # dummy column to match with fasta
  mutate(species = "mouse\nbrain" )%>% 
  left_join(Fasta_combined, by = "species") %>%
  mutate(mot = purrr::map2(psm, Fasta, .f = aggrregated_motifsearch)) %>%
  select(-Fasta, -psm) %>%
  unnest()
```

## Plot results

```{r}
motif_species %>%
  ggplot(aes(y = reorder(motif, score), x = species, fill = score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("lightsalmon", "indianred", "darkred")) +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))

motif_glycan_type %>%
  filter(grepl("mouse", species)) %>%
  ggplot(aes(y = reorder(motif, score), x = glycan_type, fill = score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("lightsalmon", "indianred", "darkred")) +
  # facet_wrap(~species) +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(y = "motif")

motif_glycan_bin %>%
  filter(grepl("mouse", species)) %>%
  ggplot(aes(y = reorder(motif, score), x = bin, fill = score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("lightsalmon", "indianred", "darkred")) +
  # facet_wrap(~species, scales = "free") +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(y = "motif")
```

## Save

```{r, eval = F}
micro_anno <- data.frame(bin = unique(motif_glycan_bin$bin), microheterogeneity = c("high", "medium", "low"))

motifs_fullcode <- bind_rows(
  "all_glycans" = motif_species,
  "microheterogeneity" = motif_glycan_bin,
  "glycan_types" = motif_glycan_type,
  .id = "search_type"
) %>%
  filter(grepl("mouse", species)) %>%
  left_join(micro_anno) %>%
  mutate(background = "full_code")

#save(motifs_fullcode, file = paste0("data/paper/motifs/", Sys.Date(),"_all_motifs_fullcode.RData"))
```

# Motif search with degenerated amino acid code

## motif search

```{r}
motif_species <- res_mousebrain_fractionated_Nglyco$psm %>%
  mutate(Peptide.ID = paste0(Protein.ID, "_", Modified.Peptide)) %>%
  distinct(Peptide.ID, Protein.ID, n_position, Modified.Peptide)%>%
  nest(psm = c(Peptide.ID, Protein.ID, Modified.Peptide, n_position)) %>%
  mutate(species = "mouse\nbrain" )%>% 
  left_join(Fasta_combined, by = "species") %>%
  mutate(mot = purrr::map2(psm, Fasta, .f = aggrregated_motifsearch2)) %>%
  select(-Fasta, -psm) %>%
  unnest(cols = c(mot))


motif_glycan_type <- res_mousebrain_fractionated_Nglyco$psm %>%
  mutate(Peptide.ID = paste0(Protein.ID, "_", Modified.Peptide)) %>%
  distinct(Peptide.ID, Protein.ID, Modified.Peptide, n_position, glycan_type) %>%
  nest(psm = c(Peptide.ID, Protein.ID, Modified.Peptide, n_position)) %>%
  mutate(species = "mouse\nbrain" )%>% 
  left_join(Fasta_combined, by = "species") %>%
  mutate(mot = purrr::map2(psm, Fasta, .f = aggrregated_motifsearch2)) %>%
  select(-Fasta, -psm) %>%
  unnest()

motif_glycan_bin <- res_mousebrain_fractionated_Nglyco$psm %>%
  mutate(species = "mouse\nbrain" )%>% 
  distinct(species, Protein.ID, n_position, Observed.Modifications, .keep_all = T) %>%
  select(species, Protein.ID, n_position, Observed.Modifications, Modified.Peptide) %>%
  group_by(species, Protein.ID, n_position) %>%
  mutate(n = n()) %>%
  ungroup() %>%
  distinct(species, Protein.ID, n_position, n, .keep_all = T) %>%
  group_by(species) %>%
  mutate(bin = cut_number(n, n = 3), Peptide.ID = (paste0(Protein.ID, "_", Modified.Peptide))) %>%
  select(-Observed.Modifications, -n) %>%
  nest(psm = c(Peptide.ID, Protein.ID, Modified.Peptide, n_position)) %>%
  left_join(Fasta_combined, by = "species") %>%
  mutate(mot = purrr::map2(psm, Fasta, .f = aggrregated_motifsearch2)) %>%
  select(-Fasta, -psm) %>%
  unnest()
```

## Plot results

```{r}
motif_species %>%
  mutate(motif = str_replace_all(motif, "R", "B")) %>%
  ggplot(aes(y = reorder(motif, score), x = species, fill = score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("lightsalmon", "indianred", "darkred")) +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))


motif_glycan_type %>%
  mutate(motif = str_replace_all(motif, "R", "B")) %>%
  ggplot(aes(y = reorder(motif, score), x = glycan_type, fill = score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("lightsalmon", "indianred", "darkred")) +
  # facet_wrap(~species) +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(y = "motif")

motif_glycan_bin %>%
  mutate(motif = str_replace_all(motif, "R", "B")) %>%
  ggplot(aes(y = reorder(motif, score), x = bin, fill = score)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("lightsalmon", "indianred", "darkred")) +
  # facet_wrap(~species, scales = "free") +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(y = "motif")
```

## Save

```{r, eval = F}
bins <- motif_glycan_bin %>%
  filter(grepl("mouse", species))
micro_anno <- data.frame(bin = unique(bins$bin), microheterogeneity = c("high", "medium", "low"))

motifs_degen_code <- bind_rows(
  "all_glycans" = motif_species,
  "microheterogeneity" = motif_glycan_bin,
  "glycan_types" = motif_glycan_type,
  .id = "search_type"
) %>%
  filter(grepl("mouse", species)) %>%
  left_join(micro_anno) %>%
  mutate(background = "degenerated_code", motif = str_replace_all(motif, "R", "B"))

#save(motifs_degen_code, file = paste0("data/paper/motifs/", Sys.Date(),"_all_motifs_degen_code.RData"))
```

# End

```{r}
```
