---
title: "Preprocess glycoSPP of mouse microbiome brain project"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2, 
  fig.height = 4,
  dpi = 700
)
knitr::opts_knit$set(root.dir = "/Users/burtsche/Documents/01_repos/Glycoproteomics/")
```

## Packages

```{r, message=F, warning =F}
library(tidyverse)
library(ggplot2);theme_set(cowplot::theme_cowplot(font_size = 15))
library("reshape2")
library(OmnipathR)
library(viridis)
library(ggrepel)
options(connectionObserver = NULL)
library(org.Mm.eg.db)
library(knitr)
library(ggpubr)
library(RColorBrewer)
library(ReactomePA)
library(clusterProfiler)
library(reactome.db)
library(ggvenn)
library(Biostrings)
library(limma)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
filter <- dplyr::filter
```

```{r}
options(ggplot2.discrete.colour= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))
options(ggplot2.discrete.fill= c("orange3", "darkslateblue", "darkred", "darkgreen", "darkgrey", "darkviolet"))

order <- c("sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small")
pal <- c("lightgrey", "#d7301fff", "#377eb8ff", "#41ae76ff","#88419dff" ,"#ff7f00ff", "#ffe300ff","#f781bfff", "mediumseagreen", "grey")
names(pal) <- c("FP", "sialylated", "fucosylated", "high mannose", "complex/hybrid", "phospho", "paucimannose", "small", "hit", "no hit")

order_O <- c("O-sialylated", "O-fucosylated", "O-hexose", "O-hybrid", "O-phospho", "O-sulfated", "O-GlcNac")
pal_O <- c("lightgrey",  "#d7301fff", "#377eb8ff", "#41ae76ff","#88419dff" ,"#ff7f00ff", '#a65628ff',"#f781bfff")
names(pal_O) <- c("FP", "O-sialylated", "O-fucosylated", "O-hexose", "O-hybrid", "O-phospho", "O-sulfated", "O-GlcNac")
tilted <-  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust = 1))
blank <- theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
```

# Functions

```{r}
source("~/Documents/generic_functions.R")
#load("results/2023-10-17_res_mouse_brain_solubility.RData")
```

# 1. Glyco

## A Preprocess

```{r}
res_mouse_brain_solubility <- glyco_psm_processing(path = "/Users/burtsche/Documents/01_repos/Glycoproteomics/data/glycoSPP/Mouse_brain_SPP_glyco_psm.tsv", 
                                                     dataset = "mousebrain_SPP_glyco", 
                                                     species = "Mus musculus", version = "20", qvalue_cutoff = 0.05, purity_cutoff = 0.5)

names(res_mouse_brain_solubility) <- paste0("glyco_", names(res_mouse_brain_solubility))


res_mouse_brain_solubility$TMT_info <- data.frame(
  sample_glyco = c(
    "sample09", "sample10", "sample11",
    "sample12", "sample13", "sample14",
    "sample15", "sample16"
  ),
  sample_FP = c(
    "sample01", "sample02", "sample03",
    "sample04", "sample05", "sample06",
    "sample07", "sample08"
  ),
  solubility = c(
    "NP40", "NP40", "NP40", "NP40",
    "SDS", "SDS", "SDS", "SDS"
  ),
  replicate = c(
    "rep1","rep2","rep3", "rep4",
    "rep1","rep2","rep3", "rep4"
  )
)
```

## B Normalisation

```{r}
res_mouse_brain_solubility$glyco_psm_normalised <-  tt <-
  res_mouse_brain_solubility$glyco_psm %>%
  select(Protein.ID, Gene, Peptide, Modified.Peptide, Observed.Modifications, n_position, glycan_type, Gene, Purity, matches("sample")) %>%
  melt(
    variable.name = "sample",
    value.name = "quant",
    id.vars = c("Protein.ID", "Gene", "Peptide", "Modified.Peptide","Observed.Modifications", "n_position", "glycan_type")
  ) %>%
  # add TMT info
  inner_join(res_mouse_brain_solubility$TMT_info,
    by = c("sample" = "sample_glyco"),
    multiple = "all"
  ) %>%
  drop_na() %>%
  mutate(
    ID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications),
    dummy = "dummy"
  ) %>%
  filter(!is.na(Protein.ID) & !is.na(Modified.Peptide) & !is.na(quant)) %>%
  mutate(quant = ifelse(quant == 0, NA, quant)) %>%
  group_by(dummy) %>%
  group_modify(~ tibble(normalisation_melted_ptmTPP(.))) %>%
  ungroup() %>%
  mutate(quant_norm_all = quant_norm) %>% 
  select(-quant_norm) %>% 
  group_by(solubility) %>%
  group_modify(~ tibble(normalisation_melted_ptmTPP(.))) %>%
  ungroup() %>% 
  mutate(quant_norm_sol = quant_norm) %>% 
  select(-quant_norm, -dummy)  
```

### check

```{r}
ggplot(res_mouse_brain_solubility$glyco_psm_normalised  %>%
         distinct(replicate,solubility, Peptide, `Modified.Peptide`,Observed.Modifications, .keep_all = T), 
       aes(x = as.factor(sample), y = log2(quant), colour = solubility)) +
  geom_boxplot() +
  cowplot::theme_cowplot() +
  #facet_wrap(~replicate) +
  labs( y = "log2 intensity") +
  tilted

ggplot(res_mouse_brain_solubility$glyco_psm_normalised  %>%
         distinct(replicate,solubility, Peptide, `Modified.Peptide`,Observed.Modifications, .keep_all = T), 
       aes(x = as.factor(sample), y = log2(quant_norm_sol), colour = solubility)) +
  geom_boxplot() +
  cowplot::theme_cowplot() +
  #facet_wrap(~replicate) +
  labs( y = "log2 quant_norm")
```

## C PCA

### raw

```{r}
m <- acast(res_mouse_brain_solubility$glyco_psm_normalised %>% drop_na(),
  sample ~ Protein.ID + Modified.Peptide + Observed.Modifications,
  value.var = "quant",
  fun.aggregate = median,
  fill = 0
)

m[m == 0] <- NA
m <- t(m)
m <- m[complete.cases(m), ]
m <- t(m)

m <- log2(m)

m <- m[matrixStats::rowVars(m, na.rm = T) %>%
  order(decreasing = T) %>%
  head(6000), ]

pca <- prcomp(m)
plot(pca)
var = (pca$sdev)^2/sum(pca$sdev^2)*100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  inner_join(res_mouse_brain_solubility$TMT_info, by = c("sample" = "sample_glyco")) %>%
  ggplot(aes(x = PC1, y = PC2, colour = as.factor(solubility), shape = replicate)) +
  geom_point(size = 3) +
  labs(title = "Glycoproteome", 
       subtitle = "raw intensities", 
       x = paste0("PC1 (", as.character(round(var[1])), "%)"),
       y = paste0("PC2 (", as.character(round(var[2])), "%)"))

pca_rotation <- pca$rotation %>% as.data.frame()
```

### normalised

```{r}
m <- acast(res_mouse_brain_solubility$glyco_psm_normalised %>% drop_na(),
  sample ~ Protein.ID + Modified.Peptide + Observed.Modifications,
  value.var = "quant_norm_sol",
  fun.aggregate = median,
  fill = 0
)

m[m == 0] <- NA
m <- t(m)
m <- m[complete.cases(m), ]
m <- t(m)

m <- log2(m)

m <- m[matrixStats::rowVars(m, na.rm = T) %>%
  order(decreasing = T) %>%
  head(6000), ]

pca <- prcomp(m)
plot(pca)
var = (pca$sdev)^2/sum(pca$sdev^2)*100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  inner_join(res_mouse_brain_solubility$TMT_info, by = c("sample" = "sample_glyco")) %>%
  ggplot(aes(x = PC1, y = PC2, colour = as.factor(solubility), shape = replicate)) +
  geom_point(size = 3) +
  labs(title = "Glycoproteome", 
       subtitle = "norm intensities - per sol", 
       x = paste0("PC1 (", as.character(round(var[1])), "%)"),
       y = paste0("PC2 (", as.character(round(var[2])), "%)"))

pca_rotation <- pca$rotation %>% as.data.frame()
```

```{r}
rm(m, pca_rotation)
```

#2. FP

## A Load and format

```{r}
res_mouse_brain_solubility$FP <- read_tsv("/Users/burtsche/Documents/01_repos/Glycoproteomics/data/glycoSPP/Mouse_brain_SPP_FP_protein.tsv")
colnames(res_mouse_brain_solubility$FP) <- gsub(" ", "\\.", colnames(res_mouse_brain_solubility$FP))

res_mouse_brain_solubility$FP <- res_mouse_brain_solubility$FP%>% 
  filter(grepl("Mus", Organism) & !(grepl("contam", Protein)))
```

## B Normalise

```{r}
res_mouse_brain_solubility$FP_normalised <- tt <- res_mouse_brain_solubility$FP %>% 
  select(Protein.ID, Gene, matches(res_mouse_brain_solubility$TMT_info$sample_FP)) %>% 
  melt(variable.name = "sample", value.name = "quant") %>% 
  inner_join(res_mouse_brain_solubility$TMT_info, by = c("sample" = "sample_FP")) %>% 
  filter(!is.na(solubility) & quant > 0 & is.finite(quant))%>% 
  drop_na() %>% 
  mutate(ID = Protein.ID)%>% ungroup() %>% 
  #group_by(solubility) %>%
  group_modify(~ tibble(normalisation_melted_ptmTPP(.))) %>%
  ungroup() %>% 
  mutate(quant_norm_sol = quant_norm) %>% 
  select(-quant_norm)  %>% 
  drop_na()
```

### check

```{r}
ggplot(res_mouse_brain_solubility$FP_normalised, 
       aes(y = log2(quant), x = replicate, colour = solubility)) +
  geom_boxplot() 

ggplot(res_mouse_brain_solubility$FP_normalised, 
       aes(y = log2(quant_norm_sol), x = replicate, colour = solubility)) +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust =1)) +
  labs(x = "", title = "Full proteome - norm per sol")

```

## C PCA

### raw

```{r}
m <- acast(res_mouse_brain_solubility$FP_normalised %>% drop_na(),
  sample ~ Protein.ID ,
  value.var = "quant",
  fun.aggregate = median,
  fill = 0
)

m[m == 0] <- NA
m <- t(m)
m <- m[complete.cases(m), ]
m <- t(m)

m <- log2(m)

m <- m[matrixStats::rowVars(m, na.rm = T) %>%
  order(decreasing = T) %>%
  head(6000), ]

pca <- prcomp(m)
plot(pca)
var = (pca$sdev)^2/sum(pca$sdev^2)*100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(res_mouse_brain_solubility$TMT_info, by = c("sample" = "sample_FP")) %>%
  ggplot(aes(x = PC1, y = PC2, colour = as.factor(solubility), shape = replicate)) +
  geom_point(size = 3) +
  labs(title = "Proteome", 
       subtitle = "raw intensities", 
       x = paste0("PC1 (", as.character(round(var[1])), "%)"),
       y = paste0("PC2 (", as.character(round(var[2])), "%)"))

pca_rotation <- pca$rotation %>% as.data.frame()
```

### normalised

```{r}
m <- acast(res_mouse_brain_solubility$FP_normalised %>% drop_na(),
  sample ~ Protein.ID ,
  value.var = "quant_norm_sol",
  fun.aggregate = median,
  fill = 0
)

m[m == 0] <- NA
m <- t(m)
m <- m[complete.cases(m), ]
m <- t(m)

m <- log2(m)

m <- m[matrixStats::rowVars(m, na.rm = T) %>%
  order(decreasing = T) %>%
  head(6000), ]

pca <- prcomp(m)
plot(pca)
var = (pca$sdev)^2/sum(pca$sdev^2)*100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(res_mouse_brain_solubility$TMT_info, by = c("sample" = "sample_FP")) %>%
  ggplot(aes(x = PC1, y = PC2, colour = as.factor(solubility), shape = replicate)) +
  geom_point(size = 3) +
  labs(title = "Proteome", 
       subtitle = "norm intensities - per sol", 
       x = paste0("PC1 (", as.character(round(var[1])), "%)"),
       y = paste0("PC2 (", as.character(round(var[2])), "%)"))

pca_rotation <- pca$rotation %>% as.data.frame()
```


```{r}
rm(m, pca_rotation)
```


#3. Ratios

```{r}
res_mouse_brain_solubility$solubility_ratios <- tt <- res_mouse_brain_solubility$glyco_psm_normalised %>%
  mutate(phosphoID = ID) %>% 
  ungroup() %>%
  filter(quant > 0) %>%
  dplyr::select(Protein.ID, Modified.Peptide, Observed.Modifications, intensity = quant, solubility, replicate) %>%
  drop_na() %>%
  dcast(Protein.ID + Modified.Peptide + Observed.Modifications + replicate ~ solubility, 
        value.var = "intensity", 
        fun.aggregate = sum) %>%
  drop_na() %>%
  filter(NP40 > 0 & SDS > 0) %>%
  mutate(ratio_glyco = (NP40 / SDS)) %>%
  select(-NP40, -SDS) %>%
  dcast(Protein.ID + Modified.Peptide + Observed.Modifications ~ replicate)%>%
  inner_join(
    res_mouse_brain_solubility$FP_normalised %>%
      ungroup() %>%
      filter(quant> 0) %>%
      dplyr::select(Protein.ID, intensity = quant, solubility, replicate) %>%
      dcast(Protein.ID + replicate ~ solubility, value.var = "intensity", fun.aggregate = sum) %>%
      filter(NP40 > 0 & SDS > 0) %>%
      mutate(ratio_FP = (NP40 / SDS)) %>%
      select(-NP40, -SDS) %>%
      dcast(Protein.ID ~ replicate),
    by = c("Protein.ID"), suffix = c("_glyco", "_FP")
  ) %>%
  melt()  %>%  inner_join(res_mouse_brain_solubility$glyco_psm %>%  distinct(Gene, Protein.ID,Observed.Modifications, Modified.Peptide, glycan_type, n_position))
```

```{r}
res_mouse_brain_solubility$solubility_ratios %>% 
  separate(variable, into = c("replicate", "dataset")) %>% 
  #filter(value < 10) %>% 
  dcast(Protein.ID + Modified.Peptide + Observed.Modifications + replicate ~ dataset)%>% 
  ggplot(aes(x = log2(FP), y= log2(glyco), colour = log2(glyco))) +
  geom_hline(yintercept = 0, colour = "grey") +
  geom_vline(xintercept = 0, colour = "grey") +
  geom_abline(slope =1, colour = "grey") +
  geom_point(size = 0.1) + 
  geom_smooth(method='lm', colour = "grey", linetype = 3)+
  #stat_cor(method="spearman")+
  scale_color_gradient2(low = "red",midpoint = 0,mid = "white",high = "blue",space="Lab") +
  facet_wrap(~replicate) +
  #lims(x = c(-6, 1), y = c(-6, 6)) +
  theme_bw() +
  labs(x = "log2 NP40/SDS FP", y = "log2 NP40/SDS glyco", title = "solubility ratios", subtitle = "raw intensities")
  

```

## PCA

```{r}
m <- acast(res_mouse_brain_solubility$solubility_ratios %>% drop_na(),
  variable ~ Protein.ID + Modified.Peptide + Observed.Modifications,
  value.var = "value",
  fun.aggregate = median,
  fill = 0
)

m[m == 0] <- NA
m <- t(m)
m <- m[complete.cases(m), ]
m <- t(m)

m <- log2(m)

# m <- m[matrixStats::rowVars(m, na.rm = T) %>%
#   order(decreasing = T) %>%
#   head(6000), ]

pca <- prcomp(m)
plot(pca)
var = (pca$sdev)^2/sum(pca$sdev^2)*100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  separate(sample, into = c("replicate", "dataset")) %>% 
  ggplot(aes(x = PC1, y = PC2, colour = dataset, shape = replicate)) +
  geom_point(size = 3) +
  labs(title = "Solubility ratios",  
       x = paste0("PC1 (", as.character(round(var[1])), "%)"),
       y = paste0("PC2 (", as.character(round(var[2])), "%)"))

pca_rotation <- pca$rotation %>% as.data.frame()
```

## Proteins with insoluble sub

```{r}
res_mouse_brain_solubility$proteins_with_insol_sub <-tt <-  res_mouse_brain_solubility$solubility_ratios %>% 
  filter(grepl("FP", variable)) %>% 
  distinct(Protein.ID, variable, value) %>% 
  group_by(Protein.ID) %>% 
  summarise(mean_ratio = mean(value)) %>% 
  filter(mean_ratio < 0.5)
```


#4. Limma

### Input

```{r}
res_mouse_brain_solubility$eset_all <-
  res_mouse_brain_solubility$solubility_ratios %>%
  # filter(Protein.ID %in% tt$Protein.ID) %>%
  acast(Protein.ID + Modified.Peptide + Observed.Modifications ~ variable) 

res_mouse_brain_solubility$eset_sites <- tt <-
  res_mouse_brain_solubility$solubility_ratios %>%
  inner_join(
    res_mouse_brain_solubility$glyco_psm %>%
      distinct(Protein.ID, Modified.Peptide, Observed.Modifications, n_position, glycan_type)
  ) %>%
  acast(Protein.ID + n_position ~ variable, fun.aggregate = mean)

res_mouse_brain_solubility$eset_types <- tt <-
  res_mouse_brain_solubility$solubility_ratios %>%
  inner_join(
    res_mouse_brain_solubility$glyco_psm %>%
      distinct(Protein.ID, Modified.Peptide, Observed.Modifications, n_position, glycan_type)
  ) %>%
  acast(Protein.ID + n_position + glycan_type ~ variable, fun.aggregate = mean)

res_mouse_brain_solubility$eset_all <- res_mouse_brain_solubility$eset_all[complete.cases(res_mouse_brain_solubility$eset_all),]%>% limma::normalizeMedianValues() %>%  log2()
res_mouse_brain_solubility$eset_sites <- res_mouse_brain_solubility$eset_sites[complete.cases(res_mouse_brain_solubility$eset_sites),]%>% limma::normalizeMedianValues() %>%  log2()
res_mouse_brain_solubility$eset_types <- res_mouse_brain_solubility$eset_types[complete.cases(res_mouse_brain_solubility$eset_types),]%>% limma::normalizeMedianValues() %>%  log2()
```


```{r}
boxplot(res_mouse_brain_solubility$eset_all)
boxplot(res_mouse_brain_solubility$eset_sites)
boxplot(res_mouse_brain_solubility$eset_types)

```

### Design


```{r}
s <- str_extract(colnames(res_mouse_brain_solubility$eset_all), "glyco|FP")

res_mouse_brain_solubility$design <- model.matrix(~ 0 + s)

colnames(res_mouse_brain_solubility$design) <- c("FP", "glyco")
res_mouse_brain_solubility$design

res_mouse_brain_solubility$contr <-
  cbind(
    "glyco-FP" = c(-1, 1)
  )

rm(s)
```

### Fit

```{r}
fit <- lmFit(res_mouse_brain_solubility$eset_all, res_mouse_brain_solubility$design)
fit <- contrasts.fit(fit, res_mouse_brain_solubility$contr)
fit_all <- eBayes(fit)

fit <- lmFit(res_mouse_brain_solubility$eset_sites, res_mouse_brain_solubility$design)
fit <- contrasts.fit(fit, res_mouse_brain_solubility$contr)
fit_sites <- eBayes(fit)

fit <- lmFit(res_mouse_brain_solubility$eset_types, res_mouse_brain_solubility$design)
fit <- contrasts.fit(fit, res_mouse_brain_solubility$contr)
fit_types <- eBayes(fit)
```

### Extract results

```{r}
mapping <- res_mouse_brain_solubility$glyco_psm %>%
  mutate(ID = paste(Protein.ID, Modified.Peptide, Observed.Modifications, sep = "_")) %>%
  distinct(Protein.ID, Gene, ID, n_position, glycan_type, bin) %>%
  bind_rows(res_mouse_brain_solubility$FP_normalised %>%
    distinct(Protein.ID, Gene) %>%
    mutate(ID = Protein.ID)) %>%
  distinct(Protein.ID, Gene, ID, n_position, glycan_type, bin) %>% 
  mutate(n_position = as.character(n_position))

res_mouse_brain_solubility$limma_results <- tt <-
  lapply(colnames(fit_all$coefficients), function(x) {
    limma::topTable(fit_all, coef = x, number = Inf) %>%
      mutate(contrast = x, dataset = "ratio_comp_all") %>%
      rownames_to_column(var = "ID")
  }) %>%
  bind_rows() %>%
  separate(ID, into = c("Protein.ID", "Modified.Peptide", "Observed.Modifications"), sep = "_", remove = F) %>% 
  left_join(mapping, by = c("Protein.ID", "ID")) %>%
  bind_rows(
    lapply(colnames(fit_types$coefficients), function(x) {
      limma::topTable(fit_types, coef = x, number = Inf) %>%
        mutate(contrast = x, dataset = "ratio_comp_types") %>%
        rownames_to_column(var = "ID")
    }) %>%
      bind_rows() %>%
      separate(ID, into = c("Protein.ID", "n_position", "glycan_type"), sep = "_", remove = F) %>% 
      left_join(mapping %>%  distinct(Protein.ID, n_position, glycan_type, Gene), 
                by = c("Protein.ID", "n_position", "glycan_type")),
    lapply(colnames(fit_sites$coefficients), function(x) {
      limma::topTable(fit_sites, coef = x, number = Inf) %>%
        mutate(contrast = x, dataset = "ratio_comp_sites") %>%
        rownames_to_column(var = "ID")
    }) %>%
      bind_rows() %>%
      separate(ID, into = c("Protein.ID", "n_position"), sep = "_", remove = F)%>% 
      left_join(mapping %>%  distinct(Protein.ID, n_position, Gene), 
                by = c("Protein.ID", "n_position"))
  ) %>%
  mutate(
    hit = ifelse((logFC > log2(1.5) & adj.P.Val < 0.05), "hit", "no hit"),
    hit = ifelse((logFC < log2(1 / 1.5) & adj.P.Val < 0.05), "hit", hit)
  ) %>%
  mutate(
    direction = ifelse(logFC > 0 & hit == "hit", "up", "no change"),
    direction = ifelse(logFC < 0 & hit == "hit", "down", direction)
  )
```

## Hits

```{r}
ggplot(res_mouse_brain_solubility$limma_results %>% filter(hit == "hit"), 
       aes(glycan_type, fill = direction)) +
  geom_bar(pos = "dodge") +
  scale_fill_manual(values = c("dodgerblue4", "indianred")) +
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust = 1)) +
  facet_wrap(~dataset, scales = "free") +
  labs(y = " # hits")

ggplot(res_mouse_brain_solubility$limma_results %>% filter(hit == "hit" & dataset =="ratio_comp_all"), 
       aes(bin, fill = direction)) +
  geom_bar(pos = "dodge") +
  scale_fill_manual(values = c("dodgerblue4", "indianred")) +
  theme(axis.text.x = element_text(angle = 45, hjust =1, vjust = 1)) +
  facet_wrap(glycan_type~dataset, scales = "free") +
  labs(y = " # hits")
```

## Volcano

```{r}
ggplot(
  res_mouse_brain_solubility$limma_results,
  aes(x = logFC, y = -log10(P.Value), colour = direction)
) +
  geom_point(alpha = 0.5, stroke = 0) +
  #geom_text_repel(aes(label = Gene), size =2) +
  scale_colour_manual(values = c("dodgerblue4", "lightgrey", "indianred")) +
  facet_grid( ~ dataset, scales = "free") +
  # theme_bw(14) +
  theme(panel.grid = element_blank()) +
  labs(x = "log2 fold-change\nglyco_solubility-protein_solubility", y = "-log10 p-value")

ggplot(
  res_mouse_brain_solubility$limma_results %>% 
    filter(Protein.ID %in% res_mouse_brain_solubility$proteins_with_insol_sub$Protein.ID),
  aes(x = logFC, y = -log10(P.Value), colour = direction)
) +
  geom_point(alpha = 0.5, stroke = 0) +
  #geom_text_repel(aes(label = Gene), size =2) +
  scale_colour_manual(values = c("dodgerblue4", "lightgrey", "indianred")) +
  facet_grid( ~ dataset, scales = "free") +
  # theme_bw(14) +
  theme(panel.grid = element_blank()) +
  labs(title = "proteins with at least 25% insoluble",
    x = "log2 fold-change\nglyco_solubility-protein_solubility", y = "-log10 p-value")
```


```{r}
ggplot(
  res_mouse_brain_solubility$limma_results %>% 
    mutate(n_fuc = str_extract(Observed.Modifications, "Fuc\\(\\d+\\)"),
           n_sia = str_extract(Observed.Modifications, "NeuGc\\(\\d+\\)|NeuAc\\(\\d+\\)")) %>% 
    filter(!is.na(n_fuc)),
  aes(x = logFC, y = -log10(P.Value), colour = direction)
) +
  geom_vline(xintercept = 0, size = 0.2) +
  geom_point(alpha = 0.5, stroke = 0) +
  #geom_text_repel(aes(label = Gene), size =2) +
  scale_colour_manual(values = c("dodgerblue4", "lightgrey", "indianred")) +
  facet_grid(dataset ~ n_fuc) +
  theme_bw() +
  theme(panel.grid = element_blank()) +
  labs(x = "log2 fold-change sol_glyco-sol_total", y = "-log10 p-value") 


ggplot(
  res_mouse_brain_solubility$limma_results %>% 
    mutate(n_fuc = str_extract(Observed.Modifications, "Fuc\\(\\d+\\)"),
           n_sia = str_extract(Observed.Modifications, "NeuGc\\(\\d+\\)|NeuAc\\(\\d+\\)")) %>% 
    filter(!is.na(n_sia)),
  aes(x = logFC, y = -log10(P.Value), colour = direction)
) +
  geom_vline(xintercept = 0, size = 0.2) +
  geom_point(alpha = 0.5, stroke = 0) +
  #geom_text_repel(aes(label = Gene), size =2) +
  scale_colour_manual(values = c("dodgerblue4", "lightgrey", "indianred")) +
  facet_grid(dataset ~ n_sia) +
    theme_bw() +
  theme(panel.grid = element_blank()) +
  labs(x = "log2 fold-change sol_glyco-sol_total", y = "-log10 p-value")
```

```{r}
res_mouse_brain_solubility$limma_results %>% 
  group_by(dataset,glycan_type, hit, direction) %>% 
  count() %>% 
  group_by(dataset,hit, direction) %>% 
  mutate(total = sum(n)) %>% 
  ungroup() %>% 
  mutate(fraction = n/total) %>% 
  #filter(hit == "hit") %>% 
  mutate(glycan_type = factor(glycan_type, levels = order)) %>% 
  ggplot(aes(x = glycan_type, y = fraction, fill = glycan_type)) +
  geom_bar(stat = "identity", pos = "dodge") +
  geom_text(aes(label = round(fraction, 2)), angle = 90, hjust = -0.2) +
  facet_grid(dataset~direction) +
  lims(y = c(0, 0.8)) +
  scale_fill_manual(values = pal) +
  theme(axis.text.x = element_blank())

res_mouse_brain_solubility$limma_results %>% 
  filter(Protein.ID %in% res_mouse_brain_solubility$proteins_with_insol_sub$Protein.ID) %>% 
  mutate(glycan_type = factor(glycan_type, levels = order)) %>% 
  ggplot(aes(x = glycan_type, y = logFC, colour = glycan_type)) +
  geom_hline(yintercept = 0) +
  ggforce::geom_sina() +
  geom_boxplot(colour = "grey", outlier.colour = "white", fill =NA, width = 0.3) +
  geom_violin(fill = NA) +
  scale_colour_brewer(palette = "Set1") +
  facet_wrap(~dataset) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(title = "proteins with at least 25% insoluble")

```

```{r}
res_mouse_brain_solubility$limma_results %>% 
  filter(dataset == "ratio_comp_all" & !(glycan_type %in% c("paucimannose", "small", "phospho"))) %>% 
   filter(Protein.ID %in% res_mouse_brain_solubility$proteins_with_insol_sub$Protein.ID) %>% 
  mutate(glycan_type = factor(glycan_type, levels = order)) %>% 
  ggplot(aes(x = glycan_type, y = logFC, colour = glycan_type)) +
  geom_hline(yintercept = 0) +
  ggforce::geom_sina() +
  geom_boxplot(colour = "grey", outlier.colour = "white", fill =NA, width = 0.3) +
  geom_violin(fill = NA) +
  scale_colour_brewer(palette = "Set1") +
  stat_compare_means(method = "anova", label.y = 5)+      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = "sialylated")  +
  facet_wrap(~dataset) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(title = "proteins with at least 25% insoluble") 
```


```{r}
library(org.Mm.eg.db)
#GO:0015630 microtub cy
# 0007010

# GO:0007155



# ceell_junction GO:0030054
# 
# GO:0099561

ecm <- AnnotationDbi::select(org.Mm.eg.db, keytype="GOALL", keys=c("GO:0030054"), columns="UNIPROT")

ioncahnnel <- AnnotationDbi::select(org.Mm.eg.db, keytype="GOALL", keys=c("GO:0034702"), columns="UNIPROT")

# res_mouse_brain_solubility$limma_results %>% 
#   filter(dataset == "ratio_comp_all" & Protein.ID %in% res_mouse_brain_solubility$proteins_with_insol_sub$Protein.ID) %>% 
#    mutate(anno = ifelse(Protein.ID %in% ecm$UNIPROT, "ion_channel_complex", "other")) %>%
#   mutate(anno = ifelse(Protein.ID %in% ioncahnnel$UNIPROT, "cell_junction", anno)) %>%
#   mutate(glycan_type = factor(glycan_type, levels = order)) %>% 
#   ggplot(aes(x = anno, y = logFC, colour = anno)) +
#   geom_hline(yintercept = 0) +
#   ggforce::geom_sina() +
#   geom_boxplot(colour = "grey", outlier.colour = NA, fill =NA, width = 0.3) +
#   geom_violin(fill = NA) +
#   scale_colour_brewer(palette = "Set1") +
#   facet_wrap(~dataset) +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) 

res_mouse_brain_solubility$limma_results %>% 
  filter(dataset == "ratio_comp_all" & Protein.ID %in% res_mouse_brain_solubility$proteins_with_insol_sub$Protein.ID & Protein.ID %in% ecm$UNIPROT) %>% 
  mutate(pathway = "synaptic membrane adhesion\nto extracellular matrix") %>%
  bind_rows(
  # res_mouse_brain_solubility$limma_results %>% 
  # filter(dataset == "ratio_comp_all" & Protein.ID %in% res_mouse_brain_solubility$proteins_with_insol_sub$Protein.ID & Protein.ID %in% ioncahnnel$UNIPROT) %>% 
  # mutate(pathway = "ion channel complex"), 
  res_mouse_brain_solubility$limma_results %>% 
  filter(dataset == "ratio_comp_all" & Protein.ID %in% res_mouse_brain_solubility$proteins_with_insol_sub$Protein.ID & !(Protein.ID %in% ioncahnnel$UNIPROT) & !(Protein.ID %in% ecm$UNIPROT)) %>% 
  mutate(pathway = "other")
  ) %>% 
  ggplot(aes(x = pathway, y = logFC, colour = pathway)) +
  geom_hline(yintercept = 0) +
  ggforce::geom_sina(size = 2, alpha = 0.2) +
  stat_summary(aes(group = pathway), fun = median, geom = "crossbar", colour = "black") +
  geom_boxplot(colour = "black", outlier.colour = NA, fill =NA, width = 0.3) +
  ggpubr::stat_compare_means(aes(group = pathway), label = "p.signif",method = "t.test", comparisons = list(c("other","synaptic membrane adhesion\nto extracellular matrix"))) +
  #facet_wrap(~glycan_type) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), legend.position = "none") 
  ylim(c(-2.5, 6.5))
```


# Save

```{r}
save(res_mouse_brain_solubility, file = paste0("results/", Sys.Date(), "_res_mouse_brain_solubility.RData"))
```

