---
title: "Quantitative N-glycoprotoeme profiling of mouse brain conditioned with different microbiome groups"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2,
  fig.height = 4,
  dpi = 700
)
knitr::opts_knit$set(root.dir = "~/Documents/Glycoproteomics/")
```


## Packages

```{r, message=F, warning =F, include=F}
library(tidyverse)
library(ggplot2)
theme_set(cowplot::theme_cowplot())
library("reshape2")
library(ggrepel)
options(connectionObserver = NULL)
library(org.Mm.eg.db)
library(knitr)
library(ggpubr)
library(RColorBrewer)
library(ReactomePA)
library(clusterProfiler)
library(reactome.db)
library(ggvenn)
library(ComplexHeatmap)
library(STRINGdb)
library(babelgene)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by

source("~/Documents/generic_functions.R")
```

```{r}
options(ggplot2.discrete.colour = c("purple4", "goldenrod", "darkgreen", "darkorange", "darkblue", "darkred", "darkcyan"))
```

## Functions

```{r}
normalisation_melted_FP <- function(df) {
  m <- acast(df, Protein.ID ~ sample, value.var = "quant", fill = NA)
  m_norm <- limma::normalizeVSN(m)
  vsn::meanSdPlot(m_norm)
  df <- reshape2::melt(m_norm, value.name = "quant_norm", varnames = c("Protein.ID", "sample")) %>% left_join(df, by = c("Protein.ID", "sample"))
  return(tibble::tibble(df))
}

normalisation_melted_glyco <- function(df) {
  m <- acast(df, ID ~ sample, value.var = "quant", fun.aggregate = median)
  m_norm <- limma::normalizeVSN(m)
  vsn::meanSdPlot(m_norm)
  df <- melt(m_norm, value.name = "quant_norm", varnames = c("ID", "sample")) %>% left_join(df, by = c("ID", "sample"))
  return(tibble::tibble(df))
}
```

# 1. Glyco data

```{r}
# glyco psm
load("data/paper/processed/2023-05-16_microbiome_mousebrain_Nglyco_psm_processed.RData")
```

## TMT anno

Build info for each TMT channel

```{r}
res_microbiome_mousebrain_Nglyco$TMT_info_glyco <- data.frame(
  sample = c(
    "sample01", "sample02", "sample03",
    "sample04", "sample05", "sample06",
    "sample07", "sample08", "sample09",
    "sample10", "sample11", "sample12",
    "sample13", "sample14", "sample15",
    "sample16", "sample17", "sample18"
  ),
  microbiome = c(
    "germfree", "germfree", "germfree",
    "germfree", "germfree", "germfree",
    "monocolonized", "monocolonized", "monocolonized",
    "monocolonized", "monocolonized", "monocolonized",
    "community", "community", "community",
    "community", "community", "community"
  ),
  sex = c(
    "male", "female", "female",
    "male", "male", "female",
    "male", "male", "male",
    "female", "female", "female",
    "male", "male", "male",
    "female", "female", "female"
  ),
  replicate = c(
    "rep1", "rep2", "rep3",
    "rep4", "rep5", "rep6",
    "rep1", "rep2", "rep3",
    "rep4", "rep5", "rep6",
    "rep1", "rep2", "rep3",
    "rep4", "rep5", "rep6"
  )
)
```

## Normalisation

Basic formatting. Get relevant columns and normalize using normalisation_melted_glyco()

```{r}
res_microbiome_mousebrain_Nglyco$psm_anno <- res_microbiome_mousebrain_Nglyco$psm %>%
  # filter for glycopeptides
  filter(grepl("\\d+N", Assigned.Modifications)) %>%
  select(Peptide, Protein.ID, Protein.Start, n_position, matches("Sample"), `Modified.Peptide`, Gene, Assigned.Modifications, Observed.Modifications, glycan_type) %>%
  melt(
    id.vars = c("Peptide", "Protein.ID", "Protein.Start", "n_position", "Modified.Peptide", "Gene", "Assigned.Modifications", "Observed.Modifications", "glycan_type"),
    variable.name = "sample",
    value.name = "quant"
  ) %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_glyco, by = "sample") %>%
  filter(!is.na(sex)) %>%
  mutate(ID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>%
  # group_by(condition) %>%
  group_modify(~ tibble(normalisation_melted_glyco(.))) %>%
  mutate(
    glycosylation = as.character(str_extract_all(Assigned.Modifications, "\\d+N")),
    #glyco_position = n_position,
    condition = paste0(microbiome, "_", sex)
  )
```

## Check normalization

```{r}
ggplot(res_microbiome_mousebrain_Nglyco$psm_anno, aes(y = log2(quant), x = paste0(sex, "_", replicate), colour = microbiome)) +
  geom_boxplot() +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  facet_wrap(~microbiome, scales = "free")

ggplot(res_microbiome_mousebrain_Nglyco$psm_anno, aes(y = quant_norm, x = paste0(sex, "_", replicate), colour = microbiome)) +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"))+
  facet_wrap(~microbiome, scales = "free_x") +
  labs(x = "", title = "Glycoproteome")
```

## Numbers

```{r}
res_microbiome_mousebrain_Nglyco$psm_anno %>%
  mutate(siteID = paste0(Protein.ID, "_", n_position)) %>%
  summarise(
    n_glycoproteins = n_distinct(Protein.ID),
    n_glycopeptides = n_distinct(ID),
    n_glycosites = n_distinct(siteID)
  ) %>%
  melt()
```

## Correlations

### heatmap
of raw intensities, with complex heatmap

```{r}
m <- res_microbiome_mousebrain_Nglyco$psm %>%
  select(matches(res_microbiome_mousebrain_Nglyco$TMT_info_glyco$sample)) %>%
  setNames(paste0(res_microbiome_mousebrain_Nglyco$TMT_info_glyco$microbiome, "_", res_microbiome_mousebrain_Nglyco$TMT_info_glyco$replicate))

# remove glycopsms with missing quantification information
m[m == 0] <- NA
m <- drop_na(m)

ComplexHeatmap::Heatmap(m %>% cor(m, method = "spearman"),
  col = corrplot::COL1("Reds"),
  # cluster_columns = F,
  show_column_names = FALSE,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  rect_gp = gpar(col = "white"),
  right_annotation = row_anno <- HeatmapAnnotation(
    "microbiome" = res_microbiome_mousebrain_Nglyco$TMT_info_glyco$microbiome,
    "sex" = res_microbiome_mousebrain_Nglyco$TMT_info_glyco$sex,
    col = list(
      microbiome = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"),
      sex = c("male" = "#E1BE6A", "female" = "#40B0A6")
    ),
    which = "row",
    show_annotation_name = FALSE
  ),
  name = "corr"
)
```

### pairwise scatterplots

```{r, eval = F}
GGscatterhex <- function(data, mapping, ...) {
  # function to do pairwise correlation plot with hex instead og points
  x <- GGally::eval_data_col(data, mapping$x)
  y <- GGally::eval_data_col(data, mapping$y)
  df <- data.frame(x = x, y = y)
  pp <- ggplot(df, aes(x = x, y = y)) +
    geom_hex()
  return(pp)
}

GGally::ggpairs(log10(m),
  columns = c(1:18),
  upper = list(continuous = GGally::wrap("cor", method = "spearman", size = 2)),
  lower = list(continuous = GGally::wrap(GGscatterhex))
) +
  cowplot::theme_cowplot(9)
```

### correlation summary table

Check correlation o samples of differnet microbiomes and sex

```{r}
m %>%
  cor(method = "spearman") %>%
  melt() %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_glyco %>% mutate(Var1 = paste0(microbiome, "_", replicate)) %>% distinct(`Var1`, `sex`)) %>%
  mutate(sex1 = sex) %>%
  select(-sex) %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_glyco %>% mutate(Var2 = paste0(microbiome, "_", replicate)) %>% distinct(`Var2`, `sex`)) %>%
  mutate(sex2 = sex) %>%
  select(-sex) %>%
  mutate(
    Var1 = str_replace(Var1, "_rep\\d", ""),
    Var2 = str_replace(Var2, "_rep\\d", "")
  ) %>%
  filter(Var1 == Var2 & sex1 == sex2) %>%
  distinct(Var1, sex1, value) %>%
  group_by(Var1, sex1) %>%
  summarise(mean_corr = mean(value), min_corr = min(value))
```

# 2. Full proteome data

For full proteome, start from protein.tsv output from MSFragger.

```{r}
res_microbiome_mousebrain_Nglyco$psm_FP <- read_tsv("data/paper/psm_files/TMT/Glyco_mousebrain_threeconditions_FP_protein_1protpergene.tsv")
colnames(res_microbiome_mousebrain_Nglyco$psm_FP) <- gsub(" ", "\\.", colnames(res_microbiome_mousebrain_Nglyco$psm_FP))
```

The TMT columns have a different name everywhere. Hard overwriting here, but I don't want to rewrite everything.

```{r}
colnames(res_microbiome_mousebrain_Nglyco$psm_FP)[24:41] <- c(
  "sample01", "sample02", "sample03",
  "sample04", "sample05", "sample06",
  "sample07", "sample08", "sample09",
  "sample10", "sample11", "sample12",
  "sample13", "sample14", "sample15",
  "sample16", "sample17", "sample18"
)

colnames(res_microbiome_mousebrain_Nglyco$psm_FP)
```

## Remove contaminants

Reomve some proteins which are not Mus musculus.

```{r}
res_microbiome_mousebrain_Nglyco$psm_FP <- res_microbiome_mousebrain_Nglyco$psm_FP %>%
  filter(grepl("Mus musculus", Organism))
```

## TMT anno

Build info for each TMT channel

```{r}
res_microbiome_mousebrain_Nglyco$TMT_info_FP <- data.frame(
  sample = c(
    "sample01", "sample02", "sample03",
    "sample04", "sample05", "sample06",
    "sample07", "sample08", "sample09",
    "sample10", "sample11", "sample12",
    "sample13", "sample14", "sample15",
    "sample16", "sample17", "sample18"
  ),
  microbiome = c(
    "germfree", "germfree", "germfree",
    "germfree", "germfree", "germfree",
    "monocolonized", "monocolonized", "monocolonized",
    "monocolonized", "monocolonized", "monocolonized",
    "community", "community", "community",
    "community", "community", "community"
  ),
  sex = c(
    "male", "female", "female",
    "male", "male", "female",
    "male", "male", "male",
    "female", "female", "female",
    "male", "male", "male",
    "female", "female", "female"
  ),
  replicate = c(
    "rep1", "rep2", "rep3",
    "rep4", "rep5", "rep6",
    "rep1", "rep2", "rep3",
    "rep4", "rep5", "rep6",
    "rep1", "rep2", "rep3",
    "rep4", "rep5", "rep6"
  )
)
```

## Normalisation

```{r}
res_microbiome_mousebrain_Nglyco$psm_FP_anno <- res_microbiome_mousebrain_Nglyco$psm_FP %>%
  select(`Protein.ID`, Gene, matches(res_microbiome_mousebrain_Nglyco$TMT_info_FP$sample)) %>%
  melt(variable.name = "sample", value.name = "quant") %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_FP, by = c("sample")) %>%
  filter(!is.na(sex) & quant > 0 & is.finite(quant)) %>%
  mutate(dummy = "dummy") %>%
  group_by(dummy) %>%
  group_modify(~ tibble(normalisation_melted_FP(.))) %>%
  ungroup() %>%
  select(-dummy) %>%
  mutate(condition = paste0(microbiome, "_", sex)) %>%
  filter(!is.na(quant) & !is.na(quant_norm))
```

## Check normalization

```{r}
ggplot(res_microbiome_mousebrain_Nglyco$psm_FP_anno, aes(y = log2(quant), x = paste0(sex, "_", replicate), colour = microbiome)) +
  geom_boxplot() +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"))+
  facet_wrap(~microbiome, scales = "free_x")

ggplot(res_microbiome_mousebrain_Nglyco$psm_FP_anno, aes(y = quant_norm, x = paste0(sex, "_", replicate), colour = microbiome)) +
  geom_boxplot() +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  facet_wrap(~microbiome, scales = "free_x") +
  labs(x = "", title = "Full proteome")
```

## correlations
### Heatmap

```{r}
ComplexHeatmap::Heatmap(
  res_microbiome_mousebrain_Nglyco$psm_FP %>%
    select(Protein.ID, matches(res_microbiome_mousebrain_Nglyco$TMT_info_FP$sample)) %>%
    group_by(Protein.ID) %>%
    summarise(across(everything(), list(sum))) %>%
    select(-Protein.ID) %>%
    setNames(paste0(res_microbiome_mousebrain_Nglyco$TMT_info_FP$microbiome, "_", res_microbiome_mousebrain_Nglyco$TMT_info_FP$replicate)) %>%
    cor(method = "spearman"),
  col = corrplot::COL1("Reds"),
  cluster_columns = F,
  cluster_rows = F,
  show_column_names = FALSE,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  rect_gp = gpar(col = "white"),
  right_annotation = HeatmapAnnotation(
    "microbiome" = res_microbiome_mousebrain_Nglyco$TMT_info_FP$microbiome,
    "sex" = res_microbiome_mousebrain_Nglyco$TMT_info_FP$sex,
    col = list(
      microbiome = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen"),
      sex = c("male" = "#E1BE6A", "female" = "#40B0A6")
    ),
    which = "row",
    show_annotation_name = FALSE
  ),
  name = "corr"
)
```

# 3. Correct data: Fit lm

## Combine glyco and total

```{r}
length(unique(res_microbiome_mousebrain_Nglyco$psm_anno$Protein.ID))
length(unique(res_microbiome_mousebrain_Nglyco$psm_FP_anno$Protein.ID))

length(intersect(unique(res_microbiome_mousebrain_Nglyco$psm_anno$Protein.ID), unique(res_microbiome_mousebrain_Nglyco$psm_FP_anno$Protein.ID)))
```


```{r}
data_fit <- res_microbiome_mousebrain_Nglyco$psm_anno %>%
  # format glyco data
  filter(quant_norm > 0) %>%
  # remove psm duplicates which arise from normalisation
  distinct(ID, sample, quant_norm, .keep_all = T) %>%
  dplyr::select(Protein.ID, glycoID = ID, sample, glyco_intensity = quant_norm) %>%
  # add FP data
  inner_join(res_microbiome_mousebrain_Nglyco$psm_FP_anno %>%
    filter(quant_norm > 0) %>%
    # mutate(value = log2(quant_norm)) %>%
    dplyr::select(Protein.ID, sample, total_intensity = quant_norm), by = c("sample", "Protein.ID"))
```

## Fit individual lm per glycopeptide

```{r}
data_lm <- data_fit %>%
  filter(!is.na(glyco_intensity), !is.na(total_intensity)) %>%
  nest(data = c(sample, glyco_intensity, total_intensity)) %>%
  mutate(fit = purrr::map(data, function(df) lm(glyco_intensity ~ total_intensity, data = df)))

data_lm <- data_lm %>%
  mutate(model = purrr::map(fit, broom::tidy))

data_lm <- data_lm %>%
  mutate(glyco_pred = map(fit, predict))

data_lm <- data_lm %>%
  dplyr::select(-fit) %>%
  unnest(c(data, glyco_pred))

data_lm <- data_lm %>%
  dplyr::select(-model) %>%
  mutate(cleaned_glycosignal = glyco_intensity - glyco_pred)
```

### Comparison corrected to not corrected

```{r}
data_cor <- data_fit %>%
  filter(!is.na(glyco_intensity), !is.na(total_intensity)) %>%
  group_by(Protein.ID, glycoID) %>%
  summarise(cor_intensity = cor(glyco_intensity, total_intensity)) %>%
  ungroup()

data_cor_after <- data_lm %>%
  ungroup() %>%
  filter(!is.na(cleaned_glycosignal), !is.na(total_intensity)) %>%
  group_by(glycoID) %>%
  summarise(cor_intensity = cor(cleaned_glycosignal, total_intensity)) %>%
  ungroup()

ggplot(data_cor_after, aes(x = cor_intensity)) +
  geom_histogram(aes(y = ..density..), alpha = 0.3, colour = "gray") +
  geom_density(colour = "maroon") +
  scale_x_continuous(limits = c(-0.1, 0.1)) +
  theme_bw(base_size = 16) +
  labs(x = "R2 corrected glycoproteome vs total proteome", title = "Correlation after")

ggplot(data_cor, aes(x = cor_intensity)) +
  geom_histogram(aes(y = ..density..), alpha = 0.3, colour = "gray") +
  geom_density(colour = "maroon") +
  theme_bw(base_size = 16) +
  labs(x = "R2 glycoproteome vs total proteome", title = "Correlation before")

rm(data_cor, data_cor_after)
```

```{r}
res_microbiome_mousebrain_Nglyco$psm_anno <- data_lm %>%
  left_join(res_microbiome_mousebrain_Nglyco$psm_anno, by = c("glycoID" = "ID", "sample", "Protein.ID"))
```

```{r}
rm(data_fit, data_lm)
```

# 4. PCA

## glyco

```{r}
m <- acast(res_microbiome_mousebrain_Nglyco$psm_anno, glycoID ~ sample, value.var = "cleaned_glycosignal", fill = 0, fun.aggregate = median)

m[m == 0] <- NA

m <- m %>%
  as.data.frame() %>%
  drop_na() %>%
  as.matrix() %>%
  t()

# m <- m[matrixStats::rowVars(m, na.rm = T) %>% order(decreasing = T) %>%
#        head(6000), ]

pca <- prcomp(m)
plot(pca)
var <- (pca$sdev)^2 / sum(pca$sdev^2) * 100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_glyco) %>%
  ggplot(aes(x = PC1, y = PC2, colour = microbiome, shape = sex)) +
  geom_point(size = 3) +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
  scale_shape_manual(values = c(15, 17)) +
  labs(
    title = "Glycoproteome",
    subtitle = "normalized and corrected",
    x = paste0("PC1 (", as.character(round(var[1])), "%)"),
    y = paste0("PC2 (", as.character(round(var[2])), "%)")
  )

pca_rotation <- pca$rotation %>% as.data.frame()

rm(m, pca_rotation)
```

## full proteome

```{r}
m <- dcast(res_microbiome_mousebrain_Nglyco$psm_FP_anno, Protein.ID ~ sample, value.var = "quant_norm", fill = NA) %>%
  drop_na() %>%
  column_to_rownames("Protein.ID") %>%
  as.matrix() %>%
  t()

pca <- prcomp(m)
plot(pca)

var <- (pca$sdev)^2 / sum(pca$sdev^2) * 100

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(res_microbiome_mousebrain_Nglyco$TMT_info_FP) %>%
  ggplot(aes(x = round(PC1), y = PC2, colour = microbiome, shape = sex)) +
  geom_point(size = 3) +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
  scale_shape_manual(values = c(15, 17)) +
  labs(
    title = "Proteome",
    subtitle = "normalized",
    x = paste0("PC1 (", as.character(round(var[1])), "%)"),
    y = paste0("PC2 (", as.character(round(var[2])), "%)")
  )

pca_rotation <- pca$rotation %>% as.data.frame()
rm(m, pca_rotation)
```


# 5. Limma

## Input

```{r}
# remove features with incomplete quantification
pdata_FP <- res_microbiome_mousebrain_Nglyco$psm_FP_anno %>%
  acast(`Protein.ID` ~ microbiome + replicate, value.var = "quant_norm")

pdata_FP[is.na(pdata_FP)] <- 0
pdata_FP <- pdata_FP[rowSums(pdata_FP == 0) <= 0, ]

pdata_glyco <- res_microbiome_mousebrain_Nglyco$psm_anno %>%
  acast(glycoID + Observed.Modifications ~ microbiome + replicate, value.var = "cleaned_glycosignal", fun.aggregate = median)

pdata_glyco[is.na(pdata_glyco)] <- 0
pdata_glyco <- pdata_glyco[rowSums(pdata_glyco == 0) <= 0, ]


boxplot(pdata_FP)
boxplot(pdata_glyco)
```

## FP

### Design matrix

```{r}
# microbiome group
g <- gsub("_rep\\d+$", "", colnames(pdata_FP))
# sex
sex <-
  c(
    "male", "female", "female",
    "male", "male", "female",
    "male", "male", "male",
    "female", "female", "female",
    "male", "male", "male",
    "female", "female", "female"
  )

# build design matrix
design_matrix <- model.matrix(~ 0 + g + sex)
colnames(design_matrix) <- c("community", "germfree", "monocolonized", "sex")
design_matrix
```

### Fit

```{r}
fit <- limma::lmFit(pdata_FP, design_matrix)

contr <- cbind(
  "community-germfree" = c(1, -1, 0, 0),
  "monocolonized-germfree" = c(0, -1, 1, 0),
  "community-monocolonized" = c(1, 0, -1, 0)
)

fit <- limma::contrasts.fit(fit, contr)
fit <- limma::eBayes(fit)
```

Extract results

```{r}
topTable_FP <- lapply(colnames(fit$coefficients), function(x) {
  limma::topTable(fit, coef = x, number = Inf) %>%
    mutate(contrast = x, dataset = "full proteome") %>%
    rownames_to_column(var = "Protein.ID")
}) %>%
  bind_rows()
```

## glyco

### Design matrix

stays the same

```{r}
# microbiome group
g <- gsub("_rep\\d+$", "", colnames(pdata_glyco))
# sex
sex <-
  c(
    "male", "female", "female",
    "male", "male", "female",
    "male", "male", "male",
    "female", "female", "female",
    "male", "male", "male",
    "female", "female", "female"
  )


design_matrix <- model.matrix(~ 0 + g + sex)
colnames(design_matrix) <- c("community", "germfree", "monocolonized", "sex")

design_matrix
```

### Fit

```{r}
fit <- limma::lmFit(pdata_glyco, design_matrix)

contr <- cbind(
  "community-germfree" = c(1, -1, 0, 0),
  "monocolonized-germfree" = c(0, -1, 1, 0),
  "community-monocolonized" = c(1, 0, -1, 0)
)

fit <- limma::contrasts.fit(fit, contr)
fit <- limma::eBayes(fit)
```

Extract results

```{r}
topTable_glyco <- lapply(colnames(fit$coefficients), function(x) {
  limma::topTable(fit, coef = x, number = Inf) %>%
    mutate(contrast = x, dataset = "glycoproteome") %>%
    rownames_to_column(var = "ID")
}) %>%
  bind_rows() %>%
  separate(ID, into = c("Protein.ID", "Modified.Peptide", "Observed.Modifications"), sep = "_", remove = F)
```

## Results

### combine results

```{r}
res_microbiome_mousebrain_Nglyco$limma_results <- bind_rows(topTable_glyco, topTable_FP) %>%
  mutate(
    hit = ifelse((abs(logFC) > log2(1.5) & adj.P.Val < 0.05), "hit", "no hit")
  ) %>%
  filter(!is.na(hit)) %>%
  mutate(contrast = factor(contrast, levels = c("monocolonized-germfree", "community-germfree", "community-monocolonized"))) %>%
  left_join(
    # add additional information
    # from FP data
    distinct(res_microbiome_mousebrain_Nglyco$psm_FP, Protein.ID, Gene, Protein.Description) %>%
      bind_rows(
        # from glyco data
        distinct(res_microbiome_mousebrain_Nglyco$psm, Protein.ID, Gene, Protein.Description, Modified.Peptide, Observed.Modifications, glycan_type, n_position)
      ) %>%
      distinct(Protein.ID, Gene, Protein.Description, Modified.Peptide, Observed.Modifications, glycan_type, n_position)
  ) %>%
  mutate(
    ID = paste(Protein.ID, Modified.Peptide, Observed.Modifications, sep = "_"),
    n_position = n_position + 1
  ) %>% 
  mutate(
      direction = ifelse(logFC > 0 & hit == "hit", "up", "no change"),
      direction = ifelse(logFC < 0 & hit == "hit", "down", direction)
    )
```

### Numbers


```{r}
res_microbiome_mousebrain_Nglyco$limma_results %>%
  distinct(dataset, contrast, Protein.ID, hit) %>%
  group_by(dataset, contrast, hit) %>%
  count()
```

```{r}
rm(design_matrix, contr, fit, pdata_FP, pdata_glyco, g, topTable_FP, topTable_glyco, sex, var, row_anno)
```

### Volcano

```{r}
ggplot(
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    mutate(
      contrast = str_replace(contrast, "-", "-\n"),
      contrast = factor(contrast, levels = c("monocolonized-\ngermfree", "community-\ngermfree", "community-\nmonocolonized"))
    ),
  aes(x = logFC, y = -log10(adj.P.Val), colour = direction)
) +
  geom_point(alpha = 0.5, stroke = 0) +
  scale_colour_manual(values = c("dodgerblue4", "lightgrey", "indianred")) +
  facet_grid(dataset ~ contrast) +
  # theme_bw(14) +
  theme(panel.grid = element_blank()) +
  labs(x = "log2 fold-change", y = "-log10 adjusted p-value")
```

### up and down per protein

```{r, eval = F}
res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "glycoproteome" & hit == "hit") %>%
  mutate(ext_anno = ifelse(logFC > 0, "hit_up", "hit_down")) %>%
  group_by(Protein.ID, Gene, Protein.Description, contrast, ext_anno) %>%
  count() %>%
  dcast(Protein.ID + Gene + Protein.Description + contrast ~ ext_anno)
```

### heatmap of hits

```{r}
hits <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(hit == "hit")

m <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "glycoproteome" & contrast != "community-monocolonized" & ID %in% hits$ID) %>%
  acast(ID ~ contrast, value.var = "logFC")

m %>%
  ComplexHeatmap::Heatmap(
    cluster_columns = T,
    # number of clusters
    km = 5,
    show_row_dend = F,
    show_column_dend = T,
    show_row_names = F,
    row_names_gp = grid::gpar(fontsize = 8),
    col = circlize::colorRamp2(c(-3, 0, 3), c("darkblue", "white", "darkred")),
    name = "Overview"
  )

rm(m, hits)
```



### Overlaps

#### glyco-FP

```{r}
res_microbiome_mousebrain_Nglyco$limma_results %>%
  mutate(dataset = ifelse(dataset == "full proteome", "FP", "glyco")) %>%
  filter(hit == "hit" & contrast == "community-germfree") %>%
  mutate(class = ifelse(logFC > 0, "up", "down")) %>%
  distinct(dataset, class, Protein.ID) %>%
  dcast(Protein.ID ~ class + dataset) %>%
  mutate(
    down_FP = ifelse(is.na(down_FP), FALSE, TRUE),
    down_glyco = ifelse(is.na(down_glyco), FALSE, TRUE),
    up_FP = ifelse(is.na(up_FP), FALSE, TRUE),
    up_glyco = ifelse(is.na(up_glyco), FALSE, TRUE)
  ) %>%
  ggplot() +
  ggvenn::geom_venn(
    aes(
      A = `down_FP`,
      B = `down_glyco`,
      C = `up_FP`,
      D = `up_glyco`,
    ),
    fill_color = c("darkorange", "darkblue", "darkred", "darkgreen"),
    fill_alpha = 1,
    stroke_alpha = 1,
    stroke_color = "white",
    set_name_size = 5,
    # set_names = c("HEK", "HeLa"),
    text_size = 4,
    text_color = "white",
    show_percentage = FALSE
  ) +
  coord_fixed() +
  theme_void() +
  labs(title = "Full proteome and glycoproteome community-germfree")

res_microbiome_mousebrain_Nglyco$limma_results %>%
  mutate(dataset = ifelse(dataset == "full proteome", "FP", "glyco")) %>%
  filter(hit == "hit") %>%
  # mutate(class =ifelse(logFC > 0, "up", "down")) %>%
  distinct(dataset, Protein.ID) %>%
  dcast(Protein.ID ~ dataset) %>%
  mutate(
    FP = ifelse(is.na(FP), FALSE, TRUE),
    glyco = ifelse(is.na(glyco), FALSE, TRUE)
  ) %>%
  ggplot() +
  ggvenn::geom_venn(
    aes(
      A = `FP`,
      B = `glyco`
    ),
    fill_color = c("darkorange", "darkblue", "darkred", "darkgreen"),
    fill_alpha = 1,
    stroke_alpha = 1,
    stroke_color = "white",
    set_name_size = 5,
    # set_names = c("HEK", "HeLa"),
    text_size = 4,
    text_color = "white",
    show_percentage = FALSE
  ) +
  coord_fixed() +
  theme_void() +
  labs(title = "Full proteome and \nglycoproteome all contrasts")
```

#6. Enrichments

```{r}
res_microbiome_Nglyco_ER <- list()
```


## Reactome

### glyco

```{r}
input <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "glycoproteome", hit == "hit") %>%
  distinct(contrast, Gene)


universe <- mapIds(org.Mm.eg.db, unique(c(res_microbiome_mousebrain_Nglyco$psm$Gene, res_microbiome_mousebrain_Nglyco$psm_FP$Gene)), "ENTREZID", "SYMBOL") %>%
  unlist() %>%
  unname()

res_microbiome_Nglyco_ER$reactomeER_glyco <- input %>%
  group_by(contrast) %>%
  nest() %>%
  mutate(ER = map(data, function(df) {
    ER <- ReactomePA::enrichPathway(
      gene = mapIds(org.Mm.eg.db, df$Gene, "ENTREZID", "SYMBOL"),
      organism = "mouse",
      pvalueCutoff = 1,
      pAdjustMethod = "BH",
      qvalueCutoff = 1,
      universe = universe,
      minGSSize = 5,
      maxGSSize = 500,
      readable = TRUE
    )@result
  })) %>%
  select(-data) %>%
  unnest(cols = c(ER))

interesting_pathways <- res_microbiome_Nglyco_ER$reactomeER_glyco %>%
  mutate(Description = str_replace(Description, "Mus musculus\r:", "")) %>%
  filter(Count > 3 & p.adjust < 0.05 & contrast != "community-monocolonized") %>%
  group_by(contrast) %>%
  arrange(p.adjust) %>%
  slice_head(n = 10)

res_microbiome_Nglyco_ER$reactomeER_glyco %>%
  filter(Description %in% interesting_pathways$Description & contrast != "community-monocolonized") %>%
  ungroup() %>%
  arrange(p.adjust) %>%
  ggplot(aes(x = -log10(p.adjust), y = reorder(Description, log10(p.adjust)), size = Count, colour = -log10(p.adjust))) +
  geom_point() +
  geom_vline(xintercept = -log10(0.05)) +
  scale_size(range = c(4, 10)) +
  scale_color_gradientn(colors = corrplot::COL1("Reds")[50:200]) +
  theme(legend.position = "right") +
  facet_wrap(~contrast, nrow = 1) +
  theme(panel.grid.major.y = element_line(colour = "grey")) +
  labs(y = "", x = "-log10 adjusted p-value", title = "Reactomme ORA")

rm(input, interesting_pathways, universe)
```

Get subset of proteins involved in protein interactions at synapse

```{r}
res_microbiome_mousebrain_Nglyco$limma_result %>%
  filter(dataset == "glycoproteome") %>%
  left_join(res_microbiome_mousebrain_Nglyco$psm_anno %>% distinct(Protein.ID, Gene)) %>%
  filter(Gene %in% c("Grin1", "Ntrk3", "Grm5", "Nrxn1", "Lrrc4b", "Grin2b", "Nlgn1", "Grm1", "Gria1", "Grin2a", "Slitrk2", "Gria4", "Nrxn3", "Gria3", "Lrfn1", "Slitrk3")) %>%
  mutate(contrast = factor(contrast, levels = c("monocolonized-germfree", "community-germfree", "community-monocolonized"))) %>%
  ggplot(aes(x = logFC, y = -log10(adj.P.Val), colour = hit)) +
  geom_point(alpha = 0.5) +
  scale_colour_manual(values = c("#964E7C", "#99A0A9")) +
  facet_wrap(~contrast) +
  theme_bw(14) +
  theme(panel.grid = element_blank()) +
  labs(title = "Synapse protein interactions", x = "log2 fold-change", y = "-log10 adjusted p-value")
```


### FP

```{r}
input <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "full proteome", hit == "hit") %>%
  distinct(contrast, Gene)


universe <- mapIds(org.Mm.eg.db, unique(c(res_microbiome_mousebrain_Nglyco$psm$Gene, res_microbiome_mousebrain_Nglyco$psm_FP$Gene)), "ENTREZID", "SYMBOL") %>%
  unlist() %>%
  unname()

res_microbiome_Nglyco_ER$reactomeER_FP <- input %>%
  group_by(contrast) %>%
  nest() %>%
  mutate(ER = map(data, function(df) {
    ER <- ReactomePA::enrichPathway(
      gene = mapIds(org.Mm.eg.db, df$Gene, "ENTREZID", "SYMBOL"),
      organism = "mouse",
      pvalueCutoff = 1,
      pAdjustMethod = "BH",
      qvalueCutoff = 1,
      universe = universe,
      minGSSize = 5,
      maxGSSize = 500,
      readable = TRUE
    )@result
  })) %>%
  select(-data) %>%
  unnest(cols = c(ER))

interesting_pathways <- res_microbiome_Nglyco_ER$reactomeER_FP %>%
  mutate(Description = str_replace(Description, "Mus musculus\r:", "")) %>%
  filter(Count > 3 & p.adjust < 0.05 & contrast != "community-monocolonized") %>%
  group_by(contrast) %>%
  arrange(p.adjust) %>%
  slice_head(n = 10)

res_microbiome_Nglyco_ER$reactomeER_FP %>%
  filter(Description %in% interesting_pathways$Description & contrast != "community-monocolonized") %>%
  ungroup() %>%
  arrange(p.adjust) %>%
  ggplot(aes(x = -log10(p.adjust), y = reorder(Description, log10(p.adjust)), size = Count, colour = -log10(p.adjust))) +
  geom_point() +
  geom_vline(xintercept = -log10(0.05)) +
  scale_size(range = c(4, 10)) +
  scale_color_gradientn(colors = corrplot::COL1("Reds")[50:200]) +
  theme(legend.position = "right") +
  facet_wrap(~contrast, nrow = 1) +
  theme(panel.grid.major.y = element_line(colour = "grey")) +
  labs(y = "", x = "-log10 adjusted p-value", title = "Reactomme ORA")


rm(input, interesting_pathways, universe)
```

## Nice visualisation of interesting proteins

### glyco

```{r}
processes <- bind_rows(
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(grepl("Gria|Gabra|Glra|Grin|Grik|Grm|Gabbr|Gabrb|Gabrg|Gabrd", Gene)) %>%
    distinct(Gene, Protein.Description) %>%
    mutate(
      biol_process = str_extract(Protein.Description, "Glutamate|glutamate|Gamma"),
      biol_process = ifelse(grepl("Glutamate|glutamate", biol_process), "Glutamate\nreceptor", "GABA\nreceptor")
    ),
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(hit == "hit") %>%
    filter(grepl("channel", Protein.Description)) %>% distinct(Gene, Protein.Description) %>%
    mutate(biol_process = "Ion\nchannels"),
  data.frame(
    Gene = c("Sema4a", "Plxna4", "Dscam", "Grin2b", "Epha4", "Col6a1", "Ncstn", "Sema4d", "Dcc", "Dpysl2", "Plxnb3", "Plxna1", "Dag1", "Col6a2", "Itgb3", "Reln", "Grin1", "Itgb1"),
    biol_process = "Axon\nguidance"
  )
) %>%
  mutate(biol_process = factor(biol_process, levels = c("Axon\nguidance", "Glutamate\nreceptor", "Ion\nchannels", "GABA\nreceptor")))



hits <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "glycoproteome" & hit == "hit") %>%
  mutate(ID = paste0(Gene, "_", Modified.Peptide))

col_fun <- circlize::colorRamp2(c(-1.5, 0, 1.5), c("dodgerblue4", "white", "indianred"), transparency = 0)

anno <- processes %>%
  inner_join(res_microbiome_mousebrain_Nglyco$limma_results %>% filter(dataset == "glycoproteome") %>% distinct(Gene, Modified.Peptide, contrast, logFC), multiple = "all") %>%
  mutate(ID = paste0(Gene, "_", Modified.Peptide)) %>%
  as.data.frame() %>%
  filter(ID %in% hits$ID) %>%
  dcast(Gene + ID + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median)

ComplexHeatmap::Heatmap(
  processes %>%
    inner_join(res_microbiome_mousebrain_Nglyco$limma_results %>%
      filter(dataset == "glycoproteome") %>%
      distinct(Gene, Modified.Peptide, contrast, logFC), multiple = "all") %>%
    mutate(ID = paste0(Gene, "_", Modified.Peptide)) %>%
    filter(ID %in% hits$ID & contrast != "community-monocolonized") %>%
    mutate(contrast = str_replace(contrast, "-", "-\n")) %>%
    acast(ID + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median),
  col = col_fun,
  cluster_rows = T,
  show_column_names = T,
  row_split = anno$biol_process,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  right_annotation = HeatmapAnnotation(
    "process" = anno$biol_process,
    col = list(
      process = c("GABA\nreceptor" = "goldenrod", "Glutamate\nreceptor" = "darkred", "Ion\nchannels" = "darkcyan", "Axon\nguidance" = "mediumvioletred")
    ),
    which = "row",
    show_annotation_name = FALSE
  ),
  name = "log2 fold-change"
)
```

### FP

```{r}
processes <- bind_rows(
  res_microbiome_mousebrain_Nglyco$limma_results %>%
    filter(grepl("Gng", Gene)) %>%
    distinct(Gene, Protein.Description) %>%
    mutate(
      biol_process = "Gng\nproteins"
    ),
  data.frame(
    Gene = c("Ndufab1", "Cox19", "Slc25a14", "Uqcrh", "Uqcr11", "Cox14", "Cox7a2l", "Ndufa4", "Sdhaf4", "Cox5b", "Coa6", "Sdhaf2", "Ppp1r1a", "Mt3", "S100b", "Cox17", "Sdhd", "Adrb1", "Pdha2", "Cox7a2", "Cox7a1", "Sdhc"),
    biol_process = "Cellular\nrespiration"
  )
)

hits <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "full proteome") %>%
  mutate(ID = Gene)

col_fun <- circlize::colorRamp2(c(-1.5, 0, 1.5), c("dodgerblue4", "white", "indianred"), transparency = 0)

anno <- processes %>%
  inner_join(res_microbiome_mousebrain_Nglyco$limma_results %>% filter(dataset == "full proteome") %>% distinct(Gene, contrast, logFC), multiple = "all") %>%
  mutate(ID = Gene) %>%
  as.data.frame() %>%
  filter(ID %in% hits$ID) %>%
  dcast(Gene + ID + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median)

ComplexHeatmap::Heatmap(
  processes %>%
    inner_join(res_microbiome_mousebrain_Nglyco$limma_results %>%
      filter(dataset == "full proteome") %>%
      distinct(Gene, contrast, logFC), multiple = "all") %>%
    mutate(ID = Gene) %>%
    filter(ID %in% hits$ID & contrast != "community-monocolonized") %>%
    mutate(contrast = str_replace(contrast, "-", "-\n")) %>%
    acast(ID + biol_process ~ contrast, value.var = "logFC", fun.aggregate = median),
  col = col_fun,
  cluster_rows = T,
  show_column_names = T,
  row_split = anno$biol_process,
  show_column_dend = F,
  show_row_dend = F,
  show_row_names = F,
  right_annotation = HeatmapAnnotation(
    "process" = anno$biol_process,
    col = list(
      process = c("Cellular\nrespiration" = "grey60", "Gng\nproteins" = "palegreen4")
    ),
    which = "row",
    show_annotation_name = FALSE
  ),
  name = "log2 fold-change"
)
```

## String

```{r}
string_db <- STRINGdb$new(
  version = "11.5", species = 10090,
  score_threshold = 200, network_type = "full", input_directory = ""
)

background <- string_db$map(res_microbiome_mousebrain_Nglyco$limma_results, "Gene", removeUnmappedRows = TRUE)

string_db <- STRINGdb$new(
  version = "11.5", species = 10090,
  score_threshold = 200, network_type = "full", input_directory = "",
  backgroundV = unique(background$STRING_id)
)

res_microbiome_Nglyco_ER$ER_string_glyco <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "glycoproteome") %>%
  filter(hit == "hit") %>%
  distinct(contrast, Gene) %>%
  group_by(contrast) %>%
  nest(Gene) %>%
  mutate(ER = map(data, function(df) {
    ER <- string_db$get_enrichment(df$Gene)
  })) %>%
  select(-data) %>%
  unnest(cols = c(ER))
```




#7. Intensity plots

These plot show the glyco intensities per site protein-wise for the different conditions

## Receptors

Neurotransmitter receptors: Gria4/Gria3/Gabra3/Glrb/Glra2
Gria|Gabra|Glra|Grin|Grik|Grm|Gabbr|Gabrb|Gabrg|Gabrd

```{r}
res_microbiome_mousebrain_Nglyco$psm_anno %>%
   filter(grepl("Gria|Gabra|Glra|Grin|Grn|Gria|Gabra|Glra|Grin|Grik|Grm|Gabbr|Gabrb|Gabrg|Gabrd", Gene) & Gene %in% hits$Gene)  %>%
  ggplot(aes(
    x = as.factor(n_position),
    y = cleaned_glycosignal,
    colour = microbiome,
    group = Modified.Peptide
  )) +
  # geom_boxplot() +
  geom_point(position = position_dodge(width = 0.7), alpha = 0.5, size = 2) +
  theme(legend.position = "right") +
  scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
  facet_wrap(~Gene, scales = "free", ncol = 4) +
  labs(x = "glycopeptides per site")
```

## All hits

This writes all proteins with at least one hit into a pdf for exploration

```{r, eval = F}
hits <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "glyco" & hit == "hit") %>%
  distinct(Gene) %>%
  filter(!is.na(Gene))

FCs <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(dataset == "glyco" & contrast == "community-germfree") %>%
  mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>%
  distinct(glycoID, logFC, hit)

pdf("20230417_mousebrains_hits_globalplots.pdf", width = 15, height = 8)
for (hit in hits$Gene) {
  p <- res_microbiome_mousebrain_Nglyco$psm_anno %>%
    filter(Gene == hit) %>%
    mutate(glycoID = paste0(Protein.ID, "_", Modified.Peptide, "_", Observed.Modifications)) %>%
    left_join(FCs, by = "glycoID") %>%
    filter(!is.na(Gene)) %>%
    ungroup() %>%
    arrange(abs(logFC)) %>%
    ggplot(aes(x = as.factor(n_position), y = cleaned_glycosignal, colour = microbiome, group = reorder(Modified.Peptide, -abs(logFC)))) +
    # geom_boxplot() +
    geom_point(position = position_dodge(width = 0.7), alpha = 1, size = 3) +
    theme(legend.position = "right") +
    scale_colour_manual(values = c("germfree" = "#7D9FC5", "community" = "#C11555", "monocolonized" = "darkseagreen")) +
    facet_wrap(~Gene, scales = "free", ncol = 2) +
    labs(
      x = "N position in protein", y = "log2 normalised glycopeptide intensity",
      title = unique(na.omit(res_microbiome_mousebrain_Nglyco$limma_results$Protein.Description[res_microbiome_mousebrain_Nglyco$limma_results$Gene == hit]))
    )
  plot(p)
}
dev.off()

rm(hits, hit, p, FCs)
```

# 8. Glycosyltrasnferases - Ortholog comparison

Check human glycosylases

```{r}
conversion <- distinct(res_microbiome_mousebrain_Nglyco$limma_results, Gene)
conversion <- orthologs(genes = conversion$Gene, human = FALSE, species = "Mus musculus")

res_microbiome_mousebrain_Nglyco$limma_results <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  left_join(conversion %>% select(Gene = symbol, human_symbol))

glycosltransferases <- read_tsv("data/paper/uniprot/220926_human_glycosltransferases.txt") %>%
  select(Gene = `Approved symbol`)

 res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(human_symbol %in% glycosltransferases$Gene & dataset == "glycoproteome" & hit == "hit") %>%
  distinct(Gene)

res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(human_symbol %in% glycosltransferases$Gene & dataset == "full proteome" & hit == "hit") %>%
  distinct(Gene)

```

## ALG 12

accumulation of GlcNAc2Man5–7 and decreased levels of GlcNAc2Man8–9, which indicated impaired ALG12 enzymatic activity. 

```{r}
alg_glycan_pattern <- res_microbiome_mousebrain_Nglyco$limma_results %>%
  filter(grepl("HexNAc\\(2\\)Hex\\(9\\) |HexNAc\\(2\\)Hex\\(8\\) |HexNAc\\(2\\)Hex\\(5\\) |HexNAc\\(2\\)Hex\\(6\\) |HexNAc\\(2\\)Hex\\(7\\) ", Observed.Modifications)) %>%
  mutate(classification = ifelse(grepl("Hex\\(9\\)|Hex\\(8\\)", Observed.Modifications), "potentially up", "potentially down"))

alg_glycan_pattern %>%
  # filter(hit == "hit") %>%
  ggplot(aes(x = Observed.Modifications, y = logFC, colour = classification)) +
  geom_boxplot() +
  geom_hline(yintercept = 0) +
  facet_wrap(~contrast) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
``` 

# Save

```{r, eval =F}
save(res_microbiome_mousebrain_Nglyco, file = paste0("data/paper/microbiome/", Sys.Date(), "res_microbiome_mousebrain_Nglyco.RData"))

save(res_microbiome_Nglyco_ER, file = paste0("data/paper/microbiome/", Sys.Date(), "res_microbiome_mousebrain_Nglyco_enrichments.RData"))
```
